// @msdyn365-commerce-modules/utilities/src/labeled-dropdown/LabeledDropdown.props.ts

export interface ILabeledDropdownOption {
    key: string;
    value: string;
}

export interface ILabeledDropdownOnChangeNotification {
    selectedOption: ILabeledDropdownOption;
}

/**
 * Labeled Dropdown properties.
 */
export interface ILabeledDropdownProps extends React.DOMAttributes<HTMLElement> {
    /** Label to display (required) */
    labelText: string;
    /** ID of dropdown (required) */
    dropdownId: string;
    /** Options (required) */
    dropdownOptions: ILabeledDropdownOption[];
    /** Selected option (required) */
    selectedOption: ILabeledDropdownOption;
    /** Sets wether the slected state is controlled by parent or itself */
    controlledByParent: boolean;

    /** Classname for label component (optional) */
    labelClassname?: string;
    /** Color for toggle component (optional, defaults to 'link') */
    toggleColor?: string;
    /** Classname for dropdown component (optional) */
    dropdownClassname?: string;

    /** Function called the select updates */
    onSelectOption(notification: ILabeledDropdownOnChangeNotification): void;
}
