// @msdyn365-commerce-modules/utilities/src/labeled-dropdown/LabeledDropdown.examples.tsx
import * as React from 'react';
import LabeledDropdown from './LabeledDropdown';
import { ILabeledDropdownOnChangeNotification } from './LabeledDropdown.props';

/**
 * Labeled Dropdown example
 */
export class LabeledDropdownExample extends React.Component {
    constructor(props: {}) {
        super(props);
        this._onSelectedOption = this._onSelectedOption.bind(this);
    }

    public render(): JSX.Element {
        const options = [
            {key: 'a', value: 'apple'},
            {key: 'b', value: 'banana'},
            {key: 'c', value: 'cucumber'},
            {key: 'delta', value: 'pear'},
            {key: 'e', value: 'tomato'},
        ];

        return (
            <LabeledDropdown
                dropdownOptions={options}
                selectedOption={options[1]}
                onSelectOption={this._onSelectedOption}
                labelText='fruit'
                dropdownId='testDropdownId'
            />
        );
    }

    private _onSelectedOption = (notification: ILabeledDropdownOnChangeNotification) => {
        alert(`Selected option key=${notification.selectedOption.key} value=${notification.selectedOption.value}`);
    }
}
