// @msdyn365-commerce-modules/core-components/src/modules/default-container-2-slots/default-container-2-slots.props.autogenerated.ts
/**
 * Copyright (c) 2018 Microsoft Corporation
 * IDefaultContainer2Slots containerModule Interface Properties
 * THIS FILE IS AUTO-GENERATED - MANUAL MODIFICATIONS WILL BE LOST
 */

import * as Msdyn365 from '@msdyn365-commerce/core';
import * as React from 'react';

export const enum containerType {
    fluid = 'fluid',
    container = 'container'
}

export const enum xsmall {
    col_auto_auto = 'col_auto_auto',
    col_12_12 = 'col_12_12',
    col_9_3 = 'col_9_3',
    col_8_4 = 'col_8_4',
    col_6_6 = 'col_6_6',
    col_4_8 = 'col_4_8',
    col_3_9 = 'col_3_9'
}

export const enum small {
    col_auto_auto = 'col_auto_auto',
    col_12_12 = 'col_12_12',
    col_9_3 = 'col_9_3',
    col_8_4 = 'col_8_4',
    col_6_6 = 'col_6_6',
    col_4_8 = 'col_4_8',
    col_3_9 = 'col_3_9'
}

export const enum medium {
    col_auto_auto = 'col_auto_auto',
    col_12_12 = 'col_12_12',
    col_9_3 = 'col_9_3',
    col_8_4 = 'col_8_4',
    col_6_6 = 'col_6_6',
    col_4_8 = 'col_4_8',
    col_3_9 = 'col_3_9'
}

export const enum large {
    col_auto_auto = 'col_auto_auto',
    col_12_12 = 'col_12_12',
    col_9_3 = 'col_9_3',
    col_8_4 = 'col_8_4',
    col_6_6 = 'col_6_6',
    col_4_8 = 'col_4_8',
    col_3_9 = 'col_3_9'
}

export interface IDefaultContainer2SlotsConfig extends Msdyn365.IModuleConfig {
    heading?: IHeadingData;
    containerType?: containerType;
    xsmall?: xsmall;
    small?: small;
    medium?: medium;
    large?: large;
    className?: string;
}

export const enum HeadingTag {
    h1 = 'h1',
    h2 = 'h2',
    h3 = 'h3',
    h4 = 'h4',
    h5 = 'h5',
    h6 = 'h6'
}

export interface IHeadingData {
    headingText?: string;
    tag?: HeadingTag;
}

export interface IDefaultContainer2SlotsProps<T> extends Msdyn365.IModule<T> {
    config: IDefaultContainer2SlotsConfig;
    slots: {
        content: React.ReactNode[];
    };
}
