// @msdyn365-commerce-modules/utilities/src/fade/Fade.examples.tsx
import * as React from 'react';
import Fade, { IFadeState } from './Fade';
import { IFadeProps } from './Fade.props';

/**
 * Fade Examples
 */
export class FadeExample extends React.Component<IFadeProps, IFadeState> {
    public props: IFadeProps;
    public state: IFadeState;

    constructor(props: IFadeProps) {
        super(props);
        this.props = props;
        this.toggle = this.toggle.bind(this);
        this.state = {
            fadeIn: false
        };
    }

    public render(): JSX.Element {
        return (
            <div>
                <button onClick={this.toggle} aria-controls='idFadeHeading' aria-expanded={this.state.fadeIn}>
                    Toggle Fade
                </button>
                <Fade in={this.state.fadeIn} tag='h5' className='mt-3'>
                    This content will fade in and out as the button is pressed
                </Fade>
            </div>
        );
    }

    private toggle = () => {
        this.setState((prevState: IFadeState) => ({
            fadeIn: !prevState.fadeIn
        }));
    }
}
