// @msdyn365-commerce-modules/account-management/dist/lib/modules/account-welcome-tile/account-welcome-tile.js
import classnames from 'classnames';
import * as React from 'react';
import * as Msdyn365 from '@msdyn365-commerce/core';
class AccountWelcomeTile extends React.PureComponent {
    constructor() {
        super(...arguments);
        this.handleHeadingChange = (event) => this.props.config.heading.text = event.target.value;
    }
    render() {
        const { config, data } = this.props;
        const { className } = config;
        const accountInfo = data.accountInformation.result;
        const heading = {
            ...config.heading,
            ...{
                text: `${config.heading.text} ${accountInfo && accountInfo.FirstName ? accountInfo.FirstName : ''}`
            }
        };
        const viewProps = {
            ...this.props,
            className: className,
            AccountWelcomeTile: {
                moduleProps: this.props,
                className: classnames('ms-account-welcome-tile', config.className)
            },
            heading: heading && heading.text && (React.createElement(Msdyn365.Text, { className: 'ms-account-welcome-tile__heading', tag: heading.tag, text: heading.text, editProps: { onEdit: this.handleHeadingChange, requestContext: this.props.context.request } })),
        };
        return this.props.renderView(viewProps);
    }
}
export default AccountWelcomeTile;
