// @msdyn365-commerce-modules/core-components/dist/lib/modules/default-container-3-slots/default-container-3-slots.js
/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */
import { LogLevel } from '@msdyn365-commerce/telemetry-internal';
import React from 'react';
import { Container, HeadingComponent } from '../../components/base-components';
import { getColClassArray, showErrors } from '../../utils/classnames';
class Default3SlotContainer extends React.PureComponent {
    constructor(props) {
        super(props);
        this.handleTextChange = (event) => (this.props.config.heading.headingText = event.target.value);
        this._getColClassName = () => {
            const { config } = this.props;
            const classNames = getColClassArray(config);
            if (!classNames.length) {
                return {
                    left: ['col'],
                    center: ['col'],
                    right: ['col']
                };
            }
            return classNames.reduce((memo, [left, center, right]) => {
                memo.left.push(left);
                memo.center.push(center);
                memo.right.push(right);
                return memo;
            }, { left: [], center: [], right: [] });
        };
    }
    render() {
        const { slots, telemetry, context, config, data } = this.props;
        const slotsClassNames = this._getColClassName();
        const content = slots && slots.content;
        telemetry.log(LogLevel.Debug, "Default 3 Slot Container rendering for '{id}/{typeName}'", {
            values: [this.props.id, this.props.typeName]
        });
        if (content && content.length > 3) {
            telemetry.log(LogLevel.Warning, 'Default 3 Slot Container passed more than 3 slots');
        }
        const heading = config.heading || data.heading;
        const headingProps = {
            ...heading,
            handleTextChange: this.handleTextChange,
            requestContext: this.props.context.request
        };
        return (React.createElement(Container, { containerType: config.containerType, className: config.className, typeName: this.props.typeName, id: this.props.id, request: this.props.context.request },
            heading && React.createElement(HeadingComponent, Object.assign({ className: 'container-heading' }, headingProps)),
            content && content[0] && React.createElement("div", { className: slotsClassNames.left.join(' ') }, content[0]),
            content && content[1] && React.createElement("div", { className: slotsClassNames.center.join(' ') }, content[1]),
            content && content[2] && React.createElement("div", { className: slotsClassNames.right.join(' ') }, content[2]),
            React.createElement(React.Fragment, null, showErrors(context) &&
                content &&
                content.slice(2, content.length - 1).map((slot, index) => {
                    return (React.createElement("div", { className: 'col-12', key: index },
                        React.createElement("div", { className: 'card' },
                            React.createElement("div", { className: 'card-header text-white bg-danger' }, "Slot exceeded maximum number of slots supported by container. (3)"))));
                }))));
    }
}
export default Default3SlotContainer;
