// @msdyn365-commerce-modules/order-management/src/modules/order-details/order-details.props.autogenerated.ts
/**
 * Copyright (c) 2018 Microsoft Corporation
 * IOrderDetails contentModule Interface Properties
 * THIS FILE IS AUTO-GENERATED - MANUAL MODIFICATIONS WILL BE LOST
 */

import * as Msdyn365 from '@msdyn365-commerce/core';

export const enum addressOptions {
    showAddressWithSingleLine = 'showAddressWithSingleLine',
    showAddressWithMultiplelines = 'showAddressWithMultiplelines'
}

export interface IOrderDetailsConfig extends Msdyn365.IModuleConfig {
    heading?: IHeadingData;
    contactNumber?: string;
    imageSettings?: Msdyn365.IImageSettings;
    showChannelInfo?: boolean;
    addressOptions?: addressOptions;
    className?: string;
}

export interface IOrderDetailsResources {
    loadingLabel: string;
    genericErrorMessage: string;
    orderItemLabel: string;
    orderItemsLabel: string;
    orderIdLabel: string;
    confirmationIdLabel: string;
    orderSummaryTitle: string;
    paymentMethodsTitle: string;
    shippingAddressTitle: string;
    noSalesOrderDetailsText: string;
    creditCardEndingLabel: string;
    giftCardEndingLabel: string;
    amountCoveredLabel: string;
    loyaltyCardUsedLabel: string;
    cashUsedLabel: string;
    customerAccountUsedLabel: string;
    freePriceText: string;
    orderSummaryItemsTotalLabel: string;
    orderSummaryShippingFeeLabel: string;
    orderSummaryTaxLabel: string;
    orderSummaryGrandTotalLabel: string;
    salesLineQuantityText: string;
    processingLabel: string;
    needHelpLabel: string;
    helpLineNumberLabel: string;
    helpLineContactAriaLabel: string;
    pickedUpSalesLines: string;
    deliveredSalesLines: string;
    carryOutSalesLines: string;
    orderStatusReadyForPickup: string;
    orderStatusProcessing: string;
    orderStatusShipped: string;
    orderStatusPickedUp: string;
    orderStatusCanceled: string;
    phoneAriaLabel: string;
    buyItAgainLabel: string;
    buyItAgainAriaLabel: string;
    shipToLabel: string;
    storeLabel: string;
    productDimensionTypeColor: string;
    productDimensionTypeSize: string;
    productDimensionTypeStyle: string;
    trackingLabel: string;
    trackingComingSoonLabel: string;
    trackingAriaLabel: string;
    pointsEarnedLabel: string;
    receiptEmailMessage: string;
    receiptIdLabel: string;
    backToShopping: string;
    configString: string;
    phoneLabel: string;
    orderSummaryHeading: string;
    discountStringText: string;
    posChannelNameText: string;
    onlineStoreChannelNameText: string;
}

export const enum HeadingTag {
    h1 = 'h1',
    h2 = 'h2',
    h3 = 'h3',
    h4 = 'h4',
    h5 = 'h5',
    h6 = 'h6'
}

export interface IHeadingData {
    text: string;
    tag?: HeadingTag;
}

export interface IOrderDetailsProps<T> extends Msdyn365.IModule<T> {
    resources: IOrderDetailsResources;
    config: IOrderDetailsConfig;
}
