// @msdyn365-commerce-modules/retail-actions/dist/lib/get-full-products-by-refine-search-text.js
import { createObservableDataAction } from '@msdyn365-commerce/core';
import { refineSearchByTextAsync } from '@msdyn365-commerce/retail-proxy/dist/DataActions/ProductsDataActions.g';
import { parseSearchData } from './utilities/input-data-parser';
import { QueryResultSettingsProxy } from './utilities/QueryResultSettingsProxy';
import { getRefinedFullProducts } from './utilities/refiner-utils';
import { getProductDetailsCriteriaFromActionInput } from './utilities/utils';
/**
 * Input for refining products returned by the search text.
 */
export class FullProductsRefineSearchByTextInput {
    constructor(queryResultSettingsProxy, searchText, channelId, refinementCriteria, catalogId, criteria) {
        this.getCacheKey = () => `FullProductsRefineSearchByTextInputCache`;
        this.getCacheObjectType = () => 'FullProductsRefineSearchByTextInput';
        this.dataCacheType = () => 'none';
        this.queryResultSettingsProxy = queryResultSettingsProxy;
        this.searchText = searchText || '';
        this.channelId = channelId;
        this.refinementCriteria = refinementCriteria || [];
        this.catalogId = catalogId || 0;
        this.productDetailsCriteria = criteria;
    }
}
/**
 * Creates the input required to make the core action calls
 */
export const createFullProductsRefineSearchByTextInput = (inputData) => {
    const refinementCriteria = inputData.config && inputData.config.refinementCriteria;
    const queryResultSettingsProxy = QueryResultSettingsProxy.fromInputData(inputData);
    if (!Array.isArray(refinementCriteria)) {
        return new FullProductsRefineSearchByTextInput(queryResultSettingsProxy);
    }
    const searchInputData = parseSearchData(inputData);
    const catalogId = inputData.requestContext.apiSettings.catalogId;
    const productDetailsCriteria = getProductDetailsCriteriaFromActionInput(inputData);
    return new FullProductsRefineSearchByTextInput(queryResultSettingsProxy, searchInputData.q, searchInputData.channelId, refinementCriteria, catalogId, productDetailsCriteria);
};
/**
 * Calls the refine-search-by-text action.
 * Based on search result calls get-full-products to get all the product details.
 */
export function getFullProductsByRefineSearchTextAction(input, ctx) {
    return getRefinedFullProducts(input, ctx, 
    // @ts-ignore: Promise.then typing conflict
    () => {
        return refineSearchByTextAsync({ callerContext: ctx, queryResultSettings: input.queryResultSettingsProxy.QueryResultSettings }, input.channelId || 0, input.catalogId, input.searchText, input.refinementCriteria);
    });
}
export default createObservableDataAction({
    id: '@msdyn365-commerce-modules/retail-actions/get-full-products-by-refine-search-text',
    action: getFullProductsByRefineSearchTextAction,
    input: createFullProductsRefineSearchByTextInput,
    isBatched: false
});
