// @msdyn365-commerce-modules/retail-actions/dist/lib/get-refiners-by-category.js
import { createObservableDataAction } from '@msdyn365-commerce/core';
import getCurrentCategory, { CurrentCategoryInput } from './get-current-category';
import { getRefinersByCategoryAsync } from '@msdyn365-commerce/retail-proxy/dist/DataActions/ProductsDataActions.g';
import { QueryResultSettingsProxy } from './utilities/QueryResultSettingsProxy';
/**
 * Action Input class for the getRefinersByCategory data action
 */
export class RefinersByCategoryInput {
    constructor(category, queryResultSettingsProxy, catalogId) {
        this.getCacheKey = () => `${this.currentCategory.getCacheKey()}|${this.catalogId}|${this.queryResultSettingsProxy.cacheKeyHint}`;
        this.getCacheObjectType = () => 'ProductRefiner';
        this.shouldCacheOutput = () => true;
        this.queryResultSettingsProxy = queryResultSettingsProxy;
        this.catalogId = catalogId || 0;
        this.currentCategory = category;
    }
}
/**
 * Creates the input required to make the retail api call
 */
export const createRefinersByCategoryInput = (inputData) => {
    if (inputData && inputData.requestContext) {
        const catalogId = inputData.config ? Number(inputData.config.catalogId) : 0;
        const currentCategory = new CurrentCategoryInput(inputData.requestContext);
        return new RefinersByCategoryInput(currentCategory, QueryResultSettingsProxy.fromInputData(inputData), Number.isNaN(catalogId) ? 0 : catalogId);
    }
    throw new Error('Please specify categoryId query string in request.');
};
/**
 * Calls the Retail API and returns all refiners by category
 */
export async function getRefinersByCategoryAction(input, ctx) {
    let categoryId = input.currentCategory.categoryId;
    if (input.currentCategory.categorySlug && !categoryId) {
        const category = await getCurrentCategory(input.currentCategory, ctx);
        if (!category) {
            ctx.trace('[getRefinersByCategoryAction] Unable to find category');
            return [];
        }
        categoryId = category.RecordId;
    }
    return getRefinersByCategoryAsync({ callerContext: ctx, queryResultSettings: input.queryResultSettingsProxy.QueryResultSettings }, input.catalogId, categoryId || 0);
}
export default createObservableDataAction({
    id: '@msdyn365-commerce-modules/retail-actions/get-refiners-by-category',
    action: getRefinersByCategoryAction,
    input: createRefinersByCategoryInput
});
