// @msdyn365-commerce-modules/retail-actions/dist/lib/get-categories.js
import { createObservableDataAction } from '@msdyn365-commerce/core';
import { getCategoriesAsync } from '@msdyn365-commerce/retail-proxy/dist/DataActions/CategoriesDataActions.g';
/**
 * Input for get-categories data action
 */
export class CategoriesInput {
    constructor(context, mappedToHierarchy, maxItems) {
        this.getCacheKey = () => `${this.channelId}|top-${this.maxItems || 250}`;
        this.getCacheObjectType = () => (this._mappedToHierarchy ? 'CategoryHierarchy' : 'Category');
        this.dataCacheType = () => 'application';
        this._mappedToHierarchy = mappedToHierarchy;
        this.maxItems = maxItems || 250;
        this.channelId = context && context.apiSettings && context.apiSettings.channelId ? Number(context.apiSettings.channelId) : 0;
    }
}
/**
 * Creates the input required to make the retail api call
 */
export const createCategoriesInput = (inputData) => {
    const topItems = inputData.config && inputData.config.topCategories && parseInt(inputData.config.topCategories, 10);
    return new CategoriesInput(inputData.requestContext, false, topItems);
};
/**
 * Calls the Retail API and returns all the categories as a flat list
 */
export async function getCategoryAction(input, ctx) {
    return (getCategoriesAsync({ callerContext: ctx }, input.channelId));
}
export default createObservableDataAction({
    id: '@msdyn365-commerce-modules/retail-actions/get-categories',
    action: getCategoryAction,
    input: createCategoriesInput
});
