// @msdyn365-commerce-modules/retail-actions/dist/lib/get-search-configuration.js
import { createObservableDataAction } from '@msdyn365-commerce/core';
import { buildCacheKey } from '.';
import { getSearchConfigurationAsync } from '@msdyn365-commerce/retail-proxy/dist/DataActions/StoreOperationsDataActions.g';
/**
 * Action Input class for the getSearchConfiguration data action
 */
export class SearchConfigurationInput {
    constructor(apiSettings) {
        this.getCacheKey = () => buildCacheKey('SearchConfig', this.apiSettings);
        this.getCacheObjectType = () => 'SearchConfiguration';
        this.dataCacheType = () => 'request';
        this.apiSettings = apiSettings;
    }
}
/**
 * Creates the input required to make the retail api call
 */
export const createSearchConfigurationInput = (inputData) => {
    return new SearchConfigurationInput(inputData.requestContext.apiSettings);
};
/**
 * Calls the Retail API
 */
export async function getSearchConfigurationAction(input, ctx) {
    return getSearchConfigurationAsync({ callerContext: ctx });
}
export default createObservableDataAction({
    id: '@msdyn365-commerce-modules/retail-actions/get-search-configurations',
    action: getSearchConfigurationAction,
    input: createSearchConfigurationInput
});
