// @msdyn365-commerce-modules/retail-actions/dist/lib/get-organization-users.js
import { createObservableDataAction } from '@msdyn365-commerce/core';
import { getCurrentAsync, getUsersAsync } from '@msdyn365-commerce/retail-proxy/dist/DataActions/BusinessPartnersDataActions.g';
import { buildCacheKey } from './utilities/utils';
/**
 * Input class for get customer data action
 */
export class OrganizationUsersInput {
    constructor(apiSettings) {
        this.getCacheKey = () => buildCacheKey('OrganizationUsers', this.apiSettings);
        this.getCacheObjectType = () => 'BusinessPartner';
        this.dataCacheType = () => 'request';
        this.apiSettings = apiSettings;
    }
}
export const createGetOrganizationUsersInput = (inputData) => {
    const { requestContext } = inputData;
    // Check if the user is authenticated.
    if (!requestContext.user.isAuthenticated) {
        throw new Error('User is not logged in.');
    }
    return new OrganizationUsersInput(requestContext.apiSettings);
};
export async function getOrganizationUsers(input, ctx) {
    return getCurrentAsync({ callerContext: ctx })
        .then(businessPartner => {
        return getUsersAsync({ callerContext: ctx }, businessPartner.BusinessPartnerId)
            .then((users) => {
            return users;
        })
            .catch(error => {
            ctx.telemetry.exception(error);
            ctx.trace('[get-organization-users] Unable to load users');
            throw new Error('[get-organization-users] Unable to load users');
        });
    })
        .catch(error => {
        ctx.telemetry.exception(error);
        ctx.telemetry.debug('[get-organization-users] Unable to get BusinessPartner');
        throw new Error('[get-organization-users] Unable to get BusinessPartner');
    });
}
export default createObservableDataAction({
    id: 'get-organization-users',
    action: getOrganizationUsers,
    input: createGetOrganizationUsersInput
});
