// @msdyn365-commerce-modules/retail-actions/dist/lib/refine-search-by-category.js
import { createObservableDataAction } from '@msdyn365-commerce/core';
import { createRefineSearchByCategoryInput } from '@msdyn365-commerce/retail-proxy/dist/DataActions/ProductsDataActions.g';
import { retailAction } from '@msdyn365-commerce/retail-proxy/dist/retail-action';
import { QueryResultSettingsProxy } from './utilities/QueryResultSettingsProxy';
import { buildCacheKey } from './utilities/utils';
/**
 * Action Input class for the refineSearchByCategory data action
 */
export class RefineSearchByCriteriaInput {
    constructor(queryResultSettingsProxy, apiSettings, categoryId, channelId, refinementCriteria, catalogId) {
        this.getCacheKey = () => buildCacheKey(`RefineSearchByCriteriaInputCache`, this.apiSettings);
        this.getCacheObjectType = () => 'RefineSearchByCriteriaInput';
        this.dataCacheType = () => 'request';
        this.categoryId = categoryId;
        this.channelId = channelId;
        this.refinementCriteria = refinementCriteria || [];
        this.queryResultSettingsProxy = queryResultSettingsProxy;
        this.catalogId = catalogId || 0;
        this.apiSettings = apiSettings;
    }
}
/**
 * createInput method for the refineSearchByCategory data action
 */
export const createInput = (inputData) => {
    const refinementCriteria = inputData.config && inputData.config.refinementCriteria;
    if (inputData && inputData.requestContext && inputData.requestContext.query && inputData.requestContext.query.categoryId) {
        return createRefineSearchByCategoryInput(QueryResultSettingsProxy.fromInputData(inputData).QueryResultSettings, +inputData.requestContext.apiSettings.channelId, inputData.config ? Number(inputData.config.catalogId) : 0, +inputData.requestContext.query.categoryId, !Array.isArray(refinementCriteria) ? [] : refinementCriteria.map((refinementCriterion) => refinementCriterion));
    }
    throw new Error('Please specify categoryId, refinerId, and refinerSourceValue query string in request.');
};
/**
 * The RefineSearchByCategory data action
 * Uses categoryId query string parameter to call the RefinerSearchByCategory
 * RetailServer API to get information about products
 */
export default createObservableDataAction({
    id: '@msdyn365-commerce-modules/retail-actions/regin-search-by-category',
    action: retailAction,
    input: createInput
});
