// @msdyn365-commerce-modules/cookie-compliance/dist/lib/modules/cookie-compliance/components/links.js
import { getPayloadObject, getTelemetryAttributes, onTelemetryClick } from '@msdyn365-commerce-modules/utilities';
import * as Msdyn365 from '@msdyn365-commerce/core';
import * as React from 'react';
export class CookieComplianceLinks extends React.PureComponent {
    constructor() {
        super(...arguments);
        this._mapEditableLinks = (linkdata) => {
            if (!linkdata || linkdata.length === 0) {
                return null;
            }
            const editableLinks = [];
            linkdata.forEach((link, index) => {
                const payLoad = getPayloadObject('click', this.props.telemetryContent, '', '');
                const linkText = link.linkText ? link.linkText : '';
                payLoad.contentAction.etext = linkText;
                const attributes = getTelemetryAttributes(this.props.telemetryContent, payLoad);
                const editableLink = {
                    ariaLabel: link.ariaLabel,
                    className: 'link',
                    linkText: link.linkText,
                    linkUrl: link.linkUrl.destinationUrl,
                    openInNewTab: link.openInNewTab,
                    role: 'button',
                    additionalProperties: attributes,
                    onClick: onTelemetryClick(this.props.telemetryContent, payLoad, linkText)
                };
                editableLinks.push(editableLink);
            });
            return editableLinks;
        };
    }
    render() {
        const editableLinks = this._mapEditableLinks(this.props.links);
        return (React.createElement("span", { className: 'ms-cookie-compliance__cta-layer' }, editableLinks && editableLinks.length ?
            React.createElement(Msdyn365.Links, { links: editableLinks, editProps: { onTextChange: this.props.onTextChange, requestContext: this.props.requestContext } })
            : null));
    }
}
export default CookieComplianceLinks;
