// @msdyn365-commerce/runtime-internal/dist/lib/app-initialization/initialization.js
/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */
import { MSDyn365Commerce, msdyn365Commerce } from '@msdyn365-commerce/core-internal';
// tslint:disable:no-invalid-this
export const initializeApp = (bindings) => {
    /**
     * Why we are using 'Prototype'. MSDyn365Commerce is defined in core packages and used in modules and C1 development.
     * whereas all registration, build information, commerce details etc available in runtime-internal package.
     * prototype help us declaring properties in core package and implementation in runtime-internal package.
     *
     * Following properties are not accessible by module or c1 developer.
     */
    MSDyn365Commerce.prototype.bindings = bindings;
    MSDyn365Commerce.prototype.moduleBinder = function (name) {
        return this.bindings && this.bindings.modules && this.bindings.modules[name && name.toLowerCase()];
    };
    MSDyn365Commerce.prototype.viewBinder = function (name) {
        return this.bindings && this.bindings.views && this.bindings.views[name];
    };
    MSDyn365Commerce.prototype.componentBinder = function (name) {
        return this.bindings && this.bindings.components && this.bindings.components[name];
    };
    MSDyn365Commerce.prototype.dataAction = function (name) {
        return this.bindings && this.bindings.dataActions && this.bindings.dataActions[name];
    };
    // tslint:disable-next-line: no-any
    MSDyn365Commerce.prototype.getConnector = function (name) {
        return this.bindings && this.bindings.connectors && this.bindings.connectors[name];
    };
    MSDyn365Commerce.prototype.getAllModuleBinder = function () {
        return this.bindings && this.bindings.modules && Object.values(this.bindings.modules);
    };
    MSDyn365Commerce.prototype.updatePrivateVariable = function () {
        // @ts-ignore
        this._env = process.env.NODE_ENV;
    };
    msdyn365Commerce.updatePrivateVariable();
};
