// @msdyn365-commerce-modules/account-management/src/modules/loyalty-sign-up/loyalty-sign-up.props.autogenerated.ts
/**
 * Copyright (c) 2018 Microsoft Corporation
 * ILoyaltySignUp contentModule Interface Properties
 * THIS FILE IS AUTO-GENERATED - MANUAL MODIFICATIONS WILL BE LOST
 */

import * as Msdyn365 from '@msdyn365-commerce/core';

export interface ILoyaltySignUpConfig extends Msdyn365.IModuleConfig {
    heading?: IHeadingData;
    memberHeading?: IMemberHeadingData;
    className?: string;
}

export interface ILoyaltySignUpResources {
    loyaltyProgramText: string;
    loyaltyTermsLinkText: string;
    joinLoyaltyButtonText: string;
    viewLoyaltyPageText: string;
}

export const enum HeadingTag {
    h1 = 'h1',
    h2 = 'h2',
    h3 = 'h3',
    h4 = 'h4',
    h5 = 'h5',
    h6 = 'h6'
}

export interface IHeadingData {
    text: string;
    tag?: HeadingTag;
}

export const enum MemberHeadingTag {
    h1 = 'h1',
    h2 = 'h2',
    h3 = 'h3',
    h4 = 'h4',
    h5 = 'h5',
    h6 = 'h6'
}

export interface IMemberHeadingData {
    text: string;
    tag?: MemberHeadingTag;
}

export interface ILoyaltySignUpProps<T> extends Msdyn365.IModule<T> {
    resources: ILoyaltySignUpResources;
    config: ILoyaltySignUpConfig;
}
