// @msdyn365-commerce-modules/utilities/src/tooltip/Tooltip.examples.tsx
import * as React from 'react';
import Tooltip from './Tooltip';

export interface ITooltipExampleState {
    tooltipOpen1: boolean;
    tooltipOpen2: boolean;
}

/**
 * Tooltip examples
 */
export default class TooltipExample extends React.Component<{}, ITooltipExampleState> {
    private ref: React.RefObject<HTMLAnchorElement>;

    constructor(props: {}) {
        super(props);

        this.ref = React.createRef();
        this._toggle1 = this._toggle1.bind(this);
        this._toggle2 = this._toggle2.bind(this);
        this.state = {
          tooltipOpen1: false,
          tooltipOpen2: false
        };
      }

    public render(): JSX.Element {
        return (
            <div>
                <div>
                    There is a <a href='#tooltip1' id='TooltipExample' aria-describedby='TooltipOne'>tooltip</a> in here.
                </div>
                <Tooltip placement='right' isOpen={this.state.tooltipOpen1} id='TooltipOne' target='TooltipExample' toggle={this._toggle1}>
                    Test tooltip
                </Tooltip>
                <div>
                    A second one below <a href='#tooltip2' id='TooltipExample2' ref={this.ref} aria-describedby='TooltipTwo'>this one</a>!
                </div>
                <Tooltip placement='bottom' isOpen={this.state.tooltipOpen2} id='TooltipTwo' target={this.ref} hideArrow toggle={this._toggle2}>
                    A second tooltip
                </Tooltip>
            </div>
        );
    }

    private _toggle1(): void {
        this.setState({
            tooltipOpen1: !this.state.tooltipOpen1
        });
    }

    private _toggle2(): void {
        this.setState({
            tooltipOpen2: !this.state.tooltipOpen2
        });
    }
}
