// @msdyn365-commerce-modules/wishlist/dist/lib/modules/wishlist-items/actions/get-product-availabilities-wishlist-items.js
import { buildCacheKey, mapProductInventoryInformation } from '@msdyn365-commerce-modules/retail-actions';
import { createObservableDataAction } from '@msdyn365-commerce/core';
import { getEstimatedAvailabilityAsync } from '@msdyn365-commerce/retail-proxy/dist/DataActions/ProductsDataActions.g';
import { ActiveWishlistInput, getActiveWishlistItems } from './get-items-in-wishlists';
export class ProductAvailabilitiesForWishlistItems {
    constructor(apiSettings) {
        this.getCacheKey = () => buildCacheKey(`ActiveWishlistItemsAvailability`, this.apiSettings);
        this.getCacheObjectType = () => `ActiveWishlistItemsAvailability`;
        this.dataCacheType = () => 'request';
        this.apiSettings = apiSettings;
    }
}
const createInput = (inputData) => {
    return new ProductAvailabilitiesForWishlistItems(inputData.requestContext.apiSettings);
};
export async function getAvailabilitiesForWishlistItems(input, ctx) {
    if (!input) {
        throw new Error('[getAvailabilitiesForWishlistItems]No valid Input was provided, failing');
    }
    const products = await getActiveWishlistItems(new ActiveWishlistInput(), ctx);
    if (!products) {
        ctx.trace('[getAvailabilitiesForWishlistItems] Not able to get products in wishlist');
        return [];
    }
    const validProducts = [];
    for (const product of products) {
        if (product.ProductDetails && product.ProductDetails.RecordId) {
            validProducts.push(product.ProductDetails.RecordId);
        }
    }
    if (validProducts.length === 0) {
        ctx.trace('[getAvailabilitiesForWishlistItems] No products in wishlist');
        return [];
    }
    const productAvailabilites = await getEstimatedAvailabilityAsync({ callerContext: ctx }, { ProductIds: validProducts, DefaultWarehouseOnly: true });
    if (productAvailabilites && productAvailabilites) {
        return mapProductInventoryInformation(ctx, productAvailabilites?.ProductWarehouseInventoryAvailabilities);
    }
    else {
        ctx.trace('[getAvailabilitiesForWishlistItems] unable to get availabilites for product');
        return [];
    }
}
export default createObservableDataAction({
    id: '@msdyn365-commerce-modules/wishlist/wishlist-items/get-product-availabilities-wishlist-items',
    action: getAvailabilitiesForWishlistItems,
    input: createInput
});
