// @msdyn365-commerce-modules/fabrikam-design-kit/dist/lib/modules/fabrikam/views/order-details.view.js
import { Module, Node } from '@msdyn365-commerce-modules/utilities';
import * as React from 'react';
export const OrderInfomation = ({ orderInformationProps, salesId, receiptId, createdDate, count, amount, channelReferenceId, channelName, channelAddress }) => (React.createElement(Node, Object.assign({}, orderInformationProps),
    channelName,
    channelAddress,
    salesId,
    receiptId,
    createdDate,
    count,
    amount,
    channelReferenceId));
export const SalesLine = ({ salesLineProps, salesLine, buyAgainButton }) => (React.createElement(Node, Object.assign({}, salesLineProps),
    salesLine,
    buyAgainButton));
export const GroupDelivery = ({ deliveryProps, heading, count, trackingInfo }) => (React.createElement(Node, Object.assign({}, deliveryProps),
    heading,
    count,
    trackingInfo));
export const Group = ({ groupProps, delivery, address, salesLinesProps, salesLines }) => (React.createElement(Node, Object.assign({}, groupProps),
    delivery && React.createElement(GroupDelivery, Object.assign({}, delivery)),
    React.createElement(Node, { className: 'ms-order-details__sales-lines_container' },
        salesLines && (React.createElement(Node, Object.assign({}, salesLinesProps), salesLines.map(salesLine => (React.createElement(React.Fragment, { key: salesLine.data.salesLine.LineId },
            React.createElement(SalesLine, Object.assign({}, salesLine)),
            salesLine.salesStatus))))),
        address)));
export const Groups = ({ groupsProps, groups }) => (React.createElement(Node, Object.assign({}, groupsProps), groups.map((group, index) => (React.createElement(Group, Object.assign({ key: index }, group))))));
export const OrderSummary = ({ orderSummaryProps, heading, subtotal, shipping, tax, totalAmount, earnedPoints }) => (React.createElement(Node, Object.assign({}, orderSummaryProps),
    heading,
    subtotal,
    shipping,
    tax,
    totalAmount,
    earnedPoints));
export const Payment = ({ paymentMethodsProps, title, methods }) => (React.createElement(Node, Object.assign({}, paymentMethodsProps),
    title,
    methods));
export const Help = ({ helpProps, needHelpLabel, helpLineNumberLabel, contactNumber }) => (React.createElement(Node, Object.assign({}, helpProps),
    needHelpLabel,
    helpLineNumberLabel,
    contactNumber));
const OrderDetailsView = ({ moduleProps, heading, alert, loading, orderInfomation, orderSummary, payment, help, groups }) => {
    return (React.createElement(Module, Object.assign({}, moduleProps),
        heading,
        alert,
        loading,
        orderInfomation && React.createElement(OrderInfomation, Object.assign({}, orderInfomation)),
        groups && React.createElement(Groups, Object.assign({}, groups)),
        React.createElement(Node, { className: 'ms-order-details__order-summary-container' },
            orderSummary && React.createElement(OrderSummary, Object.assign({}, orderSummary)),
            payment && React.createElement(Payment, Object.assign({}, payment))),
        help && React.createElement(Help, Object.assign({}, help))));
};
export default OrderDetailsView;
