// @msdyn365-commerce-modules/checkout/src/modules/gift-card-balance-check/gift-card-balance-check.props.autogenerated.ts
/**
 * Copyright (c) 2018 Microsoft Corporation
 * IGiftCardBalanceCheck contentModule Interface Properties
 * THIS FILE IS AUTO-GENERATED - MANUAL MODIFICATIONS WILL BE LOST
 */

import * as Msdyn365 from '@msdyn365-commerce/core';

export const enum showAdditionalFields {
    pin = 'pin',
    expirationDate = 'expirationDate',
    pinAndExpirationDate = 'pinAndExpirationDate'
}

export interface IGiftCardBalanceCheckConfig extends Msdyn365.IModuleConfig {
    className?: string;
    showAdditionalFields?: showAdditionalFields;
}

export interface IGiftCardBalanceCheckResources {
    emptyInputError: string;
    invalidCodeError: string;
    invalidCardInfoError: string;
    invalidCardTypeError: string;
    noCardPinError: string;
    noCardExpError: string;
    noBalanceError: string;
    giftCardFormLabel: string;
    giftCardNumberLabel: string;
    giftCardPinLabel: string;
    giftCardExpLabel: string;
    giftCardPinPlaceholderText: string;
    giftCardExpPlaceholderText: string;
    giftCardAlertLabel: string;
    balanceCheckButton: string;
    giftCardAvailableBalanceLabel: string;
    enteredGiftCardLabel: string;
}

export interface IGiftCardBalanceCheckProps<T> extends Msdyn365.IModule<T> {
    resources: IGiftCardBalanceCheckResources;
    config: IGiftCardBalanceCheckConfig;
}
