// @msdyn365-commerce-modules/account-management/src/modules/account-order-templates-tile/account-order-templates-tile.props.autogenerated.ts
/**
 * Copyright (c) 2018 Microsoft Corporation
 * IAccountOrderTemplatesTile contentModule Interface Properties
 * THIS FILE IS AUTO-GENERATED - MANUAL MODIFICATIONS WILL BE LOST
 */

import * as Msdyn365 from '@msdyn365-commerce/core';

export interface IAccountOrderTemplatesTileConfig extends Msdyn365.IModuleConfig {
    heading: IHeadingData;
    links: ILinksData[];
    className?: string;
    featureName?: string;
}

export interface IAccountOrderTemplatesTileResources {
    accountOrderTemplatesTileDescription: string;
}

export const enum HeadingTag {
    h1 = 'h1',
    h2 = 'h2',
    h3 = 'h3',
    h4 = 'h4',
    h5 = 'h5',
    h6 = 'h6'
}

export interface IHeadingData {
    text: string;
    tag?: HeadingTag;
}

export interface ILinksData {
    linkText?: string;
    linkUrl: Msdyn365.ILinkData;
    ariaLabel?: string;
    openInNewTab?: boolean;
}

export interface IAccountOrderTemplatesTileProps<T> extends Msdyn365.IModule<T> {
    resources: IAccountOrderTemplatesTileResources;
    config: IAccountOrderTemplatesTileConfig;
}
