// @msdyn365-commerce/components/dist/lib/videoplayer/components/closed-captions/ttml-settings.js
import { extend } from '../utilities/utilities';
export const xmlNS = 'http://www.w3.org/XML/1998/namespace';
export class TtmlSettings {
    constructor(settingsOverrides) {
        this.ttmlNamespace = 'http://www.w3.org/ns/ttml';
        this.ttmlStyleNamespace = 'http://www.w3.org/ns/ttml#styling';
        this.ttmlParameterNamespace = 'http://www.w3.org/ns/ttml#parameter';
        this.ttmlMetaNamespace = 'http://www.w3.org/ns/ttml#metadata';
        this.idPrefix = '';
        this.mediaFrameRate = 30;
        this.mediaFrameRateMultiplier = 1;
        this.mediaSubFrameRate = 1;
        this.mediaTickRate = 1000;
        this.supportedTimeBase = 'media';
        this.rootContainerRegionDimensions = null;
        this.relatedMediaObjectRegion = null;
        this.cellResolution = { rows: 15, columns: 32 };
        this.defaultRegionStyle = {
            backgroundColor: 'transparent',
            color: '#E8E9EA',
            direction: 'ltr',
            display: 'auto',
            displayAlign: 'before',
            extent: 'auto',
            fontFamily: 'default',
            fontSize: '1c',
            fontStyle: 'normal',
            fontWeight: 'normal',
            lineHeight: 'normal',
            opacity: '1',
            origin: 'auto',
            overflow: 'hidden',
            padding: '0',
            showBackground: 'always',
            textAlign: 'start',
            textDecoration: 'none',
            textOutline: 'none',
            unicodeBidi: 'normal',
            visibility: 'visible',
            wrapOption: 'normal',
            writingMode: 'lrtb'
        };
        this.fontMap = {};
        this.fontMap.default = 'Lucida sans typewriter, Lucida console, Consolas';
        this.fontMap.monospaceSerif = 'Courier';
        this.fontMap.proportionalSerif = 'Times New Roman, Serif';
        this.fontMap.monospaceSansSerif = 'Lucida sans typewriter, Lucida console, Consolas';
        this.fontMap.proportionalSansSerif = 'Arial, Sans-serif';
        this.fontMap.casual = 'Verdana';
        this.fontMap.cursive = 'Zapf-Chancery, Segoe script, Cursive';
        this.fontMap.smallCaps = 'Arial, Helvetica';
        this.fontMap.monospace = 'Lucida sans typewriter, Lucida console, Consolas';
        this.fontMap.sansSerif = 'Arial, Sans-serif';
        this.fontMap.serif = 'Times New Roman, Serif';
        if (settingsOverrides) {
            extend(this, settingsOverrides);
        }
    }
}
