// @msdyn365-commerce-modules/checkout/src/modules/checkout-gift-card/tests/__mock__.ts
import { Cart, GiftCard } from '@msdyn365-commerce/retail-proxy/dist/Entities/CommerceTypes.g';
import { ICheckoutGiftCardResources } from '../checkout-gift-card.props.autogenerated';

export const mockResources: ICheckoutGiftCardResources = {
    emptyInputError: 'Please enterd a gift card number',
    invalidCodeError: 'Invalid gift card code entered',
    invalidCardInfoError: 'Invalid gift card information entered',
    invalidCardTypeError: 'The gift card type is not suported',
    noCardPinError: 'Gift card PIN is required',
    noCardExpError: 'Gift card expiration date is required',
    duplicatedCodeError: 'You already entered this gift card',
    noBalanceError: 'Gift card has no balance left',
    giftCardFormLabel: 'Gift card',
    giftCardNumberLabel: 'Gift card number',
    giftCardPinLabel: 'PIN',
    giftCardExpLabel: 'Expiration date',
    giftCardAlertLabel: 'You gift card may or may not include these fields',
    removeGiftCardButton: 'Remove',
    applyGiftCardButton: 'Apply',
    enteredGiftCardLabel: 'Gift card ending in',
    giftCardAvailableBalanceLabel: 'Available balance',
    giftCardPinPlaceholderText: 'PIN',
    giftCardExpPlaceholderText: 'MM/YYYY'
};

export const mockCart: Cart = {
    Id: '123',
    TotalAmount: 500
};

export const mockGiftCard1: GiftCard = {
    Id: 'abc',
    Balance: 100,
    BalanceCurrencyCode: 'USD',
    BalanceInCardCurrency: 100,
    CardCurrencyCode: 'USD'
};

export const mockGiftCard2: GiftCard = {
    Id: 'efg',
    Balance: 100,
    BalanceCurrencyCode: 'USD',
    BalanceInCardCurrency: 100,
    CardCurrencyCode: 'USD'
};
