// @msdyn365-commerce-modules/account-management/src/modules/account-loyalty/account-loyalty.props.autogenerated.ts
/**
 * Copyright (c) 2018 Microsoft Corporation
 * IAccountLoyalty contentModule Interface Properties
 * THIS FILE IS AUTO-GENERATED - MANUAL MODIFICATIONS WILL BE LOST
 */

import * as Msdyn365 from '@msdyn365-commerce/core';

export interface IAccountLoyaltyConfig extends Msdyn365.IModuleConfig {
    heading?: IHeadingData;
    className?: string;
}

export interface IAccountLoyaltyResources {
    loyaltyCardLabel: string;
    joinDateFormatted: string;
    totalAvailablePointsLabel: string;
    expiringPointsFormatLabel: string;
    availablePointsLabel: string;
    loyaltyActivityDateLabel: string;
    loyaltyActivityBalanceLabel: string;
    redeemedPointsLabel: string;
    earnedPointsLabel: string;
    pendingPointsLabel: string;
    adjustedPointsLabel: string;
    backToShoppingText: string;
}

export const enum HeadingTag {
    h1 = 'h1',
    h2 = 'h2',
    h3 = 'h3',
    h4 = 'h4',
    h5 = 'h5',
    h6 = 'h6'
}

export interface IHeadingData {
    text: string;
    tag?: HeadingTag;
}

export interface IAccountLoyaltyProps<T> extends Msdyn365.IModule<T> {
    resources: IAccountLoyaltyResources;
    config: IAccountLoyaltyConfig;
}
