// @msdyn365-commerce/components/dist/lib/videoplayer/components/utilities/utilities.js
export function clamp(value, min, max) {
    if (min <= max) {
        return Math.min(Math.max(value, min), max);
    }
    else {
        return 0;
    }
}
export function toUpperCaseFirstChar(value) {
    if (value.length > 1) {
        return value.charAt(0).toUpperCase() + value.substr(1);
    }
    if (value.length === 1) {
        return value.charAt(0).toUpperCase();
    }
    else {
        return '';
    }
}
export function htmlCollectionToArray(nodeList) {
    return Array.prototype.slice.call(nodeList);
}
export function getCSS(element, property) {
    let style = element.style[property];
    if (!style) {
        style = getComputedStyle(element)[property];
    }
    return style;
}
export function createThrottledEvent(listener, threshold) {
    let timer;
    let last = 0;
    return (event) => {
        const now = Date.now();
        if (timer) {
            clearTimeout(timer);
            timer = undefined;
        }
        if (!!last && (now < last + threshold)) {
            timer = setTimeout(() => {
                last = Date.now();
                listener(event);
            }, threshold - (now - last));
        }
        else {
            last = now;
            listener(event);
        }
    };
}
export function addThrottledEvent(target, name, listener, threshold = 150) {
    const throttledEventHandler = createThrottledEvent(listener, threshold);
    target.addEventListener(name, throttledEventHandler, false);
    return throttledEventHandler;
}
export function mapToCssModules(className = '', cssModule) {
    if (!cssModule) {
        return className;
    }
    return className
        .split(' ')
        .map((c) => cssModule[c] || c)
        .join(' ');
}
export function isRef(target) {
    if (!target) {
        return false;
    }
    return target.current !== undefined;
}
export function isElement(target) {
    if (!target) {
        return false;
    }
    return target.addEventListener !== undefined;
}
export const keyCodes = {
    esc: 27,
    space: 32,
    tab: 9,
    up: 38,
    down: 40,
    enter: 13
};
const warned = {};
export function warnOnce(message) {
    if (!warned[message]) {
        if (typeof console !== 'undefined') {
            throw new Error(message);
        }
        warned[message] = true;
    }
}
export const canUseDOM = !!(typeof window !== 'undefined' &&
    window.document &&
    window.document.createElement);
export function findDOMElements(target) {
    try {
        if (typeof target === 'string' && canUseDOM) {
            let selection = document.querySelectorAll(target);
            if (!selection.length) {
                selection = document.querySelectorAll(`#${target}`);
            }
            if (!selection.length) {
                throw new Error();
            }
            return selection;
        }
        return target;
    }
    catch {
        console.error(`The target '${target}' could not be identified in the dom, tip: check spelling`);
        return null;
    }
}
export function getTarget(target) {
    const selection = findDOMElements(target);
    if (selection && selection.length) {
        return selection[0];
    }
    return target;
}
export function getTargetElement(target) {
    const targetElement = getTarget(target);
    if (isRef(targetElement)) {
        return targetElement.current;
    }
    if (isElement(targetElement)) {
        return targetElement;
    }
    return null;
}
export const TransitionTimeouts = {
    Fade: 150,
    Collapse: 350,
    Modal: 300,
    Carousel: 600
};
export const focusableElements = [
    'a[href]',
    'area[href]',
    'input:not([disabled]):not([type=hidden])',
    'select:not([disabled])',
    'textarea:not([disabled])',
    'button:not([disabled])',
    'iframe',
    'object',
    'embed',
    '[tabindex]:not(.modal)',
    'audio[controls]',
    'video[controls]',
    '[contenteditable]:not([contenteditable="false"])',
];
export function extend(...parameters) {
    if (!parameters || !parameters.length) {
        return null;
    }
    const recursive = (typeof (parameters[0]) === 'boolean') && parameters[0];
    if (parameters.length < 2) {
        return recursive ? null : parameters[0];
    }
    if (recursive && (parameters.length < 3)) {
        return parameters[1];
    }
    const target = recursive ? parameters[1] : parameters[0];
    for (let paramIndex = recursive ? 2 : 1; paramIndex < parameters.length; paramIndex++) {
        if (parameters[paramIndex] !== undefined) {
            for (const key of Object.keys(parameters[paramIndex])) {
                if (parameters[paramIndex].hasOwnProperty(key)) {
                    const source = parameters[paramIndex][key];
                    if (recursive) {
                        let isSourceArray = Array.isArray
                            ? Array.isArray(source)
                            : ({}).toString.call(source) === '[object Array]';
                        let isTargetArray = !!target[key] && (Array.isArray
                            ? Array.isArray(target[key])
                            : ({}).toString.call(target[key]) === '[object Array]');
                        let isSourceObject = !isSourceArray && (typeof source === 'object');
                        let isTargetObject = !isTargetArray && !!target[key] && (typeof target[key] === 'object');
                        if (isSourceArray && isTargetArray) {
                            for (let arrayIndex = 0; arrayIndex < source.length; arrayIndex++) {
                                isSourceArray = Array.isArray
                                    ? Array.isArray(source[arrayIndex])
                                    : ({}).toString.call(source[arrayIndex]) === '[object Array]';
                                isTargetArray = !!target[key][arrayIndex] && (Array.isArray
                                    ? Array.isArray(target[key][arrayIndex])
                                    : ({}).toString.call(target[key][arrayIndex]) === '[object Array]');
                                isSourceObject = !isSourceArray && (typeof source[arrayIndex] === 'object');
                                isTargetObject = !isTargetArray && !!target[key][arrayIndex] && (typeof target[key][arrayIndex] === 'object');
                                if (isSourceArray) {
                                    target[key][arrayIndex] = extend(true, isTargetArray ? target[key][arrayIndex] : [], source[arrayIndex]);
                                }
                                else if (isSourceObject) {
                                    target[key][arrayIndex] = extend(true, isTargetObject ? target[key][arrayIndex] : {}, source[arrayIndex]);
                                }
                                else {
                                    target[key][arrayIndex] = source[arrayIndex];
                                }
                            }
                            continue;
                        }
                        else if (isSourceArray) {
                            target[key] = extend(true, [], source);
                            continue;
                        }
                        else if (isSourceObject) {
                            target[key] = extend(true, isTargetObject ? target[key] : {}, source);
                            continue;
                        }
                    }
                    target[key] = source;
                }
            }
        }
    }
    return target;
}
export function getDimensions(containerElement) {
    if (containerElement === null) {
        return null;
    }
    return {
        width: containerElement.clientWidth,
        height: containerElement.clientHeight
    };
}
