// @msdyn365-commerce-modules/search-result-container/dist/lib/modules/search-result-container/list-page-state.js
import { __decorate } from "tslib";
import { createObservableDataAction } from '@msdyn365-commerce/core';
import { observable } from 'mobx';
export class ListPageState {
    constructor() {
        this.activeFilters = null;
        this.activeProducts = [];
        this.sortingCritera = null;
        this.currentPageNumber = null;
        this.pageSize = 10;
        this.pageType = 'Unknown';
        this.featureProduct = null;
    }
}
__decorate([
    observable
], ListPageState.prototype, "activeFilters", void 0);
__decorate([
    observable
], ListPageState.prototype, "activeProducts", void 0);
__decorate([
    observable
], ListPageState.prototype, "sortingCritera", void 0);
__decorate([
    observable
], ListPageState.prototype, "currentPageNumber", void 0);
__decorate([
    observable
], ListPageState.prototype, "totalProductCount", void 0);
__decorate([
    observable
], ListPageState.prototype, "pageSize", void 0);
__decorate([
    observable
], ListPageState.prototype, "pageType", void 0);
__decorate([
    observable
], ListPageState.prototype, "featureProduct", void 0);
export class ListPageStateInput {
    constructor() {
        this.getCacheKey = () => `ListPageState`;
        this.getCacheObjectType = () => 'ListPageState';
        this.dataCacheType = () => 'request';
    }
}
export async function listPageStateAction(input, ctx) {
    return new ListPageState();
}
export const createListPageInput = (inputData) => {
    return new ListPageStateInput();
};
export default createObservableDataAction({
    action: listPageStateAction,
    input: createListPageInput
});
