// @msdyn365-commerce-modules/utilities/dist/lib/fade/Fade.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var classnames_1 = tslib_1.__importDefault(require("classnames"));
var React = tslib_1.__importStar(require("react"));
var Transition_1 = tslib_1.__importDefault(require("react-transition-group/Transition"));
var utilities_1 = require("../common/utilities");
var Fade = (function (_super) {
    tslib_1.__extends(Fade, _super);
    function Fade() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Fade.prototype.render = function () {
        var _a = this.props, tag = _a.tag, className = _a.className, cssModule = _a.cssModule, baseClass = _a.baseClass, baseClassActive = _a.baseClassActive, appear = _a.appear, enter = _a.enter, exit = _a.exit, fadeIn = _a.in, mountOnEnter = _a.mountOnEnter, unmountOnExit = _a.unmountOnExit, timeoutOrUndefined = _a.timeout, addEndListener = _a.addEndListener, onEnter = _a.onEnter, onEntering = _a.onEntering, onEntered = _a.onEntered, onExit = _a.onExit, onExiting = _a.onExiting, onExited = _a.onExited, children = _a.children, otherProps = tslib_1.__rest(_a, ["tag", "className", "cssModule", "baseClass", "baseClassActive", "appear", "enter", "exit", "in", "mountOnEnter", "unmountOnExit", "timeout", "addEndListener", "onEnter", "onEntering", "onEntered", "onExit", "onExiting", "onExited", "children"]);
        var timeout = timeoutOrUndefined !== undefined ? timeoutOrUndefined : utilities_1.TransitionTimeouts.Fade;
        var transitionProps = {
            appear: appear,
            enter: enter,
            exit: exit,
            in: fadeIn,
            mountOnEnter: mountOnEnter,
            unmountOnExit: unmountOnExit,
            timeout: timeout,
            addEndListener: addEndListener,
            onEnter: onEnter,
            onEntering: onEntering,
            onEntered: onEntered,
            onExit: onExit,
            onExiting: onExiting,
            onExited: onExited,
            children: children
        };
        var Tag = this.props.tag ? this.props.tag : 'div';
        var baseClassAlways = baseClass || 'fade';
        var baseClassActiveAlways = baseClassActive || 'show';
        return (React.createElement(Transition_1.default, tslib_1.__assign({}, transitionProps), function (status) {
            var isActive = status === 'entered';
            var classes = utilities_1.mapToCssModules(classnames_1.default([
                className,
                baseClassAlways,
                isActive && baseClassActiveAlways
            ]), cssModule);
            return (React.createElement(Tag, tslib_1.__assign({ className: classes }, otherProps), children));
        }));
    };
    Fade.defaultProps = {
        tag: 'div',
        baseClass: 'fade',
        baseClassActive: 'show',
        timeout: utilities_1.TransitionTimeouts.Fade,
        in: true,
        appear: true,
        enter: true,
        exit: true
    };
    return Fade;
}(React.Component));
exports.default = Fade;
