// @msdyn365-commerce-modules/account-management/src/modules/organization-users-tile/organization-users-tile.props.autogenerated.ts
/**
 * Copyright (c) 2018 Microsoft Corporation
 * IOrganizationUsersTile contentModule Interface Properties
 * THIS FILE IS AUTO-GENERATED - MANUAL MODIFICATIONS WILL BE LOST
 */

import * as Msdyn365 from '@msdyn365-commerce/core';

export interface IOrganizationUsersTileConfig extends Msdyn365.IModuleConfig {
    heading: IHeadingData;
    links: ILinksData[];
    className?: string;
}

export interface IOrganizationUsersTileResources {
    accountOrganizationUsersTileDescription: string;
}

export const enum HeadingTag {
    h1 = 'h1',
    h2 = 'h2',
    h3 = 'h3',
    h4 = 'h4',
    h5 = 'h5',
    h6 = 'h6'
}

export interface IHeadingData {
    text: string;
    tag?: HeadingTag;
}

export interface ILinksData {
    linkText?: string;
    linkUrl: Msdyn365.ILinkData;
    ariaLabel?: string;
    openInNewTab?: boolean;
}

export interface IOrganizationUsersTileProps<T> extends Msdyn365.IModule<T> {
    resources: IOrganizationUsersTileResources;
    config: IOrganizationUsersTileConfig;
}
