// @msdyn365-commerce/core-internal/dist/lib/utilities/links.js
/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */
import React from 'react';
import { EditableField, FieldType } from './editable-fields/editable-field';
/**
 * Component class for action links
 */
export class Links extends React.Component {
    constructor(props) {
        super(props);
        this.render = () => {
            if (this.props.links && this.props.links.length > 0) {
                return (React.createElement(React.Fragment, null, this.props.links.map((l, i) => (React.createElement(EditableField, { requestContext: this.props.editProps?.requestContext, type: FieldType.Link, onChange: this.props.editProps?.onTextChange
                        ? this.props.editProps.onTextChange(i)
                        : (event) => null, fieldProps: {
                        title: l.linkText,
                        text: l.linkText,
                        key: l.key || l.linkUrl,
                        openInNewTab: l.openInNewTab,
                        className: l.className || (i === 0 ? 'msc-cta__primary' : 'msc-cta__secondary'),
                        href: l.linkUrl,
                        role: l.role || 'button',
                        ariaLabel: l.ariaLabel,
                        onClick: l.onClick,
                        additionalProperties: l.additionalProperties,
                        linkTag: l.linkTag,
                        innerClassName: l.innerClassName
                    } })))));
            }
            return null;
        };
    }
}
