// @msdyn365-commerce/core-internal/dist/lib/utilities/link.js
/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */
import React from 'react';
import { EditableField, FieldType } from './editable-fields/editable-field';
/**
 * Component class for action links
 */
export class Link extends React.Component {
    constructor(props) {
        super(props);
        this.render = () => {
            if (this.props.link) {
                return (React.createElement(EditableField, { requestContext: this.props.editProps?.requestContext, type: FieldType.Link, onChange: this.props.editProps?.onTextChange ? this.props.editProps.onTextChange : (event) => null, fieldProps: {
                        title: this.props.link.linkText,
                        text: this.props.link.linkText,
                        key: this.props.link.key || this.props.link.linkUrl,
                        openInNewTab: this.props.link.openInNewTab,
                        className: this.props.link.className,
                        href: this.props.link.linkUrl,
                        role: this.props.link.role || 'button',
                        ariaLabel: this.props.link.ariaLabel,
                        onClick: this.props.link.onClick,
                        additionalProperties: this.props.link.additionalProperties,
                        linkTag: this.props.link.linkTag,
                        innerClassName: this.props.link.innerClassName
                    } }));
            }
            return null;
        };
    }
}
