// @msdyn365-commerce-modules/footer/dist/lib/modules/back-top-footer/back-top-footer.js
import { BackToTop, getTelemetryObject } from '@msdyn365-commerce-modules/utilities';
import * as React from 'react';
class BackTopFooter extends React.PureComponent {
    constructor(props) {
        super(props);
        this.editProps = {
            requestContext: this.props.context.request,
            onEdit: (event) => this.props.config.backtoTopText = event.target.value
        };
        this.telemetryContent = getTelemetryObject(this.props.context.request.telemetryPageName, this.props.friendlyName, this.props.telemetry);
    }
    render() {
        if (!this.props.config.destination) {
            this.props.telemetry.error('BackToTop.destination has not been provided, module wont render.');
            return null;
        }
        return (React.createElement(BackToTop, { href: this.props.config.destination, title: this.props.config.title, "aria-label": this.props.config.ariaLabel, className: 'ms-back-to-top', editProps: this.editProps, text: this.props.config.backtoTopText, telemetryContent: this.telemetryContent }));
    }
}
export default BackTopFooter;
