// @msdyn365-commerce-modules/product-specification/dist/lib/dataActions/get-product-availabilities-for-selected-variant.js
import { createObservableDataAction } from '@msdyn365-commerce/core';
import { getSelectedProductIdFromActionInput, getSelectedVariant, SelectedVariantInput } from '@msdyn365-commerce-modules/retail-actions';
import { getProductAvailabilitiesAsync } from '@msdyn365-commerce/retail-proxy/dist/DataActions/ProductsDataActions.g';
export class ProductAvailabilitiesForSelectedVariantInput {
    constructor(productId, channelId) {
        this.getCacheKey = () => `ProductAvailabilitiesForSelectedVariant`;
        this.getCacheObjectType = () => 'ProductAvailabilities';
        this.dataCacheType = () => 'none';
        this.productId = typeof productId === 'string' ? +productId : productId;
        this.channelId = channelId;
    }
}
export const createProductAvailabilitiesForSelectedVariantInput = (inputData) => {
    const productId = getSelectedProductIdFromActionInput(inputData);
    if (productId) {
        return new ProductAvailabilitiesForSelectedVariantInput(+productId, +inputData.requestContext.apiSettings.channelId);
    }
    else {
        throw new Error('Unable to create ProductAvailabilitiesForSelectedVariantInput, no productId found on module config or query');
    }
};
export async function getProductAvailabilitiesForSelectedVariantAction(input, ctx) {
    const selectedVariantInput = new SelectedVariantInput(input.productId, input.channelId);
    return getSelectedVariant(selectedVariantInput, ctx)
        .then(productResult => {
        if (productResult) {
            return getProductAvailabilitiesAsync({ callerContext: ctx, queryResultSettings: {} }, [productResult.RecordId], input.channelId).then(response => {
                if (response && response.length) {
                    return response[0];
                }
                return null;
            });
        }
        return null;
    })
        .catch((error) => {
        ctx.trace(error.message);
        throw new Error('[getPriceForSelectedVariantAction]Error executing action');
    });
}
export default createObservableDataAction({
    id: '@msdyn365-commerce-modules/product-info/get-product-availabilities-for-selected-variant',
    action: getProductAvailabilitiesForSelectedVariantAction,
    input: createProductAvailabilitiesForSelectedVariantInput
});
