// @msdyn365-commerce-modules/ratings-reviews/dist/lib/actions/submit-helpfulness-vote.js
import { createObservableDataAction, sendRequest } from '@msdyn365-commerce/core';
async function submitHelpfulnessVote(input, ctx) {
    if (!inputIsValid(input)) {
        throw new Error(`[SubmitHelpfulnessVote] Invalid input. ReviewId=${input.reviewId},tenantId=${input.tenantId}, locale= ${input.locale},`);
    }
    const requestUrl = `${input.serviceEndpoint}/v2.0/votes/review/${input.reviewId}` +
        `?tenantId=${input.tenantId}` +
        `&locale=${input.locale}`;
    const requestBody = {
        isHelpful: !!input.isHelpful
    };
    const requestHeader = {
        headers: {
            Authorization: `id_token ${input.authToken}`
        }
    };
    return sendRequest(requestUrl, 'post', requestBody, requestHeader)
        .then(response => {
        return response.status === 200 || response.status === 201;
    })
        .catch((error) => {
        ctx.trace(`[SubmitHelpfulnessVote] Failed to vote on a review`);
        ctx.telemetry.exception(error);
        ctx.telemetry.debug(`[SubmitHelpfulnessVote] Failed to vote on a review`);
        return false;
    });
}
function inputIsValid(input) {
    return (input &&
        input.reviewId.length > 0 &&
        input.tenantId.length > 0 &&
        input.locale.length > 0 &&
        input.authToken.length > 0);
}
export default createObservableDataAction({
    id: '@msdyn365-commerce-modules/ratings-reviews/submit-helpfulness-vote',
    action: submitHelpfulnessVote
});
