// @msdyn365-commerce-modules/cart/dist/lib/modules/cart/__mocks__/mock-resources.js
export const mockResources = {
    pickUpAtStoreWithLocationText: 'Pick up at a store',
    bopisSelectAStore: 'Select A Store',
    orLabel: 'OR',
    backToShopping: 'backToShopping',
    outOfStockErrorMessageHeader: 'There are some issues.',
    outOfStockErrorMessage: 'This product is no longer available. Please remove it from the bag.',
    cartTitle: 'Cart',
    emptyCartText: 'Your cart is empty.',
    failToGetCartText: 'Unable to get cart!',
    removeCartButtonText: 'Remove',
    cartRemoveButtonAriaLabel: 'Remove {productName}',
    pickUpAtStoreText: 'Pick up at a store ',
    waitingClass: 'waiting waiting-circular waiting-lg',
    decrementButtonAriaLabel: 'Press to decrease the quantity by 1',
    incrementButtonAriaLabel: 'Press to increment quantity by 1',
    inputQuantityAriaLabel: 'quantity input box aria label',
    discountStringText: 'Discount ',
    saveForLaterText: 'Move to Wishlist',
    nameOfWishlist: 'My Wishlist',
    productDimensionTypeColor: 'Color',
    productDimensionTypeSize: 'Size',
    productDimensionTypeStyle: 'Style',
    productDimensionTypeConfiguration: 'Configuration',
    shipInsteadDisplayText: 'Ship this item',
    changeStoreDisplayText: 'Change store',
    quantityDisplayText: 'Quantity',
    shipToAddressDisplayText: 'Ship to an address',
    pickItUpDisplayText: 'Pick this up',
    totalSavingsText: 'Savings',
    errorMessageOutOfRangeOneLeft: 'There is only one product left. Please choose quantity within the available range',
    errorMessageOutOfRangeFormat: 'There are only {numLeft} products left. Please choose quantity within the available range',
    originalPriceText: 'Original price',
    currentPriceText: 'Current price',
    cartLoadingErrorMessage: 'Sorry we are unable to load your cart at this moment, please try again later.',
    orderSummaryTitle: 'Order summary',
    itemsWithCountLabel: 'Subtotal ',
    estimatedShippingLabel: 'Estimated Shipping',
    priceFree: 'Free',
    estimatedTaxLabel: 'Estimated Tax',
    totalAmountLabel: 'Amount due ',
    totalSavings: 'Total savings ',
    conditionsText: '*Before applicable taxes',
    yetToBeCalculatedText: 'To be calculated',
    orderTotal: 'Order total',
    notApplicationForCartOrderSummary: 'Not applicable',
    totalDiscountsLabel: 'Total discounts',
    guestCheckoutButtonTitle: 'Guest Checkout',
    checkoutButtonTitle: 'Checkout',
    backToShoppingButtonTitle: 'Back to shopping',
    addToWishlistButtonText: 'Move to wishlist',
    removeFromWishlistButtonText: 'Remove from wishlist',
    defaultWishlistName: 'My Wishlist',
    promoCodeHeadingText: 'Promo code',
    promoPlaceholderText: 'Enter a promo code',
    promoCodeApplyButtonText: 'Apply',
    removePromoAriaLabelFormat: '{0} {1}',
    removePromoText: 'Remove',
    appliedPromoCodeHeadingText: 'Applied promo code',
    invalidPromoCodeErrorText: 'Invalid promo code',
    failedToAddPromoCodeErrorText: 'Failed to add promo code',
    failedToRemovePromoCodeErrorText: 'Failed to remove promo code',
    duplicatePromotionErrorText: 'Promo code already applied',
    otherCharges: 'Other Charges',
    createOrderTemplateFromCartButton: 'Save all items to order template',
    shippingCharges: 'Shipping Charges',
    addToOrderTemplateButtonText: 'Add to order template',
    createOrderTemplateHeader: 'createOrderTemplateHeader',
    orderTemplateTitle: 'orderTemplateTitle',
    orderTemplateNameAriaLabel: 'orderTemplateNameAriaLabel',
    createOrderTemplateDescription: 'createOrderTemplateDescription',
    defaultOrderTemplateName: 'defaultOrderTemplateName',
    createOrderTemplateButtonText: 'createOrderTemplateButtonText',
    createNewOrderTemplateButtonText: 'createNewOrderTemplateButtonText',
    cancelNewOrderTemplateCreationButtonText: 'cancelNewOrderTemplateCreationButtonText',
    addToOrderTemplateHeader: 'addToOrderTemplateHeader',
    noOrderTemplatesMessage: 'noOrderTemplatesMessage',
    noOrderTemplatesDescription: 'noOrderTemplatesDescription',
    createAnOrderTemplateButtonText: 'createAnOrderTemplateButtonText',
    cancelOrderTemplateCreationButtonText: 'cancelOrderTemplateCreationButtonText',
    itemAddedToOrderTemplateHeaderItemOneText: 'itemAddedToOrderTemplateHeaderItemOneText',
    itemAddedToOrderTemplateHeaderItemFormatText: 'itemAddedToOrderTemplateHeaderItemFormatText',
    itemAddedToOrderTemplateHeaderMessageText: 'itemAddedToOrderTemplateHeaderMessageText',
    linesAddedToOrderTemplateHeaderItemsOneText: 'itemAddedToOrderTemplateHeaderItemOneText',
    linesAddedToOrderTemplateHeaderItemsFormatText: 'itemAddedToOrderTemplateHeaderItemFormatText',
    linesAddedToOrderTemplateHeaderLinesOneText: 'itemAddedToOrderTemplateHeaderItemOneText',
    linesAddedToOrderTemplateHeaderLinesFormatText: 'itemAddedToOrderTemplateHeaderItemFormatText',
    linesAddedToOrderTemplateHeaderMessageText: 'itemAddedToOrderTemplateHeaderMessageText',
    viewOrderTemplateButtonText: 'viewOrderTemplateButtonText',
    continueShoppingButtonText: 'continueShoppingButtonText',
    addItemToOrderTemplateError: 'Unable to add item to order template',
    freePriceText: 'freePriceText',
    selectTemplatesText: 'Select a template or create a new one. You have {0} template(s).',
    addToTemplateButtonText: 'Add to template',
    lineItemsText: '{0} line item(s)',
    addToOrderTemplateMessage: 'Added to your order template',
    duplicatedProductsHeader: 'duplicatedProductsHeader',
    duplicatedProductsDescription: 'duplicatedProductsDescription',
    updateQuantityButtonText: 'updateQuantityButtonText',
    cancelDuplicateItemsButtonText: 'cancelDuplicateItemsButtonText'
};
