// @msdyn365-commerce-modules/video-player/src/modules/video-player/video-player.props.autogenerated.ts
/**
 * Copyright (c) 2018 Microsoft Corporation
 * IVideoPlayer contentModule Interface Properties
 * THIS FILE IS AUTO-GENERATED - MANUAL MODIFICATIONS WILL BE LOST
 */

import * as Msdyn365 from '@msdyn365-commerce/core';

export interface IVideoPlayerConfig extends Msdyn365.IModuleConfig {
    video: Msdyn365.IVideoData;
    autoplay?: boolean;
    mute?: boolean;
    loop?: boolean;
    autoCaptions?: string;
    playPauseTrigger?: boolean;
    controls?: boolean;
    playFullScreen?: boolean;
    hidePosterFrame?: boolean;
    maskLevel?: number;
    thumbnailFullWidth?: boolean;
    lazyLoad?: boolean;
}

export interface IVideoPlayerResources {
    playLabel: string;
    pauseLabel: string;
    pausedLabel: string;
    playingLabel: string;
    unMuteLabel: string;
    muteLabel: string;
    fullScreenLabel: string;
    exitFullScreenLabel: string;
    seekBarLabel: string;
    videoTimeDurationLabel: string;
    closedCaptionLabel: string;
    optionButtonLabel: string;
}

export interface IVideoPlayerProps<T> extends Msdyn365.IModule<T> {
    resources: IVideoPlayerResources;
    config: IVideoPlayerConfig;
}
