// @msdyn365-commerce/build-scripts-internal/dist/lib/module-registration/module-dapi-helper.js
"use strict";
/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var utilities_internal_1 = require("@msdyn365-commerce/utilities-internal");
var path_1 = tslib_1.__importDefault(require("path"));
var themes_dapi_helper_1 = require("./themes-dapi-helper");
var models_1 = require("./utils/models");
var modulesDAPIHelper = null;
exports.getModulesDAPIHelperSingleton = function () {
    if (!modulesDAPIHelper) {
        modulesDAPIHelper = new ModulesDAPIHelper();
    }
    return modulesDAPIHelper;
};
/**
 * Modules DAPI Helper class
 */
var ModulesDAPIHelper = /** @class */ (function () {
    function ModulesDAPIHelper() {
        var _this = this;
        this._registeredModules = {};
        this._dapiModulesList = {};
        this._themesDapiList = [];
        this.prepareModuleDefinitionsForDAPI = function (modules) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this._addModules(modules);
                        return [4 /*yield*/, Promise.all((Object.keys(modules) || []).map(function (moduleType) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var moduleRegObj, definition, dataJson;
                                return tslib_1.__generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0:
                                            moduleRegObj = modules[moduleType];
                                            return [4 /*yield*/, this._getModuleDefinition(moduleRegObj)];
                                        case 1:
                                            definition = _a.sent();
                                            return [4 /*yield*/, this._getDataJsonAndParse(moduleRegObj)];
                                        case 2:
                                            dataJson = _a.sent();
                                            !utilities_internal_1.isEmptyOrNullObject(dataJson) && (definition["data"] = dataJson);
                                            this._dapiModulesList[moduleRegObj.name] = definition;
                                            return [2 /*return*/];
                                    }
                                });
                            }); }))];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); };
        this.generateThemesDAPI = function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var _a, themeSettings;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this;
                        return [4 /*yield*/, themes_dapi_helper_1.getThemesDAPIList(this._registeredModules, this._dapiModulesList)];
                    case 1:
                        _a._themesDapiList = _b.sent();
                        return [4 /*yield*/, utilities_internal_1.safeWriteJson(models_1.keystonePaths.KEYSTONE_APP_THEMES_DAPI_FILEPATH, this._themesDapiList)];
                    case 2:
                        _b.sent();
                        return [4 /*yield*/, themes_dapi_helper_1.getThemeSettings(this._registeredModules)];
                    case 3:
                        themeSettings = _b.sent();
                        return [4 /*yield*/, utilities_internal_1.safeWriteJson(models_1.keystonePaths.KEYSTONE_APP_SETTINGS_DAPI_FILEPATH, themeSettings)];
                    case 4:
                        _b.sent();
                        return [2 /*return*/];
                }
            });
        }); };
        this.generateModuleDefinitionDAPI = function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, utilities_internal_1.safeWriteJson(models_1.keystonePaths.KEYSTONE_APP_DAPI_FILEPATH, this._dapiModulesList)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); };
        this.getModuleRegistrations = function () {
            var modules = [];
            Object.keys(_this._registeredModules || {}).forEach(function (key) {
                modules.push(_this._registeredModules[key]);
            });
            return modules;
        };
        this.getThemes = function () {
            return _this._themesDapiList;
        };
        this._getModuleDefinition = function (moduleBinder) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var definitionFile;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, utilities_internal_1.safeReadJson(moduleBinder.definitionPath, {
                            error: function (input) {
                                console.log("Couldn't read definition file [" + moduleBinder.definitionPath + "]", { exception: input });
                            }
                        })];
                    case 1:
                        definitionFile = _a.sent();
                        if (!definitionFile || utilities_internal_1.isEmptyOrNullObject(definitionFile)) {
                            return [2 /*return*/, {
                                    error: "problem found with definition file [" + moduleBinder.definitionPath + "] for module [" + moduleBinder.name + "]"
                                }];
                        }
                        return [2 /*return*/, this._parseModuleDefinition(definitionFile, moduleBinder.definitionPath)];
                }
            });
        }); };
        this._getDataJsonPath = function (definitionPath, isNodeModule) {
            var filePath = (isNodeModule ? definitionPath : definitionPath.replace('src/modules', 'build/definitions')).replace('.definition.json', '.data.json');
            return path_1.default.resolve(filePath);
        };
        this._parseModuleDefinition = function (defN, definitionPath) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var definitionWithResolvedRef;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, utilities_internal_1.resolveRef(defN, definitionPath)];
                    case 1:
                        definitionWithResolvedRef = _a.sent();
                        definitionWithResolvedRef["module"] && delete definitionWithResolvedRef["module"];
                        this._renameKeyInObject(definitionWithResolvedRef, '$type', 'type', 0);
                        return [2 /*return*/, definitionWithResolvedRef];
                }
            });
        }); };
        this._getDataJsonAndParse = function (moduleBinder) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var dataJsonPathFromDefinitionPath, dataJsonFileExists, dataJsonFile, dataJson;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        dataJsonPathFromDefinitionPath = this._getDataJsonPath(moduleBinder.definitionPath, moduleBinder.isNodeModule);
                        return [4 /*yield*/, utilities_internal_1.safeFileExists(dataJsonPathFromDefinitionPath)];
                    case 1:
                        dataJsonFileExists = _a.sent();
                        if (!(!dataJsonFileExists && moduleBinder.parentDefinitionPath)) return [3 /*break*/, 3];
                        dataJsonPathFromDefinitionPath = this._getDataJsonPath(moduleBinder.parentDefinitionPath, moduleBinder.isNodeModule);
                        return [4 /*yield*/, utilities_internal_1.safeFileExists(dataJsonPathFromDefinitionPath)];
                    case 2:
                        dataJsonFileExists = _a.sent();
                        _a.label = 3;
                    case 3:
                        if (!dataJsonFileExists) {
                            return [2 /*return*/, {}];
                        }
                        return [4 /*yield*/, utilities_internal_1.safeReadJson(dataJsonPathFromDefinitionPath, {
                                error: function (input) {
                                    console.log("Couldn't read data json file [" + dataJsonPathFromDefinitionPath + "]", { exception: input });
                                }
                            })];
                    case 4:
                        dataJsonFile = _a.sent();
                        dataJson = dataJsonFile && dataJsonFile["properties"];
                        if (!dataJson) {
                            return [2 /*return*/, {}];
                        }
                        Object.keys(dataJson).forEach(function (eachDataTypePropertyKey) {
                            var eachDataTypeProperty = dataJson[eachDataTypePropertyKey];
                            if (eachDataTypeProperty.filterFromDAPI) {
                                delete dataJson[eachDataTypePropertyKey];
                            }
                        });
                        this._renameKeyInObject(dataJson, '$ref', 'type', 3);
                        return [2 /*return*/, dataJson];
                }
            });
        }); };
        // tslint:disable-next-line:no-any
        this._renameKeyInObject = function (value, oldkey, newkey, depth) {
            if (utilities_internal_1.isObject(value) && utilities_internal_1.isEmptyOrNullObject(value)) {
                return value;
            }
            if (oldkey !== newkey && value.hasOwnProperty(oldkey)) {
                Object.defineProperty(value, newkey, Object.getOwnPropertyDescriptor(value, oldkey) || '');
                delete value[oldkey];
            }
            // if depth === undefined, then rename till all depth
            if (depth !== undefined && depth === 0) {
                return;
            }
            Object.keys(value).forEach(function (key) {
                var childObject = value[key];
                if (utilities_internal_1.isObject(childObject)) {
                    _this._renameKeyInObject(childObject, oldkey, newkey, depth && depth - 1);
                }
            });
        };
    }
    /**
     * Extracts the name of the module from the path to its definition file
     * @param definitionPath path to the definition file
     */
    ModulesDAPIHelper.prototype.getModuleName = function (definitionPath) {
        if (!definitionPath || definitionPath === '') {
            return '';
        }
        return path_1.default.basename(definitionPath, ".definition.json" /* DEFINITION_FILE */);
    };
    /**
     * Returns the binder object for the given module
     * @param typeName name of the module
     */
    ModulesDAPIHelper.prototype.getModuleBinder = function (typeName) {
        return this._registeredModules[typeName];
    };
    /**
     * Returns name of the parent module for given module name
     * @param moduleName name of the module
     */
    ModulesDAPIHelper.prototype.getParentModuleName = function (moduleName) {
        var moduleBinder = this.getModuleBinder(moduleName);
        if (!moduleBinder) {
            return '';
        }
        return this.getModuleName(moduleBinder.parentDefinitionPath);
    };
    /**
     * Gets module definition for given module name
     * @param moduleRegistration object
     */
    ModulesDAPIHelper.prototype.getModuleDefinition = function (moduleName) {
        return this._dapiModulesList[moduleName] || {};
    };
    ModulesDAPIHelper.prototype._addModules = function (registrationEntries) {
        var _this = this;
        this._dapiModulesList = this._dapiModulesList || {};
        Object.keys(registrationEntries || {}).forEach(function (entry) {
            _this._registeredModules[entry] = registrationEntries[entry];
        });
    };
    return ModulesDAPIHelper;
}());
