// @msdyn365-commerce/build-scripts-internal/dist/lib/module-registration/resources-dapi-helper.js
"use strict";
/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var utilities_internal_1 = require("@msdyn365-commerce/utilities-internal");
var lodash_1 = require("lodash");
var path_1 = tslib_1.__importDefault(require("path"));
var module_dapi_helper_1 = require("./module-dapi-helper");
var models_1 = require("./utils/models");
var RESOURCE_DETAILS_REGEX = /\/node_modules\/(?<namespace>.*)\/(?<packageName>.*)\/dist\/lib\/.*/;
/**
 * Resource manager implementation
 */
var ResourceManagerDAPIHelper = /** @class */ (function () {
    function ResourceManagerDAPIHelper() {
        this.authoringResourcesMap = {};
        this.moduleResourcesMap = {};
        this.moduleResourceKeys = {};
        this.moduleNamespaceMap = {};
    }
    /**
     * @function        {init}          - Method to initialize and load all the resource strings
     */
    ResourceManagerDAPIHelper.prototype.generateResourcesDAPI = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var localAuthoringResourceFiles, localModulesResourceFiles;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this._readAuthoringResourceFiles()];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, utilities_internal_1.safeGetAllFilesPath(models_1.localAuthoringResourcesPathPattern)];
                    case 2:
                        localAuthoringResourceFiles = _a.sent();
                        if (!(localAuthoringResourceFiles.length > 0)) return [3 /*break*/, 4];
                        return [4 /*yield*/, Promise.all(localAuthoringResourceFiles.map(function (resourceFile) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var locale, content;
                                var _this = this;
                                return tslib_1.__generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0:
                                            locale = path_1.default.basename(resourceFile, '.json').toLocaleLowerCase();
                                            return [4 /*yield*/, utilities_internal_1.safeReadJson(resourceFile)];
                                        case 1:
                                            content = _a.sent();
                                            if (!content || !content.modules) {
                                                return [2 /*return*/];
                                            }
                                            this.authoringResourcesMap[locale] = this.authoringResourcesMap[locale] || {
                                                settings: {},
                                                modules: {},
                                                themes: {}
                                            };
                                            content.modules &&
                                                Object.keys(content.modules).forEach(function (moduleName) {
                                                    _this.authoringResourcesMap[locale].modules[moduleName] = _this._getMergedAuthoringResources(_this.authoringResourcesMap[locale].modules[moduleName], content.modules[moduleName]);
                                                    if (content.settings) {
                                                        _this.authoringResourcesMap[locale].settings = content.settings;
                                                    }
                                                    _this.authoringResourcesMap[locale].themes = {};
                                                });
                                            if (content.themes) {
                                                Object.keys(content.themes).forEach(function (themeName) {
                                                    _this.authoringResourcesMap[locale].themes[themeName] = lodash_1.merge(_this.authoringResourcesMap[locale].themes[themeName], content.themes[themeName]);
                                                });
                                            }
                                            return [2 /*return*/];
                                    }
                                });
                            }); }))];
                    case 3:
                        _a.sent();
                        _a.label = 4;
                    case 4: return [4 /*yield*/, this._readModuleResourceFiles()];
                    case 5:
                        _a.sent();
                        return [4 /*yield*/, utilities_internal_1.safeGetAllFilesPath(models_1.localModulesResourcesPathPattern)];
                    case 6:
                        localModulesResourceFiles = _a.sent();
                        if (!(localModulesResourceFiles.length > 0)) return [3 /*break*/, 8];
                        return [4 /*yield*/, Promise.all(localModulesResourceFiles.map(function (resourceFile) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var locale, content, resourceKeys;
                                var _this = this;
                                return tslib_1.__generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0:
                                            locale = path_1.default.basename(resourceFile, '.json').toLocaleLowerCase();
                                            return [4 /*yield*/, utilities_internal_1.safeReadJson(resourceFile)];
                                        case 1:
                                            content = _a.sent();
                                            if (!content) {
                                                return [2 /*return*/];
                                            }
                                            this.moduleResourcesMap[locale] = this.moduleResourcesMap[locale] || {};
                                            resourceKeys = Object.keys(content);
                                            resourceKeys.forEach(function (resourceKey) {
                                                if (content[resourceKey].value) {
                                                    _this.moduleResourcesMap[locale][resourceKey] = content[resourceKey].value;
                                                }
                                            });
                                            return [2 /*return*/];
                                    }
                                });
                            }); }))];
                    case 7:
                        _a.sent();
                        _a.label = 8;
                    case 8: return [4 /*yield*/, this._getAllModuleDefinitions()];
                    case 9:
                        _a.sent();
                        return [2 /*return*/, utilities_internal_1.safeWriteJson(models_1.keystonePaths.KEYSTONE_APP_RESOURCES_DAPI_FILEPATH, {
                                authoringResourcesMap: this.authoringResourcesMap,
                                moduleResourcesMap: this.moduleResourcesMap,
                                moduleResourceKeys: this.moduleResourceKeys,
                                moduleNamespaceMap: this.moduleNamespaceMap
                            })];
                }
            });
        });
    };
    ResourceManagerDAPIHelper.prototype._readAuthoringResourceFiles = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var installedAuthoringResourceFiles, _a, _b;
            var _this = this;
            return tslib_1.__generator(this, function (_c) {
                switch (_c.label) {
                    case 0: return [4 /*yield*/, utilities_internal_1.safeGetAllFilesPath(models_1.installedAuthoringResourcesPathPattern)];
                    case 1:
                        installedAuthoringResourceFiles = _c.sent();
                        _b = (_a = installedAuthoringResourceFiles).concat;
                        return [4 /*yield*/, utilities_internal_1.safeGetAllFilesPath(models_1.hoistedAuthoringResourcesPathPattern)];
                    case 2:
                        installedAuthoringResourceFiles = _b.apply(_a, [_c.sent()]);
                        return [4 /*yield*/, Promise.all(installedAuthoringResourceFiles.map(function (resourceFile) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var locale, content;
                                return tslib_1.__generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0:
                                            locale = path_1.default.basename(resourceFile, '.json').toLocaleLowerCase();
                                            return [4 /*yield*/, utilities_internal_1.safeReadJson(resourceFile)];
                                        case 1:
                                            content = _a.sent();
                                            if (content) {
                                                this.authoringResourcesMap[locale] = content;
                                            }
                                            return [2 /*return*/];
                                    }
                                });
                            }); }))];
                    case 3:
                        _c.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    ResourceManagerDAPIHelper.prototype._readModuleResourceFiles = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var installedModulesResourceFiles, _a, _b;
            var _this = this;
            return tslib_1.__generator(this, function (_c) {
                switch (_c.label) {
                    case 0: return [4 /*yield*/, utilities_internal_1.safeGetAllFilesPath(models_1.installedModulesResourcesPathPattern)];
                    case 1:
                        installedModulesResourceFiles = _c.sent();
                        _b = (_a = installedModulesResourceFiles).concat;
                        return [4 /*yield*/, utilities_internal_1.safeGetAllFilesPath(models_1.hoistedModulesResourcesPathPattern)];
                    case 2:
                        installedModulesResourceFiles = _b.apply(_a, [_c.sent()]);
                        return [4 /*yield*/, Promise.all(installedModulesResourceFiles.map(function (resourceFile) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var matchResult, moduleNamespace, locale, content;
                                var _this = this;
                                return tslib_1.__generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0:
                                            matchResult = resourceFile.match(RESOURCE_DETAILS_REGEX);
                                            moduleNamespace = '__installed__';
                                            if (matchResult !== null && matchResult.groups && matchResult.groups.namespace) {
                                                moduleNamespace = matchResult.groups.namespace;
                                            }
                                            locale = path_1.default.basename(resourceFile, '.json').toLocaleLowerCase();
                                            return [4 /*yield*/, utilities_internal_1.safeReadJson(resourceFile)];
                                        case 1:
                                            content = _a.sent();
                                            if (!content) {
                                                return [2 /*return*/];
                                            }
                                            this.moduleResourcesMap[locale] = this.moduleResourcesMap[locale] || {};
                                            Object.keys(content).forEach(function (resourceKey) {
                                                if (content[resourceKey].value) {
                                                    _this.moduleResourcesMap[locale][moduleNamespace + "." + resourceKey] = content[resourceKey].value;
                                                }
                                            });
                                            return [2 /*return*/];
                                    }
                                });
                            }); }))];
                    case 3:
                        _c.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    ResourceManagerDAPIHelper.prototype._getAllModuleDefinitions = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var helper, modules, themeModules, defExtResources;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                helper = module_dapi_helper_1.getModulesDAPIHelperSingleton();
                modules = helper.getModuleRegistrations();
                themeModules = helper.getThemes();
                defExtResources = {};
                // Get resources object from definition extensions of theme modules
                if (themeModules && Array.isArray(themeModules)) {
                    (themeModules || []).map(function (themeModule) {
                        Object.keys(themeModule.definitionExtensions || []).map(function (moduleName) {
                            if (defExtResources[moduleName]) {
                                defExtResources[moduleName] = tslib_1.__assign(tslib_1.__assign({}, defExtResources[moduleName]), (themeModule.definitionExtensions[moduleName].resources || {}));
                            }
                            else {
                                defExtResources[moduleName] = themeModule.definitionExtensions[moduleName].resources || {};
                            }
                        });
                    });
                }
                return [2 /*return*/, Promise.all(modules.map(function (moduleBinder) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                        var definition, moduleResources, resourcesObject;
                        return tslib_1.__generator(this, function (_a) {
                            definition = module_dapi_helper_1.getModulesDAPIHelperSingleton().getModuleDefinition(moduleBinder.name);
                            this.moduleNamespaceMap[moduleBinder.name] = moduleBinder.moduleNamespace;
                            moduleResources = definition.resources || {};
                            // Merge definition extension resources
                            if (defExtResources[moduleBinder.name]) {
                                // @ts-ignore
                                moduleResources = tslib_1.__assign(tslib_1.__assign({}, moduleResources), defExtResources[moduleBinder.name]);
                            }
                            resourcesObject = {};
                            Object.keys(moduleResources).map(function (key) {
                                // @ts-ignore
                                resourcesObject[key] = moduleResources[key].value;
                            });
                            this.moduleResourceKeys[definition.name] = resourcesObject;
                            return [2 /*return*/, definition];
                        });
                    }); }))];
            });
        });
    };
    ResourceManagerDAPIHelper.prototype._getMergedAuthoringResources = function (moduleAuthoringResources, overrideModuleAuthoringResources) {
        moduleAuthoringResources = moduleAuthoringResources || {};
        overrideModuleAuthoringResources = overrideModuleAuthoringResources || {};
        var mergedModuleAuthoringResources = tslib_1.__assign(tslib_1.__assign({}, moduleAuthoringResources), overrideModuleAuthoringResources);
        // Do explicit merge of each properties as '...' does not do recursive merge
        var moduleAuthoringResourceProperties = Object.keys(moduleAuthoringResources);
        if (moduleAuthoringResourceProperties) {
            moduleAuthoringResourceProperties.forEach(function (property) {
                // @ts-ignore
                if (overrideModuleAuthoringResources[property]) {
                    // @ts-ignore
                    mergedModuleAuthoringResources[property] = tslib_1.__assign(tslib_1.__assign({}, moduleAuthoringResources[property]), (overrideModuleAuthoringResources[property] || {}));
                }
            });
        }
        return mergedModuleAuthoringResources;
    };
    return ResourceManagerDAPIHelper;
}());
exports.ResourceManagerDAPIHelper = ResourceManagerDAPIHelper;
