// @msdyn365-commerce-modules/order-management/dist/lib/modules/order-confirmation/order-confirmation.view.js
import { Module, Node } from '@msdyn365-commerce-modules/utilities';
import * as React from 'react';
export const OrderConfirmationOrderInfomation = ({ orderInformationProps, receiptEmail, createdDate, channelReferenceId }) => (React.createElement(Node, Object.assign({}, orderInformationProps),
    createdDate,
    channelReferenceId,
    receiptEmail));
export const OrderConfirmationSalesLine = ({ salesLineProps, salesLine, buyAgainButton }) => (React.createElement(Node, Object.assign({}, salesLineProps), salesLine));
export const OrderConfirmationGroupDelivery = ({ deliveryProps, heading, count, processing, address, trackingInfo }) => (React.createElement(Node, Object.assign({}, deliveryProps),
    heading,
    address));
export const OrderConfirmationGroup = ({ groupProps, delivery, salesLinesProps, salesLines }) => (React.createElement(Node, Object.assign({}, groupProps),
    delivery && React.createElement(OrderConfirmationGroupDelivery, Object.assign({}, delivery)),
    delivery && delivery.shippingItemsToYou,
    salesLines && (React.createElement(Node, Object.assign({}, salesLinesProps), salesLines.map(salesLine => (React.createElement(OrderConfirmationSalesLine, Object.assign({ key: salesLine.data.salesLine.LineId }, salesLine))))))));
export const OrderConfirmationGroups = ({ groupsProps, groups }) => (React.createElement(Node, Object.assign({}, groupsProps), groups.map((group, index) => (React.createElement(OrderConfirmationGroup, Object.assign({ key: index }, group))))));
export const OrderConfirmationOrderSummary = ({ orderSummaryProps, heading, subtotal, shipping, tax, totalAmount, earnedPoints }) => (React.createElement(Node, Object.assign({}, orderSummaryProps),
    heading,
    subtotal,
    shipping,
    tax,
    totalAmount,
    earnedPoints));
export const OrderConfirmationPayment = ({ paymentMethodsProps, title, methods }) => (React.createElement(Node, Object.assign({}, paymentMethodsProps),
    title,
    methods));
export const OrderConfirmationHelp = ({ helpProps, needHelpLabel, helpLineNumberLabel, contactNumber }) => (React.createElement(Node, Object.assign({}, helpProps),
    needHelpLabel,
    helpLineNumberLabel,
    contactNumber));
const OrderConfirmationView = ({ moduleProps, heading, backToShoppingLink, alert, loading, orderInfomation, orderSummary, payment, help, groups }) => {
    return (React.createElement(Module, Object.assign({}, moduleProps),
        heading,
        alert,
        loading,
        orderInfomation && React.createElement(OrderConfirmationOrderInfomation, Object.assign({}, orderInfomation)),
        backToShoppingLink,
        groups && React.createElement(OrderConfirmationGroups, Object.assign({}, groups)),
        payment && React.createElement(OrderConfirmationPayment, Object.assign({}, payment)),
        orderSummary && React.createElement(OrderConfirmationOrderSummary, Object.assign({}, orderSummary)),
        help && React.createElement(OrderConfirmationHelp, Object.assign({}, help))));
};
export default OrderConfirmationView;
