// @msdyn365-commerce-modules/page-summary/dist/lib/modules/default-page-summary/default-page-summary.action.js
/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */
import { buildCacheKeyWithUrlTokens } from '@msdyn365-commerce-modules/retail-actions';
import { createObservableDataAction } from '@msdyn365-commerce/core';
export class PageSummaryInput {
    constructor(config, requestContext) {
        this.getCacheObjectType = () => 'PageSummary';
        this.getCacheKey = () => `${buildCacheKeyWithUrlTokens('PageSummary', this.requestContext)}-${this.config.title}`;
        this.dataCacheType = () => 'request';
        this.config = config || {};
        this.requestContext = requestContext;
    }
}
const action = async (input, context) => {
    const { config } = input;
    return {
        title: config && config.title,
        description: config && config.description,
        keywords: config && config.keywords,
        sharingImageUrl: config && config.sharingImage && config.sharingImage.src,
        canonicalUrl: context.requestContext && context.requestContext.canonicalUrl,
        faviconUrl: config && config.faviconUrl
    };
};
export default createObservableDataAction({
    id: '@msdyn365-commerce-modules/page-summary/default-page-summary',
    action: action,
    input: (args) => {
        return new PageSummaryInput(args.config, args.requestContext);
    }
});
