// @msdyn365-commerce/core-internal/dist/lib/utilities/editable-fields/image-field.js
/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */
import React from 'react';
import { placeholderImageUrl } from '../../components/image';
import SVGImages, { DEFAULT_PLACEHOLDER, LARGE_IMAGE_PLACEHOLDER, SMALL_IMAGE_PLACEHOLDER, TEST_PLACEHOLDER } from '../svg-loader';
import { getEditableFieldClassName, isEditorialMode } from './utils';
export const ImageField = (props) => {
    const onKeyPress = (event) => {
        // 'enter' key
        if (event.charCode === 13 && props.onChange) {
            event.preventDefault();
            props.onChange(event);
        }
    };
    const editableProps = props.disabled
        ? {}
        : {
            onKeyPress: onKeyPress,
            onDoubleClick: props.onChange,
            onMouseOver: props.onMouseOver,
            onFocus: props.onFocus,
            onBlur: props.onBlur,
            onMouseOut: props.onMouseOut,
            tabIndex: 0
        };
    // tslint:disable-next-line: no-any
    const fieldProps = {
        ...props.fieldProps,
        editProps: undefined
    };
    // If the image source is the placeholder image, set it's height to 1px
    if (props.fieldProps.src === placeholderImageUrl &&
        (!props.fieldProps['data-src'] || props.fieldProps['data-src'] === placeholderImageUrl)) {
        fieldProps.height = '1px';
    }
    if (fieldProps.placeholderConfig && props.requestContext && isEditorialMode(props.requestContext)) {
        switch (fieldProps.placeholderConfig.name) {
            case DEFAULT_PLACEHOLDER:
                return React.createElement(SVGImages.largePicturePlaceholder, Object.assign({}, fieldProps.placeholderConfig.svgConfig, editableProps, { className: getEditableFieldClassName(props.disabled, props.fieldProps.className, props.requestContext) }));
            // TODO:Add more case when we get full assets of placeholder SVGs
            case LARGE_IMAGE_PLACEHOLDER:
                return React.createElement(SVGImages.largePicturePlaceholder, Object.assign({}, fieldProps.placeholderConfig.svgConfig, editableProps, { className: getEditableFieldClassName(props.disabled, props.fieldProps.className, props.requestContext) }));
            case SMALL_IMAGE_PLACEHOLDER:
                return React.createElement(SVGImages.smallPicturePlaceholder, Object.assign({}, fieldProps.placeholderConfig.svgConfig, editableProps, { className: getEditableFieldClassName(props.disabled, props.fieldProps.className, props.requestContext) }));
            case TEST_PLACEHOLDER:
                return React.createElement(SVGImages.testPlaceholder, Object.assign({}, fieldProps.placeholderConfig.svgConfig, editableProps, { className: getEditableFieldClassName(props.disabled, props.fieldProps.className, props.requestContext) }));
            default:
                return React.createElement(SVGImages.largePicturePlaceholder, Object.assign({}, fieldProps.placeholderConfig.svgConfig, editableProps, { className: getEditableFieldClassName(props.disabled, props.fieldProps.className, props.requestContext) }));
        }
    }
    return (React.createElement("img", Object.assign({}, fieldProps, editableProps, { className: getEditableFieldClassName(props.disabled, props.fieldProps.className, props.requestContext) }, (props.fieldProps.additionalProperties || {}))));
};
