// @msdyn365-commerce-modules/core-components/src/modules/default-container-3-slots/default-container-3-slots.props.autogenerated.ts
/**
 * Copyright (c) 2018 Microsoft Corporation
 * IDefaultContainer3Slots containerModule Interface Properties
 * THIS FILE IS AUTO-GENERATED - MANUAL MODIFICATIONS WILL BE LOST
 */

import * as Msdyn365 from '@msdyn365-commerce/core';
import * as React from 'react';

export const enum containerType {
    fluid = 'fluid',
    container = 'container'
}

export const enum xsmall {
    col_auto_auto_auto = 'col_auto_auto_auto',
    col_4_4_4 = 'col_4_4_4',
    col_6_3_3 = 'col_6_3_3',
    col_3_6_3 = 'col_3_6_3',
    col_3_3_6 = 'col_3_3_6',
    col_2_2_8 = 'col_2_2_8',
    col_2_8_2 = 'col_2_8_2',
    col_8_2_2 = 'col_8_2_2'
}

export const enum small {
    col_auto_auto_auto = 'col_auto_auto_auto',
    col_4_4_4 = 'col_4_4_4',
    col_6_3_3 = 'col_6_3_3',
    col_3_6_3 = 'col_3_6_3',
    col_3_3_6 = 'col_3_3_6',
    col_2_2_8 = 'col_2_2_8',
    col_2_8_2 = 'col_2_8_2',
    col_8_2_2 = 'col_8_2_2'
}

export const enum medium {
    col_auto_auto_auto = 'col_auto_auto_auto',
    col_4_4_4 = 'col_4_4_4',
    col_6_3_3 = 'col_6_3_3',
    col_3_6_3 = 'col_3_6_3',
    col_3_3_6 = 'col_3_3_6',
    col_2_2_8 = 'col_2_2_8',
    col_2_8_2 = 'col_2_8_2',
    col_8_2_2 = 'col_8_2_2'
}

export const enum large {
    col_auto_auto_auto = 'col_auto_auto_auto',
    col_4_4_4 = 'col_4_4_4',
    col_6_3_3 = 'col_6_3_3',
    col_3_6_3 = 'col_3_6_3',
    col_3_3_6 = 'col_3_3_6',
    col_2_2_8 = 'col_2_2_8',
    col_2_8_2 = 'col_2_8_2',
    col_8_2_2 = 'col_8_2_2'
}

export interface IDefaultContainer3SlotsConfig extends Msdyn365.IModuleConfig {
    heading?: IHeadingData;
    containerType?: containerType;
    xsmall?: xsmall;
    small?: small;
    medium?: medium;
    large?: large;
    className?: string;
}

export const enum HeadingTag {
    h1 = 'h1',
    h2 = 'h2',
    h3 = 'h3',
    h4 = 'h4',
    h5 = 'h5',
    h6 = 'h6'
}

export interface IHeadingData {
    headingText?: string;
    tag?: HeadingTag;
}

export interface IDefaultContainer3SlotsProps<T> extends Msdyn365.IModule<T> {
    config: IDefaultContainer3SlotsConfig;
    slots: {
        content: React.ReactNode[];
    };
}
