// @msdyn365-commerce-modules/checkout/dist/lib/modules/checkout-payment-instrument/payment-instrument-message.js
export var PaymentConnectorPostMessageType;
(function (PaymentConnectorPostMessageType) {
    PaymentConnectorPostMessageType["CardPrefix"] = "msax-cc-cardprefix";
    PaymentConnectorPostMessageType["Error"] = "msax-cc-error";
    PaymentConnectorPostMessageType["Height"] = "msax-cc-height";
    PaymentConnectorPostMessageType["Result"] = "msax-cc-result";
    PaymentConnectorPostMessageType["Submit"] = "msax-cc-submit";
    PaymentConnectorPostMessageType["ExtraContext"] = "msax-cc-extracontext";
    PaymentConnectorPostMessageType["Redirect"] = "msax-cc-redirect";
    PaymentConnectorPostMessageType["Showoverlay"] = "msax-cc-showoverlay";
    PaymentConnectorPostMessageType["Hideoverlay"] = "msax-cc-hideoverlay";
})(PaymentConnectorPostMessageType || (PaymentConnectorPostMessageType = {}));
export function paymentConnectorMessage(data, messageType) {
    return {
        message: JSON.stringify({
            type: messageType,
            value: data
        }),
        targetOrigin: '*'
    };
}
export function paymentConnectorExtraContextMessage(data) {
    return paymentConnectorMessage(data, PaymentConnectorPostMessageType.ExtraContext);
}
export function paymentConnectorSubmitMessage() {
    return paymentConnectorMessage('true', PaymentConnectorPostMessageType.Submit);
}
