// @msdyn365-commerce/action-internal/dist/lib/__mocks__/client-app/product-search.js
/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */
import { __decorate, __metadata } from "tslib";
import { observer } from 'mobx-react';
import React from 'react';
import GetProductByCategory, { ProductsByCategoryInput } from '../actions/get-products-by-category';
/*@internal*/
let ProductSearchList = class ProductSearchList extends React.Component {
    constructor(props) {
        super(props);
        this.state = { p1: [], p2: [] };
    }
    async componentWillMount() {
        const ctx = this.props.actionContext;
        const [p1, p2] = await Promise.all([
            GetProductByCategory(new ProductsByCategoryInput('', 'AA3'), ctx),
            GetProductByCategory(new ProductsByCategoryInput('C103', ''), ctx)
        ]);
        this.setState({
            p1: p1,
            p2: p2
        });
    }
    render() {
        const { p1 } = this.state;
        return React.createElement("div", null, JSON.stringify(p1));
    }
};
ProductSearchList = __decorate([
    observer,
    __metadata("design:paramtypes", [Object])
], ProductSearchList);
export default ProductSearchList;
