// @msdyn365-commerce-modules/checkout/src/modules/checkout-terms-and-conditions/checkout-terms-and-conditions.props.autogenerated.ts
/**
 * Copyright (c) 2018 Microsoft Corporation
 * ICheckoutTermsAndConditions contentModule Interface Properties
 * THIS FILE IS AUTO-GENERATED - MANUAL MODIFICATIONS WILL BE LOST
 */

import * as Msdyn365 from '@msdyn365-commerce/core';

export interface ICheckoutTermsAndConditionsConfig extends Msdyn365.IModuleConfig {
    heading?: IHeadingData;
    className?: string;
    showControlButtons?: boolean;
    showCheckbox?: boolean;
    description: Msdyn365.RichText;
}

export interface ICheckoutTermsAndConditionsResources {
    errorMessageTitle: string;
    errorMessage: string;
    saveBtnLabel: string;
    editBtnLabel: string;
    cancelBtnLabel: string;
    saveBtnAriaLabel: string;
    editBtnAriaLabel: string;
    cancelBtnAriaLabel: string;
}

export const enum HeadingTag {
    h1 = 'h1',
    h2 = 'h2',
    h3 = 'h3',
    h4 = 'h4',
    h5 = 'h5',
    h6 = 'h6'
}

export interface IHeadingData {
    text: string;
    tag?: HeadingTag;
}

export interface ICheckoutTermsAndConditionsProps<T> extends Msdyn365.IModule<T> {
    resources: ICheckoutTermsAndConditionsResources;
    config: ICheckoutTermsAndConditionsConfig;
}
