// @msdyn365-commerce-modules/order-management/dist/lib/common/get-sales-lines.js
import { getSalesLine } from './get-sales-line';
export const getSalesLines = ({ moduleName, moduleComponentProps, salesLines, products, shipments, resources, isCurrentChannel, imageOnly, showCompleteAddress, telemetryContent }) => {
    return salesLines.map(salesLine => {
        const product = products.find(_product => _product.RecordId === salesLine.ProductId) || {};
        const shipment = shipments && shipments.find(_shipment => _shipment.ShipmentLines?.find(_shipmentLine => _shipmentLine.ItemId === product.ItemId));
        return getSalesLine({
            moduleName,
            moduleComponentProps,
            product,
            shipment,
            salesLine,
            isCurrentChannel,
            resources,
            imageOnly,
            showCompleteAddress,
            telemetryContent
        });
    });
};
export const getSalesLinesInput = (props, moduleName, order, products, priceContext, showCompleteAddress, telemetryContent) => {
    if (!order || !order.SalesLines || !order.SalesLines.length) {
        return;
    }
    const { resources, config: { imageSettings }, context: { actionContext, request: { gridSettings, apiSettings: { channelId } } } } = props;
    const isCurrentChannel = !order.ChannelId || order.ChannelId === channelId;
    return {
        moduleName,
        moduleComponentProps: props,
        isCurrentChannel,
        salesLines: order.SalesLines,
        products,
        shipments: order.Shipments,
        priceContext: priceContext,
        resources,
        actionContext: actionContext,
        gridSettings: gridSettings,
        imageSettings: imageSettings,
        showCompleteAddress: showCompleteAddress,
        telemetryContent: telemetryContent
    };
};
