// @msdyn365-commerce/build-scripts-internal/dist/lib/module-registration/utils/models.js
"use strict";
/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var fs = tslib_1.__importStar(require("fs"));
var path = tslib_1.__importStar(require("path"));
var appDirectory = fs.realpathSync(process.cwd());
exports.keystonePaths = {
    KEYSTONE_THEMES_DIR: path.resolve(appDirectory, 'src', 'themes'),
    KEYSTONE_APP_DEFINITION_FILEPATH: path.resolve(appDirectory, 'lib', 'module-definition-api.json'),
    KEYSTONE_APP_DAPI_FILEPATH: path.resolve(appDirectory, 'lib', 'module-definition-dapi.json'),
    KEYSTONE_APP_DAPI_THEME_PREVIEW_DATA: path.resolve(appDirectory, 'lib', 'theme-preview-dapi.json'),
    KEYSTONE_APP_DAPI_MODULE_PREVIEW_DATA: path.resolve(appDirectory, 'lib', 'module-preview-dapi.json'),
    KEYSTONE_APP_THEMES_DAPI_FILEPATH: path.resolve(appDirectory, 'lib', 'themes-definition-dapi.json'),
    KEYSTONE_APP_RESOURCES_DAPI_FILEPATH: path.resolve(appDirectory, 'lib', 'resources-definition-dapi.json'),
    KEYSTONE_APP_SETTINGS_DAPI_FILEPATH: path.resolve(appDirectory, 'lib', 'themesettings-dapi.json'),
    KEYSTONE_APP_SETTINGS_DIR: path.resolve(appDirectory, 'build', 'settings')
};
/**
 * Extracts the name of the module from the path to its definition file
 * @param definitionPath path to the definition file
 */
exports.getModuleName = function (definitionPath) {
    if (!definitionPath || definitionPath === '') {
        return '';
    }
    return path.basename(definitionPath, ".definition.json" /* DEFINITION_FILE */);
};
var currentDirectory = process.cwd();
var msdyn365CommerceModuleNamespace = '@msdyn365-commerce-modules';
var resolvedHoistedPath = path.resolve(currentDirectory, '../..').replace(/\\/gi, '/');
exports.installedModulesResourcesPathPattern = currentDirectory + "/node_modules/" + msdyn365CommerceModuleNamespace + "/*/dist/lib/**/resources/modules/*.json";
exports.installedAuthoringResourcesPathPattern = currentDirectory + "/node_modules/" + msdyn365CommerceModuleNamespace + "/*/dist/lib/**/resources/authoring/*.json";
exports.hoistedModulesResourcesPathPattern = resolvedHoistedPath + "/node_modules/" + msdyn365CommerceModuleNamespace + "/*/dist/lib/**/resources/modules/*.json";
exports.hoistedAuthoringResourcesPathPattern = resolvedHoistedPath + "/node_modules/" + msdyn365CommerceModuleNamespace + "/*/dist/lib/**/resources/authoring/*.json";
exports.localModulesResourcesPathPattern = currentDirectory + "/src/resources/modules/*.json";
exports.localAuthoringResourcesPathPattern = currentDirectory + "/src/resources/authoring/*.json";
