// @msdyn365-commerce-modules/account-management/dist/lib/modules/account-generic-tile/account-generic-tile.js
import * as Msdyn365 from '@msdyn365-commerce/core';
import classnames from 'classnames';
import * as React from 'react';
import { getTelemetryObject } from '@msdyn365-commerce-modules/utilities';
import LinksComponent from './components/account-generic-tile-links';
class AccountGenericTile extends React.PureComponent {
    constructor(props) {
        super(props);
        this.handleHeadingChange = (event) => this.props.config.heading.text = event.target.value;
        this.handleLinkTextChange = (linkIndex) => (event) => {
            if (this.props.config.links && this.props.config.links[linkIndex]) {
                this.props.config.links[linkIndex].linkText = event.target.value;
            }
        };
        this.telemetryContent = getTelemetryObject(this.props.context.request.telemetryPageName, this.props.friendlyName, this.props.telemetry);
    }
    render() {
        const { config } = this.props;
        const { className, heading, links } = config;
        const viewProps = {
            ...this.props,
            className: className,
            AccountGenericTile: {
                moduleProps: this.props,
                className: classnames('ms-account-generic-tile', config.className)
            },
            heading: heading && heading.text && (React.createElement(Msdyn365.Text, { className: 'ms-account-generic-tile__heading', tag: heading.tag || 'h2', text: heading.text, editProps: { onEdit: this.handleHeadingChange, requestContext: this.props.context.request } })),
            links: links && links.length > 0 && React.createElement(LinksComponent, Object.assign({}, { links: links, onTextChange: this.handleLinkTextChange, requestContext: this.props.context.request, telemetryContent: this.telemetryContent }))
        };
        return this.props.renderView(viewProps);
    }
}
export default AccountGenericTile;
