// @msdyn365-commerce-modules/utilities/src/collapse/Collapse.examples.tsx
import * as React from 'react';
import Button from '../button/Button';
import Collapse from './Collapse';
import { ICollapseProps } from './Collapse.props';

export interface ICollapseState {
    expanded: boolean;
}

/**
 * CollapseExample
 */
export class CollapseExample extends React.Component<ICollapseProps, ICollapseState> {
    public props: ICollapseProps;
    public state: ICollapseState;

    constructor(props: ICollapseProps) {
        super(props);
        this.props = props;
        this.toggle = this.toggle.bind(this);
        this.state = {expanded: false};
    }

    public render(): JSX.Element {
        return (
            <div>
                <Button color='primary' tag='button' onClick={this.toggle} aria-expanded={this.state.expanded}>Toggle</Button>
                <Collapse isOpen={this.state.expanded} timeout={350}>
                    <div className='card'>
                        This is an example div.
                        It will collapse and expand by clicking on 'Toggle' button.
                    </div>
                </Collapse>
            </div>
        );
    }

    private toggle = () => {
        this.setState(prevState => ({
            expanded: !prevState.expanded
        }));
    }
}
