// @msdyn365-commerce-modules/order-template/src/modules/order-template/components/order-template-table-heading.tsx
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
import * as React from 'react';

export interface IOrderTemplateTableHeading {
    productNumber: string;
    productName: string;
    productUnitPrice: string;
    productUOM: string;
    productQuantity: string;
    productActions: string;
    clickHandler(isChecked: boolean): void;
}

export interface IOrderTemplateTableShortHeading {
    productNumber: string;
    productActions: string;
}

/**
 * Order Template's table heading
 */
export class OrderTemplateTableHeading extends React.PureComponent<IOrderTemplateTableHeading> {
    public state: {
        isChecked: boolean;
    } = {
        isChecked: false
    };

    public getOnChange = () => (event: React.ChangeEvent<HTMLInputElement>) => {
        this.setState({
            isChecked: !this.state.isChecked
        }, () => {// tslint:disable-line:align
            this.props.clickHandler(this.state.isChecked);
        });
    };

    public render(): JSX.Element {
        return (
            <thead>
                <tr className='ms-order-template-table-line__heading'>
                    <th>
                        <input type='checkbox' className='bulk-selection-all' aria-checked={this.state.isChecked} onChange={this.getOnChange()}/>
                    </th>
                    <th/>
                    <th>{this.props.productNumber}</th>
                    <th>{this.props.productName}</th>
                    <th>{this.props.productUnitPrice}</th>
                    <th>{this.props.productUOM}</th>
                    <th>{this.props.productQuantity}</th>
                    <th>{this.props.productActions}</th>
                </tr>
            </thead>
        );
    }
}
