// @msdyn365-commerce-modules/social-share/dist/lib/modules/social-share/social-share.js
import classnames from 'classnames';
import * as React from 'react';
class SocialShare extends React.PureComponent {
    constructor(props) {
        super(props);
    }
    render() {
        const { slots } = this.props;
        if (slots && slots.socialShareItems.length === 0) {
            this.props.context.telemetry.error('No social media is enable for sharing, module wont render');
            return null;
        }
        const isConsentGiven = this.props.context.request &&
            this.props.context.request.cookies &&
            this.props.context.request.cookies.isConsentGiven &&
            this.props.context.request.cookies.isConsentGiven();
        if (isConsentGiven === false) {
            return null;
        }
        const viewprops = {
            ...this.props,
            SocialShare: {
                moduleProps: this.props,
                className: classnames('ms-social-share', this.props.config.className)
            },
            SocialShareList: {
                'data-title': this.props.config.caption,
                className: `ms-social-share-ul ${this.props.config.caption && 'caption'} ${this.props.config.orientation === "vertical" ? 'vertical' : ''}`,
                tag: 'ul'
            },
            SocialShareItem: {
                className: 'ms-social-share-li',
                tag: 'li',
                tabIndex: '0'
            },
            SocialShareItemElements: slots && slots.socialShareItems.length > 0 && slots.socialShareItems
        };
        return this.props.renderView(viewprops);
    }
}
export default SocialShare;
