// @msdyn365-commerce/action-internal/dist/lib/__mocks__/actions/get-categories.js
/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */
import { createDataAction } from '../../create-data-action';
/*@internal*/
export const numberOfTimesGetProductCalledDuringTest = 0;
/**
 * test class
 */
/*@internal*/
export class CategoryInput {
    constructor(id) {
        this.shouldCacheOutput = () => true;
        this.getCacheObjectType = () => 'Category';
        this.getCacheKey = () => this.channelId;
        this.channelId = id;
    }
}
async function getCategories(inputs, ctx) {
    ctx.trace(`Calling GetCategories with category: ${inputs.channelId}`);
    return Promise.resolve(mockCategories);
}
/*@internal*/
export const mockCategories = [
    { Id: 'C100', Name: 'AA1' },
    { Id: 'C101', Name: 'AA2' },
    { Id: 'C102', Name: 'AA3' },
    { Id: 'C103', Name: 'AA4' },
    { Id: 'C104', Name: 'AA5' }
];
/*@internal*/
export default createDataAction({
    action: getCategories,
    input: () => new CategoryInput('mock-chanel-id')
});
