// @msdyn365-commerce/action-internal/dist/lib/__mocks__/actions/get-cart-hook.js
/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */
// import { Cart } from '@msdyn365-commerce/retail-proxy';
export const beforeCart = async (inputs) => {
    // tslint:disable-next-line: no-console
    console.info('Pre Hook');
};
export const afterCart = async (_inputs, cartName) => {
    // tslint:disable-next-line: no-console
    console.info('Post Hook');
};
export const preReadOnlyCart = async (inputs) => {
    // tslint:disable-next-line: no-console
    console.info('Pre ReadOnly Hook');
};
export const postReadOnlyCart = async (_inputs, cartName) => {
    // tslint:disable-next-line: no-console
    console.info('Post ReadOnly Hook');
};
export const preReadOnlyCartError = async (inputs) => {
    // tslint:disable-next-line: no-console
    throw 'PreReadOnlyError';
};
export const postReadOnlyCartError = async (_inputs, cartName) => {
    // tslint:disable-next-line: no-console
    throw 'PostReadOnlyError';
};
export const afterCartError = async (_inputs, cartName) => {
    // tslint:disable-next-line: no-console
    throw new Error('afterCartHookError');
};
