// @msdyn365-commerce/action-internal/dist/lib/__mocks__/actions/get-list-data.js
/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */
import { createDataAction } from '../../create-data-action';
import GetRecoProducts, { RecoInput } from './get-recommendation-products';
/*@internal*/
export let numberOfTimesGetistDataCalledDuringTest = 0;
/*@internal*/
export class ListInput {
    constructor(id) {
        this.shouldCacheOutput = () => false;
        this.getCacheObjectType = () => 'LIST';
        this.getCacheKey = () => this.Channel;
        this.Channel = id;
    }
}
async function GetListData(input, ctx) {
    ctx.trace('List Channels: ' + input.Channel);
    numberOfTimesGetistDataCalledDuringTest++;
    // If we want to move the reco call to the next TIER, un-comment the below line
    // return await ctx.chainAction(GetRecoProducts, new RecoInput(input.Channel));
    return GetRecoProducts(new RecoInput(input.Channel), ctx);
}
const createListInput = (arg) => {
    if (arg.config && arg.data.list && arg.data.list.listType === 'Reco' && arg.data.list.channel) {
        return new RecoInput(arg.data.list.channel);
    }
    throw new Error('data is not well defined for list action input');
};
/*@internal*/
export default createDataAction({
    action: GetListData,
    input: createListInput
});
