// @msdyn365-commerce-modules/order-template/src/modules/order-template/order-template.props.autogenerated.ts
/**
 * Copyright (c) 2018 Microsoft Corporation
 * IOrderTemplate contentModule Interface Properties
 * THIS FILE IS AUTO-GENERATED - MANUAL MODIFICATIONS WILL BE LOST
 */

import * as Msdyn365 from '@msdyn365-commerce/core';

export interface IOrderTemplateConfig extends Msdyn365.IModuleConfig {
    heading?: IHeadingData;
    imageSettings?: Msdyn365.IImageSettings;
    className?: string;
    itemsPerPage?: number;
}

export interface IOrderTemplateResources {
    emptyOrderTemplateLinesText: string;
    priceFree: string;
    removeItemFromOrderTemplateButtonText: string;
    addOrderTemplateItemToBagButtonText: string;
    addSelectedToBagButtonText: string;
    removeSelectedToBagButtonText: string;
    waitingClass: string;
    addedToCartSuccessMessage: string;
    addedToCartFailureMessage: string;
    maxQuantityLimitText: string;
    cancelBtnLabel: string;
    originalPriceText: string;
    currentPriceText: string;
    productDimensionTypeColor: string;
    productDimensionTypeConfiguration: string;
    productDimensionTypeSize: string;
    productDimensionTypeStyle: string;
    errorGettingOrderTemplateLines: string;
    productNumberText: string;
    productText: string;
    productUnitPriceText: string;
    productUOMText: string;
    productQuantityText: string;
    productActionsText: string;
    addTemplateToBagButtonText: string;
    addLineModalLinkText: string;
    addItemToTemplateText: string;
    searchModalPlaceholderText: string;
    selectProductButtonText: string;
    backButtonText: string;
    deleteTemplateButtonText: string;
    orderTemplateTitle: string;
    renameTemplateButtonText: string;
    renameOrderTemplateTitleText: string;
    renameOrderTemplateButtonText: string;
    inputQuantityAriaLabel: string;
    decrementButtonAriaLabel: string;
    incrementButtonAriaLabel: string;
    addToCartFailureMessage: string;
    addToCartSuccessMessage: string;
    addToCartProcessMessage: string;
    removeSelectedSuccessMessage: string;
    removeSelectedFailureMessage: string;
    closeWindowButtonText: string;
    deleteOrderTemplateFailedText: string;
    flipperNext: string;
    flipperPrevious: string;
    paginationAriaLabel: string;
    viewCartButtonText: string;
    orderTemplateCloseButtonText: string;
    linesAddedToCartHeaderItemsOneText: string;
    linesAddedToCartHeaderLinesOneText: string;
    linesAddedToCartHeaderLinesFormatText: string;
    linesAddedToCartHeaderMessageText: string;
    itemAddedToCartHeaderItemOneText: string;
    itemAddedToCartHeaderItemFormatText: string;
    itemAddedToCartHeaderMessageText: string;
    freePriceText: string;
    quantitySelectLabel: string;
    searchErrorMessage: string;
    notFoundSearchErrorRedediation: string;
    notFoundSearchErrorNotice: string;
    addLineProductUnitPricePrefix: string;
    addLineProductUnitOfMeasurePrefix: string;
    searchButtonAriaLabel: string;
    searchInputAriaLabel: string;
    searchResultsCountVerbage: string;
    searchResultsCountSubject: string;
    addToTemplateConfirmation: string;
    addToTemplateError: string;
    totalPriceLabel: string;
    defaultOrderTemplateName: string;
    progressNotificationText: string;
}

export const enum HeadingTag {
    h1 = 'h1',
    h2 = 'h2',
    h3 = 'h3',
    h4 = 'h4',
    h5 = 'h5',
    h6 = 'h6'
}

export interface IHeadingData {
    text: string;
    tag?: HeadingTag;
}

export interface IOrderTemplateProps<T> extends Msdyn365.IModule<T> {
    resources: IOrderTemplateResources;
    config: IOrderTemplateConfig;
}
