// @msdyn365-commerce-modules/utilities/src/modal/Modal.examples.tsx
import * as React from 'react';
import Modal from './Modal';
import ModalBody from './ModalBody';
import ModalFooter from './ModalFooter';
import ModalHeader from './ModalHeader';

export interface IModalExampleProps {
  buttonLabel: string;
  className: string;
  applicationNode: string;
}
export interface IModalExampleState {
  isOpen: boolean;
}

/**
 * Modal example
 */
class ModalExample extends React.Component<IModalExampleProps, IModalExampleState> {

// tslint:disable-next-line: no-any
  public modalExampleBtn: React.RefObject<any>;

  constructor(props: IModalExampleProps) {
    super(props);
    this.state = {
      isOpen: false
    };

    this.toggle = this.toggle.bind(this);
    this.modalExampleBtn = React.createRef();
  }

  public toggle(): void {
    this.setState({
      isOpen: !this.state.isOpen
    });
  }

  public render(): JSX.Element {
    return (
      <div>
        <button ref={this.modalExampleBtn} color='danger' onClick={this.toggle}>{this.props.buttonLabel}</button>
        <Modal
          autoFocus={true}
          fade={false}
          applicationNode={'rsg-root'}
          returnFocusRef={this.modalExampleBtn}
          isOpen={this.state.isOpen}
          toggle={this.toggle}
          className={this.props.className}
          horizontalPosition={'right'}
          verticalPosition={'top'}
        >
          <ModalHeader toggle={this.toggle}>Modal title</ModalHeader>
          <ModalBody>
            Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor
            incididunt ut labore et dolore magna aliqua.
            Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip
            ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit
            esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non
            proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
          </ModalBody>
          <ModalFooter>
            <button color='primary' onClick={this.toggle}>Do Something</button>{' '}
            <button color='secondary' onClick={this.toggle}>Cancel</button>
          </ModalFooter>
        </Modal>
      </div>
    );
  }
}

export default ModalExample;
