// @msdyn365-commerce/components/src/videoplayer/components/video-controls/VideoControls.props.ts
/*
 * Copyright (c) 2018 Microsoft Corporation
 */
import { ISliderChangeNotification, ITelemetryContent } from '@msdyn365-commerce-modules/utilities';
import { ICurrentClosedCaptionsSettings, IPlayerMenuItemClickNotification, IPlayPosition } from '../../components/data/player-constants';
import { IPlayerMenu } from '../data/player-constants';
import { IPlayerResources } from '../data/player-resources';

export interface IVideoControlsProps {
    /** Classe name for the element */
    className?: string;
    /** Video is currently playing or not */
    playing?: boolean;
    /** Current Video playing State */
    playPosition: IPlayPosition;
    /** Current Video mute/unmuteState */
    isMuted?: boolean;
    /** Current Volume */
    currentVolume: number;
    /** Full Screen */
    isInFullScreen?: boolean;
    /** volumeSeek Stpes */
    volumeSeekSteps?: number;
    /** videoSeek Stpes */
    videoSeekSteps?: number;
    /** resources for video controls */
    resources: IPlayerResources;
    /** control bar visibility  */
    isControlPanelVisible: boolean;
    /** Close caption Menu */
    closedCaptionMenu?: IPlayerMenu;
    /** current closed Caption Language */
    currentClosedCaptionLanguage: string;
    /** current closed caption settings */
    currentClosedCaptionsSetting: ICurrentClosedCaptionsSettings;
    /** options menu */
    optionMenu?: IPlayerMenu;
    /** current audio track */
    currentAudioTrack: string;
    /** current audio track */
    currentAudioTrackURL: string;
    /** To set the viewport min width */
    viewportMin?: number;
    /** To set the viewport max width */
    viewportMax?: number;
    /** Is mobile devices */
    isMobileDevice: boolean;
    /** The telemetry content */
    telemetryContent?: ITelemetryContent;
    /** Click event handler for play/pause Button */
    onPlayClick?(): void;
    /** Click event handler for mute/unMute Button */
    onMuteClick?(): void;
    /** Progess changed event handler */
    onProgressChange?(notification: ISliderChangeNotification): void;
    /** Volume changed event handler */
    onVolumeChange?(notification: ISliderChangeNotification): void;
    /** toggle full screen change */
    toggleFullScreen?(): void;
    /** KeyBoard Event handler */
    handleControlsKeyDownHandler?(e: React.KeyboardEvent): void;
    /** onMenuItemClick Event Handler */
    onMenuItemClick(notification: IPlayerMenuItemClickNotification): void;
    /** on Menu Item Foucs */
    onMenuItemFocus(notification: IPlayerMenuItemClickNotification): void;
    /** on Menu Item Blur */
    onMenuItemBlur(): void;
    /** on component focus */
    onFocus?(): void;
    /** on component blur */
    onBlur?(): void;
}
