// @msdyn365-commerce/components/dist/lib/fade/Fade.js
import { mapToCssModules, TransitionTimeouts } from '@msdyn365-commerce-modules/utilities';
import classNames from 'classnames';
import * as React from 'react';
import Transition from 'react-transition-group/Transition';
export default class Fade extends React.Component {
    render() {
        const { tag, className, cssModule, baseClass, baseClassActive, appear, enter, exit, in: fadeIn, mountOnEnter, unmountOnExit, timeout: timeoutOrUndefined, addEndListener, onEnter, onEntering, onEntered, onExit, onExiting, onExited, children, ...otherProps } = this.props;
        const timeout = timeoutOrUndefined !== undefined ? timeoutOrUndefined : TransitionTimeouts.Fade;
        const transitionProps = {
            appear,
            enter,
            exit,
            in: fadeIn,
            mountOnEnter,
            unmountOnExit,
            timeout,
            addEndListener,
            onEnter,
            onEntering,
            onEntered,
            onExit,
            onExiting,
            onExited,
            children
        };
        const Tag = this.props.tag ? this.props.tag : 'div';
        const baseClassAlways = baseClass || 'fade';
        const baseClassActiveAlways = baseClassActive || 'show';
        return (React.createElement(Transition, Object.assign({}, transitionProps), (status) => {
            const isActive = status === 'entered';
            const classes = mapToCssModules(classNames([
                className,
                baseClassAlways,
                isActive && baseClassActiveAlways
            ]), cssModule);
            return (React.createElement(Tag, Object.assign({ className: classes }, otherProps), children));
        }));
    }
}
Fade.defaultProps = {
    tag: 'div',
    baseClass: 'fade',
    baseClassActive: 'show',
    timeout: TransitionTimeouts.Fade,
    in: true,
    appear: true,
    enter: true,
    exit: true
};
