// @msdyn365-commerce/components/src/videoplayer/players/player/Player.props.ts
// tslint:disable:no-single-line-block-comment
/**
 * Copyright (c) 2018 Microsoft Corporation
 */
import { ITelemetryContent } from '@msdyn365-commerce-modules/utilities';
import { IVideoMetadata } from '../../components/data/player-metadata';
import { IPlayerOptions } from '../../components/data/player-options';
import { IPlayerResources } from '../../components/data/player-resources';

/**
 * IPlayerProps
 */
export interface IPlayerProps {
    /** The player container Id */
    playerId?: string;
    /** The player data that include player options and video metadata */
    playerData: IPlayerData;
    /** The player localization resource */
    playerResources?: IPlayerResources;
    /** The telemetry content */
    telemetryContent?: ITelemetryContent;
    /** is Editor  */
    isEditor?: boolean;
}

/**
 * IPlayerData
 */
export interface IPlayerData {
    /** The player options, like autoplay, mute configuration */
    options: IPlayerOptions;
    /** The player metadata, like video source, closed caption source, poster image source */
    metaData: IVideoMetadata;
}
