// @msdyn365-commerce-modules/address/dist/lib/actions/get-address-purposes.js
import { buildCacheKey } from '@msdyn365-commerce-modules/retail-actions';
import { createObservableDataAction } from '@msdyn365-commerce/core';
import { getAddressPurposesAsync } from '@msdyn365-commerce/retail-proxy/dist/DataActions/StoreOperationsDataActions.g';
export class GetAddressPurposesInput {
    constructor(apiSettings) {
        this.getCacheKey = () => buildCacheKey('AddressPurpose', this.apiSettings);
        this.getCacheObjectType = () => 'AddressPurpose';
        this.dataCacheType = () => 'request';
        this.apiSettings = apiSettings;
    }
}
const createInput = (inputData) => {
    return new GetAddressPurposesInput(inputData.requestContext.apiSettings);
};
export async function getAddressPurposesAction(input, ctx) {
    return getAddressPurposesAsync({ callerContext: ctx, queryResultSettings: {} });
}
export default createObservableDataAction({
    id: '@msdyn365-commerce-modules/address/get-address-purposes',
    action: getAddressPurposesAction,
    input: createInput
});
