// @msdyn365-commerce-modules/account-management/dist/lib/modules/account-profile/components/account-profile-links.js
import { getPayloadObject, getTelemetryAttributes } from '@msdyn365-commerce-modules/utilities';
import * as Msdyn365 from '@msdyn365-commerce/core';
import * as React from 'react';
const AccountProfileLinks = ({ links, telemetryContent, requestContext, onTextChange }) => {
    if (links.length === 0) {
        return null;
    }
    const editableLinks = _mapEditableLinks(links, telemetryContent);
    return (React.createElement("div", { className: 'ms-account-profile__section-links' }, editableLinks && editableLinks.length ?
        React.createElement(Msdyn365.Links, { links: editableLinks, editProps: { onTextChange: onTextChange, requestContext: requestContext } })
        : null));
};
const _mapEditableLinks = (linkdata, telemetryContent) => {
    if (!linkdata || linkdata.length === 0) {
        return null;
    }
    const editableLinks = [];
    const payLoad = getPayloadObject('click', telemetryContent, '');
    linkdata.forEach((link, index) => {
        payLoad.contentAction.etext = link.linkText;
        const attributes = getTelemetryAttributes(telemetryContent, payLoad);
        const editableLink = {
            ariaLabel: link.ariaLabel,
            className: 'ms-account-profile__section-link',
            linkText: link.linkText,
            linkUrl: link.linkUrl.destinationUrl,
            openInNewTab: link.openInNewTab,
            role: 'button',
            additionalProperties: attributes
        };
        editableLinks.push(editableLink);
    });
    return editableLinks;
};
export default AccountProfileLinks;
