// @msdyn365-commerce-modules/account-management/dist/lib/modules/loyalty-terms/components/loyalty-terms-modal.js
import { Button, getPayloadObject, getTelemetryAttributes, Modal, ModalBody, ModalFooter, ModalHeader, TelemetryConstant } from '@msdyn365-commerce-modules/utilities';
import React from 'react';
export const LoyaltyTermsModal = (props) => {
    const payLoad = getPayloadObject('click', props.telemetryContent, TelemetryConstant.LoyaltyTermsCancel);
    const loyaltyTermCancelAttributes = getTelemetryAttributes(props.telemetryContent, payLoad);
    payLoad.contentAction.etext = TelemetryConstant.LoyaltyTermsSubmit;
    const loyaltyTermSubmitAttributes = getTelemetryAttributes(props.telemetryContent, payLoad);
    payLoad.contentAction.etext = TelemetryConstant.LoyaltyTermsAgree;
    const loyaltyTermCheckboxAttributes = getTelemetryAttributes(props.telemetryContent, payLoad);
    return {
        modal: { tag: Modal, className: 'ms-loyalty-terms__modal', toggle: props.onToggle, applicationNode: 'renderPage', isOpen: props.isOpen, returnFocusRef: props.returnRef },
        modalHeader: { tag: ModalHeader, className: 'ms-loyalty-terms__modal-header', toggle: props.onToggle },
        modalFooter: { tag: ModalFooter, className: 'ms-loyalty-terms__modal-footer' },
        modalBody: { tag: ModalBody, className: 'ms-loyalty-terms__modal-body' },
        cancelButton: React.createElement(Button, Object.assign({ className: 'ms-loyalty-terms__cancel', onClick: props.onToggle }, loyaltyTermCancelAttributes), props.resources.cancelLoyaltyTermsText),
        submitButton: React.createElement(Button, Object.assign({ className: 'ms-loyalty-terms__submit', onClick: props.onSubmit, disabled: !props.checked }, loyaltyTermSubmitAttributes), props.resources.joinLoyaltyTermsText),
        heading: React.createElement("p", { className: 'ms-loyalty-terms__heading' }, props.resources.loyaltyTermsHeading),
        terms: props.terms,
        agreeSection: { className: 'ms-loyalty-terms__agree-section' },
        checkbox: React.createElement("input", Object.assign({ className: 'ms-loyalty-terms__checkbox' }, loyaltyTermCheckboxAttributes, { type: 'checkbox', checked: props.checked, "aria-checked": props.checked, onChange: props.onCheck })),
        agreeText: React.createElement("p", { className: 'ms-loyalty-terms__agree' }, props.resources.loyaltyAgreeToTerms)
    };
};
