// @msdyn365-commerce-modules/business-account-management/src/modules/business-sign-up/business-sign-up.props.autogenerated.ts
/**
 * Copyright (c) 2018 Microsoft Corporation
 * IBusinessSignUp containerModule Interface Properties
 * THIS FILE IS AUTO-GENERATED - MANUAL MODIFICATIONS WILL BE LOST
 */

import * as Msdyn365 from '@msdyn365-commerce/core';
import * as React from 'react';

export interface IBusinessSignUpConfig extends Msdyn365.IModuleConfig {
    heading: IHeadingData;
    className?: string;
    disclaimer?: Msdyn365.RichText;
    confirmationLink: IConfirmationLinkData;
}

export interface IBusinessSignUpResources {
    firstNameLabel: string;
    firstNameMaxLength: string;
    lastNameLabel: string;
    lastNameMaxLength: string;
    emailAddressLabel: string;
    phoneLabelText: string;
    phoneMaxLength: string;
    companyLabel: string;
    companyMaxLength: string;
    companyAddressLabel: string;
    companyAddressMaxLength: string;
    jobTitleLabel: string;
    jobTitleMaxLength: string;
    departmentLabel: string;
    departmentMaxLength: string;
    companySizeLabel: string;
    companySizeMaxLength: string;
    commentLabel: string;
    commentMaxLength: string;
    errorMessage: string;
    signUpButtonText: string;
    signUpButtonArialabel: string;
    cancelButtonText: string;
    cancelButtonArialabel: string;
    requiredFieldMissingSummaryError: string;
    requiredFieldMissingError: string;
    fieldIncorrectError: string;
    partialAddressErrorMessage: string;
}

export const enum HeadingTag {
    h1 = 'h1',
    h2 = 'h2',
    h3 = 'h3',
    h4 = 'h4',
    h5 = 'h5',
    h6 = 'h6'
}

export interface IHeadingData {
    text: string;
    tag?: HeadingTag;
}

export interface IConfirmationLinkData {
    linkText?: string;
    linkUrl: Msdyn365.ILinkData;
    ariaLabel?: string;
    openInNewTab?: boolean;
}

export interface IBusinessSignUpProps<T> extends Msdyn365.IModule<T> {
    resources: IBusinessSignUpResources;
    config: IBusinessSignUpConfig;
    slots: {
        addressSlot: React.ReactNode[];
    };
}
