// @msdyn365-commerce/build-scripts-internal/dist/lib/tasks/buildSiteCss.js
"use strict";
/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var theming_internal_1 = require("@msdyn365-commerce/theming-internal");
var utilities_internal_1 = require("@msdyn365-commerce/utilities-internal");
var lodash_1 = require("lodash");
var path = tslib_1.__importStar(require("path"));
var IResult_1 = require("../models/IResult");
/**
 * Current directory
 * @internal
 */
exports.currentDir = path.resolve('.');
exports.siteStyleEntryFolder = path.join(exports.currentDir, 'src', 'styles');
exports.pageStyleEntries = path.join(exports.siteStyleEntryFolder, '*.style.scss');
exports.outFolder = path.join(exports.currentDir, 'build', 'public', 'static', 'css');
exports.createSassResult = function (pathToEntryPoint, isDebug) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
    var outFile, _a;
    return tslib_1.__generator(this, function (_b) {
        switch (_b.label) {
            case 0:
                outFile = path.join(exports.outFolder, "" + path.basename(path.basename(pathToEntryPoint, '.scss'), '.style') + (isDebug ? '' : '.min') + ".css");
                _a = {
                    outFile: outFile
                };
                return [4 /*yield*/, theming_internal_1.render({
                        file: pathToEntryPoint,
                        includePaths: [path.dirname(pathToEntryPoint)],
                        sourceComments: isDebug,
                        sourceMap: !isDebug,
                        sourceMapContents: true,
                        outputStyle: isDebug ? 'expanded' : 'compressed',
                        outFile: outFile,
                        // embed sourcemaps
                        sourceMapEmbed: !isDebug
                    })];
            case 1: return [2 /*return*/, (_a.result = _b.sent(),
                    _a)];
        }
    });
}); };
/**
 * Creates two build targets - one for .min.css (dev) and one for .css (debug)
 * @param pathToEntryPoint entry point for a file
 */
exports.buildSassResults = function (pathToEntryPoint) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
    var entryPointName;
    return tslib_1.__generator(this, function (_a) {
        entryPointName = path.basename(pathToEntryPoint);
        utilities_internal_1.trace.info("[" + entryPointName + "] Compiling targets for entry point.");
        return [2 /*return*/, Promise.all([true, false].map(function (isDebug) { return exports.createSassResult(pathToEntryPoint, isDebug); }))];
    });
}); };
/**
 * Writes the result to a file and sourcemap if it exists
 * @param sassResult the output form node-sas
 */
exports.writeSassResultToFile = function (sassResult) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
    var entryPointName;
    return tslib_1.__generator(this, function (_a) {
        entryPointName = path.basename(sassResult.outFile, '.style.scss');
        utilities_internal_1.trace.info("[" + entryPointName + "] Attempting to write file.");
        return [2 /*return*/, (Promise.all([
                utilities_internal_1.safeWriteFile(sassResult.outFile, sassResult.result.css),
                sassResult.result.map && utilities_internal_1.safeWriteFile(sassResult.outFile + ".map", sassResult.result.map)
            ].filter(Boolean))
                .then(function () {
                utilities_internal_1.trace.info("[" + entryPointName + "] File written successfully.");
                return IResult_1.GreatSuccess;
            })
                // tslint:disable-next-line:no-any
                .catch(function (err) {
                utilities_internal_1.trace.error("[" + entryPointName + "] Error writing " + sassResult.outFile);
                utilities_internal_1.trace.error(err);
                return IResult_1.NoSuccess;
            }))];
    });
}); };
/**
 * This task build all site level css
 * /styles
 *      /{name1}.style.scss     <------------ {name1} is an entry point for site with name {name1}
 *      /{name2}.style.scss     <------------ {name2} is an entry point for site with name {name2}
 */
exports.default = (function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
    var entryPoints, _a, e_1;
    return tslib_1.__generator(this, function (_b) {
        switch (_b.label) {
            case 0:
                _b.trys.push([0, 3, , 4]);
                return [4 /*yield*/, utilities_internal_1.safeGetAllFilesPath(exports.pageStyleEntries)];
            case 1:
                entryPoints = _b.sent();
                _a = lodash_1.flattenDeep;
                return [4 /*yield*/, Promise.all(entryPoints.map(function (pathToEntryPoint) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                        var sassResults;
                        return tslib_1.__generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, exports.buildSassResults(pathToEntryPoint)];
                                case 1:
                                    sassResults = _a.sent();
                                    return [2 /*return*/, Promise.all(sassResults.map(exports.writeSassResultToFile))];
                            }
                        });
                    }); }))];
            case 2: 
            // build each entry point
            return [2 /*return*/, _a.apply(void 0, [_b.sent()]).reduce(function (previousValue, currentValue) {
                    return { success: previousValue.success && currentValue.success };
                }, { success: true })];
            case 3:
                e_1 = _b.sent();
                utilities_internal_1.trace.error("Error in buildSiteCss task");
                utilities_internal_1.trace.error(e_1);
                return [2 /*return*/, IResult_1.NoSuccess];
            case 4: return [2 /*return*/];
        }
    });
}); });
