// @msdyn365-commerce-modules/retail-actions/dist/lib/order-templates/create-order-template.js
import { createObservableDataAction } from '@msdyn365-commerce/core';
import { createProductListAsync } from '@msdyn365-commerce/retail-proxy/dist/DataActions/ProductListsDataActions.g';
/**
 *  Input class for the CreateOrderTemplate data action.
 */
export class CreateOrderTemplateInput {
    constructor(templateName) {
        this.getCacheKey = () => 'CreateOrderTemplateInput';
        this.getCacheObjectType = () => 'CreateOrderTemplateInput';
        this.dataCacheType = () => 'none';
        this.templateName = templateName;
    }
}
/**
 * createInput method for the CreateOrderTemplate method.
 * @param {ICreateActionContext} inputData The input data passed to the createInput method.
 * @param {string} templateName The name of template.
 */
export const createOrderTemplateInput = (inputData, templateName) => {
    const { requestContext } = inputData;
    if (!requestContext.user.isAuthenticated) {
        throw new Error('Unable to create order template. User is not authenticated.');
    }
    return new CreateOrderTemplateInput(templateName);
};
/**
 * The action method for the CreateOrderTemplate data action.
 * @param {CreateOrderTemplateInput} input The action input.
 * @param {IActionContext} ctx The action context.
 */
export async function createOrderTemplateAction(input, ctx) {
    try {
        const orderTemplate = {
            Id: '',
            Name: input.templateName,
            ProductListTypeValue: 1 /* OrderTemplate */
        };
        const createdProductList = await createProductListAsync({
            callerContext: ctx
        }, orderTemplate);
        if (createdProductList && createdProductList.Id !== '') {
            return createdProductList;
        }
        else {
            ctx.telemetry.error('Created product list was null');
            throw Error('Created product list was null');
        }
    }
    catch (error) {
        ctx.telemetry.error('Not able to create order template', error);
        throw error;
    }
}
/**
 * The CreateOrderTemplate Data Action.
 * Returns created order template.
 */
export default createObservableDataAction({
    id: '@msdyn365-commerce-modules/retail-actions/order-templates/create-order-template',
    action: createOrderTemplateAction,
    input: createOrderTemplateInput
});
