// @msdyn365-commerce-modules/retail-actions/dist/lib/order-templates/get-order-templates.js
import { createObservableDataAction } from '@msdyn365-commerce/core';
import { searchAsync } from '@msdyn365-commerce/retail-proxy/dist/DataActions/ProductListsDataActions.g';
/**
 *  Input class for the GetOrderTemplates data action
 */
export class GetOrderTemplatesInput {
    constructor(productListId, queryResultSettings) {
        this.getCacheKey = () => `${this.productListId}|${this.queryResultSettings?.Paging?.Top}|${this.queryResultSettings?.Paging?.Skip}`;
        this.getCacheObjectType = () => 'ProductList';
        this.dataCacheType = () => 'instance';
        this.shouldCacheOutput = () => false;
        this.productListId = productListId || '';
        this.queryResultSettings = queryResultSettings;
    }
}
/**
 * createInput method for the GetOrderTemplate method
 * @param inputData The input data passed to the createInput method.
 * @param productListId? The product list identifier.
 * @param queryResultSettings? The query result setting required for pagination/ordering.
 */
export const createGetOrderTemplatesInput = (inputData, productListId, queryResultSettings) => {
    const { requestContext } = inputData;
    if (!requestContext.user.isAuthenticated) {
        throw new Error('Unable to create get order template. User is not authenticated.');
    }
    return new GetOrderTemplatesInput(productListId || '', queryResultSettings);
};
/**
 * The action method for the GetOrderTemplates data action
 * @param input The action input
 * @param ctx The action context
 */
export async function getOrderTemplatesAction(input, ctx) {
    try {
        const productListSearchCriteria = {
            ProductListId: input.productListId,
            CustomerId: ctx.requestContext.user.customerAccountNumber
        };
        return searchAsync({
            callerContext: ctx,
            queryResultSettings: input.queryResultSettings
        }, productListSearchCriteria);
    }
    catch (error) {
        ctx.telemetry.error('Not able to get order template(s)', error);
        throw error;
    }
}
/**
 * The GetOrderTemplates Data Action
 * Returns order templates
 */
export default createObservableDataAction({
    id: '@msdyn365-commerce-modules/retail-actions/order-templates/get-order-templates',
    action: getOrderTemplatesAction,
    input: createGetOrderTemplatesInput
});
