// @msdyn365-commerce-modules/retail-actions/dist/lib/order-templates/get-order-templates-preview.js
import { createObservableDataAction } from '@msdyn365-commerce/core';
import { QueryResultSettingsProxy } from '../utilities/QueryResultSettingsProxy';
import getFullOrderTemplates, { GetFullOrderTemplatesInput } from './get-full-order-templates';
/**
 *  Input class for the GetOrderTemplates data action
 */
export class GetOrderTemplatesPreviewInput {
    constructor(queryResultSettings) {
        this.getCacheKey = () => 'GetOrderTemplatesPreview';
        this.getCacheObjectType = () => 'GetOrderTemplatesPreview';
        this.dataCacheType = () => 'none';
        this.queryResultSettings = queryResultSettings;
        this.queryResultSettings.count = true;
        this.queryResultSettings.Paging.Top = 4;
    }
}
/**
 * createInput method for the GetOrderTemplate method
 * @param inputData The input data passed to the createInput method.
 */
export const createGetOrderTemplatesPreviewInput = (inputData) => {
    const queryResultSettings = QueryResultSettingsProxy.fromInputData(inputData).QueryResultSettings;
    return new GetOrderTemplatesPreviewInput(queryResultSettings);
};
/**
 * The action method for the GetOrderTemplatesPreview data action
 * @param input The action input
 * @param context The action context
 */
export async function getOrderTemplatesPreview(input, context) {
    return getFullOrderTemplates(new GetFullOrderTemplatesInput(input.queryResultSettings), context);
}
/**
 * The GetOrderTemplatesPreview Data Action
 * Returns order templates
 */
export default createObservableDataAction({
    id: '@msdyn365-commerce-modules/retail-actions/order-templates/get-order-templates-preview',
    action: getOrderTemplatesPreview,
    input: createGetOrderTemplatesPreviewInput
});
