// @msdyn365-commerce-modules/retail-actions/dist/lib/order-templates/get-full-order-template.js
import { createObservableDataAction } from '@msdyn365-commerce/core';
import { getOrderTemplates, GetOrderTemplatesInput, QueryResultSettingsProxy } from '..';
import { getPaginatedOrderTemplateLinesAction, GetPaginatedOrderTemplateLinesInput } from '..';
import { OrderTemplate } from '../utilities/order-templates';
/**
 *  Input class for the GetOrderTemplates data action
 */
export class GetFullOrderTemplateInput {
    constructor(orderTemplateId, queryResultSettings) {
        this.getCacheKey = () => `GetFullOrderTemplate|${this.orderTemplateId}`;
        this.getCacheObjectType = () => 'GetFullOrderTemplate';
        this.dataCacheType = () => 'none';
        this.orderTemplateId = orderTemplateId;
        this.queryResultSettings = queryResultSettings;
    }
}
/**
 * createInput method for the GetOrderTemplate method
 * @param inputData The input data passed to the createInput method.
 */
export const createGetFullOrderTemplateInput = (inputData, orderTemplateId) => {
    const { requestContext } = inputData;
    if (!requestContext.user.isAuthenticated) {
        throw new Error('Unable to get full order templates. User is not authenticated.');
    }
    const queryResultSettings = QueryResultSettingsProxy.fromInputData(inputData).QueryResultSettings;
    queryResultSettings.count = true;
    return new GetFullOrderTemplateInput(orderTemplateId, queryResultSettings);
};
/**
 * The action method for the GetFullOrderTemplate data action
 * @param input The action input
 * @param ctx The action context
 */
export async function getFullOrderTemplateAction(input, ctx) {
    if (!ctx.requestContext.user.isAuthenticated) {
        throw new Error('Unable to get full order template. User is not authenticated.');
    }
    const inputParams = new GetOrderTemplatesInput(input.orderTemplateId);
    const productLists = await getOrderTemplates(inputParams, ctx);
    if (!productLists || productLists.length !== 1) {
        throw new Error('Unable to get full order template. Order template was not found.');
    }
    const inputLines = new GetPaginatedOrderTemplateLinesInput(productLists[0].Id, input.queryResultSettings);
    const linesState = await getPaginatedOrderTemplateLinesAction(inputLines, ctx);
    return new OrderTemplate(productLists[0], linesState.lines, linesState.totalCount);
}
/**
 * The GetFullOrderTemplate Data Action
 * Returns order templates
 */
export default createObservableDataAction({
    id: '@msdyn365-commerce-modules/retail-actions/order-templates/get-full-order-template',
    action: getFullOrderTemplateAction,
    input: createGetFullOrderTemplateInput
});
