// @msdyn365-commerce-modules/address/dist/lib/common/address-format.data.js
export var AddressItemDisplayType;
(function (AddressItemDisplayType) {
    AddressItemDisplayType["Input"] = "Input";
    AddressItemDisplayType["Dropdown"] = "Dropdown";
    AddressItemDisplayType["Checkbox"] = "Checkbox";
})(AddressItemDisplayType || (AddressItemDisplayType = {}));
export var AddressValidationRuleType;
(function (AddressValidationRuleType) {
    AddressValidationRuleType["Required"] = "Required";
    AddressValidationRuleType["Format"] = "Format";
})(AddressValidationRuleType || (AddressValidationRuleType = {}));
export var AddressItemType;
(function (AddressItemType) {
    AddressItemType[AddressItemType["Name"] = 1001] = "Name";
    AddressItemType[AddressItemType["Phone"] = 1002] = "Phone";
    AddressItemType[AddressItemType["IsPrimary"] = 1003] = "IsPrimary";
    AddressItemType[AddressItemType["AddressTypeValue"] = 1004] = "AddressTypeValue";
    AddressItemType[AddressItemType["ZipCode"] = 1] = "ZipCode";
    AddressItemType[AddressItemType["City"] = 2] = "City";
    AddressItemType[AddressItemType["County"] = 3] = "County";
    AddressItemType[AddressItemType["State"] = 4] = "State";
    AddressItemType[AddressItemType["ThreeLetterISORegionName"] = 5] = "ThreeLetterISORegionName";
    AddressItemType[AddressItemType["Street"] = 6] = "Street";
    AddressItemType[AddressItemType["District"] = 7] = "District";
    AddressItemType[AddressItemType["StreetNumber"] = 8] = "StreetNumber";
    AddressItemType[AddressItemType["BuildingCompliment"] = 9] = "BuildingCompliment";
    AddressItemType[AddressItemType["Postbox"] = 10] = "Postbox";
    AddressItemType[AddressItemType["House_RU"] = 21] = "House_RU";
    AddressItemType[AddressItemType["Flat_RU"] = 22] = "Flat_RU";
    AddressItemType[AddressItemType["CountryOKSMCode_RU"] = 23] = "CountryOKSMCode_RU";
})(AddressItemType || (AddressItemType = {}));
