// @msdyn365-commerce/components/dist/lib/social-media/social-media.components.js
import * as React from 'react';
export var SocialMedia;
(function (SocialMedia) {
    SocialMedia["Facebook"] = "facebook";
    SocialMedia["Twitter"] = "twitter";
    SocialMedia["Pinterest"] = "pinterest";
    SocialMedia["Mail"] = "mail";
    SocialMedia["Linkedin"] = "linkedin";
    SocialMedia["Customization"] = "customization";
})(SocialMedia || (SocialMedia = {}));
export const SocialMediaComponent = (props) => {
    const openWindowFunc = (event) => { return setShareLinks(props); };
    const src = props.src && props.src.src;
    return (React.createElement("img", { onClick: openWindowFunc, src: src, alt: props.socialMediaName || props.socialMedia, title: props.socialMediaName || props.socialMedia }));
};
const setShareLinks = (props) => {
    const pageUrl = encodeURIComponent(props.postUrl);
    const mediaImage = document.getElementsByName('twitter:image')[0].getAttribute('content');
    let url;
    switch (props.socialMedia) {
        case SocialMedia.Facebook:
            url = `https://www.facebook.com/sharer.php?u=${pageUrl}`;
            break;
        case SocialMedia.Twitter:
            url = `https://twitter.com/share?url=${pageUrl}${props.postTitle ? `&text=${props.postTitle}` : ''}`;
            break;
        case SocialMedia.Pinterest:
            url = `https://www.pinterest.com/pin/create/button/?url=${pageUrl}${mediaImage ? `&media=${mediaImage}` : ''}${props.postTitle ? `&description=${props.postTitle}` : ''}`;
            break;
        case SocialMedia.Mail:
            url = `mailto:?body=Check%20out%20this%20site:${pageUrl}${props.postTitle ? `&subject=${props.postTitle}` : ''}`;
            break;
        case SocialMedia.Linkedin:
            url = `https://www.linkedin.com/shareArticle?mini=true&url=${pageUrl}${props.postTitle ? `&title=${props.postTitle}` : ''}`;
            break;
        case SocialMedia.Customization:
            url = props.socialShareURL && `${props.socialShareURL}${pageUrl}`;
            break;
        default:
    }
    url && openWindow(url);
};
function openWindow(url) {
    const left = (screen.width - 570) / 2;
    const top = (screen.height - 570) / 2;
    const params = `menubar=no,toolbar=no,status=no,width=570,height=570,top='${top}',left='${left}`;
    window.open(url, 'NewWindow', params);
}
