// @msdyn365-commerce/retail-proxy/dist/exceptions/ErrorHelper.js
/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */
import { compare, EMPTY, hasElements } from '../extensions/StringExtensions';
import { ProxyError } from './ProxyError';
// TODO - refactor the logging completely to use const and trace instead of console and class
/**
 * Error type enum.
 */
export class ErrorTypeEnum {
}
ErrorTypeEnum.Cash = '1';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDDEVICETOKEN = 'MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDDEVICE';
// Client error codes mapped to locale strings
ErrorTypeEnum.APPLICATION_ERROR = 'string_29000';
ErrorTypeEnum.APPLICATION_STORE_INITIALIZATION_DATA_FAILED_TO_LOAD = 'string_29007';
ErrorTypeEnum.PRICE_CHECK_INITIALIZATION_DATA_FAILED_TO_LOAD = 'string_29022';
ErrorTypeEnum.CART_LINE_MISSING_PRODUCT_NAME_ERROR = 'string_1242';
ErrorTypeEnum.OPERATOR_ID_PASSWORD_NOT_SPECIFIED = 'string_29001';
ErrorTypeEnum.OPERATOR_PASSWORD_NOT_SPECIFIED = 'string_29019';
ErrorTypeEnum.SERVER_ERROR = 'string_29002';
ErrorTypeEnum.INVALID_URL = 'string_1330';
ErrorTypeEnum.INVALID_EMAIL = 'string_1331';
ErrorTypeEnum.INVALID_PHONE = 'string_1332';
ErrorTypeEnum.INVALID_NAME = 'string_1360';
ErrorTypeEnum.INVALID_NAME_FORMAT = 'string_1361';
ErrorTypeEnum.EMPTY_STREET = 'string_1333';
ErrorTypeEnum.EMPTY_CITY = 'string_1334';
ErrorTypeEnum.EMPTY_STATE = 'string_1335';
ErrorTypeEnum.EMPTY_COUNTRY = 'string_1336';
ErrorTypeEnum.EMPTY_ZIPCODE = 'string_1337';
ErrorTypeEnum.EMPTY_NAME = 'string_1338';
ErrorTypeEnum.LINE_ITEM_MISSING_PRODUCT_NAME_ERROR = 'string_3204';
ErrorTypeEnum.LOGOFF_ERROR = 'string_1353';
ErrorTypeEnum.AAD_AUTHENTICATION_FAILED = 'string_1442';
ErrorTypeEnum.DEVICE_ACTIVATION_DETAILS_NOT_SPECIFIED = 'string_1407';
ErrorTypeEnum.DEVICE_DEACTIVATION_INCOMPLETE_TRANSACTION_ERROR = 'string_1421';
ErrorTypeEnum.CHANGE_PASSWORD_DETAILS_NOT_SPECIFIED = 'string_6805';
ErrorTypeEnum.NEW_PASSWORD_AND_CONFIRMATION_NOT_MATCHING_ERROR = 'string_6806';
ErrorTypeEnum.RESET_PASSWORD_DETAILS_NOT_SPECIFIED = 'string_6810';
ErrorTypeEnum.OLD_AND_NEW_PASSWORD_MATCHING_ERROR = 'string_6607';
ErrorTypeEnum.OPERATION_ISSUE_CREDIT_MEMO_NOT_AVAILABLE = 'string_29801';
ErrorTypeEnum.PAYMENT_INFORMATION_INCOMPLETE = 'string_1137';
ErrorTypeEnum.PAYMENT_INVALID_NUMBER = 'string_1138';
ErrorTypeEnum.PAYMENT_CARD_NOT_SUPPORTED = 'string_1139';
ErrorTypeEnum.PAYMENT_CASH_PAYMENT_NOT_AVAILABLE = 'string_1142';
ErrorTypeEnum.PAYMENT_CARD_PAYMENT_NOT_AVAILABLE = 'string_1158';
ErrorTypeEnum.PAYMENT_UNABLE_TO_LOAD_CURRENCY_AMOUNTS = 'string_1143';
ErrorTypeEnum.PAYMENT_CUSTOMER_ACCOUNT_NOT_SET = 'string_1154';
ErrorTypeEnum.PAYMENT_AMOUNT_CANNOT_BE_EMPTY = 'string_1159';
ErrorTypeEnum.PAYMENT_CARD_TRACK_DATA_EMPTY = 'string_1166';
ErrorTypeEnum.PAYMENT_CARD_NUMBER_EMPTY = 'string_1167';
ErrorTypeEnum.PAYMENT_CARD_SECURITY_CODE_EMPTY = 'string_1168';
ErrorTypeEnum.PAYMENT_CREDIT_MEMO_NUMBER_EMPTY = 'string_1169';
ErrorTypeEnum.PAYMENT_LOYALTY_CARD_NUMBER_EMPTY = 'string_1170';
ErrorTypeEnum.PAYMENT_GIFT_CARD_NUMBER_EMPTY = 'string_1171';
ErrorTypeEnum.PAYMENT_CUSTOMER_ACCOUNT_EMPTY = 'string_1172';
ErrorTypeEnum.PAYMENT_INVALID_CARD_NUMBER = 'string_1175';
ErrorTypeEnum.PAYMENT_INVALID_SECURITY_CODE = 'string_1176';
ErrorTypeEnum.PAYMENT_INVALID_ZIP_CODE = 'string_1177';
ErrorTypeEnum.CART_IS_EMPTY = 'string_29008';
ErrorTypeEnum.RETAIL_SERVER_REDIRECT_ERROR = 'RETAIL_SERVER_REDIRECT_ERROR';
ErrorTypeEnum.OPERATION_NOT_VALID_FOR_BIG_ENDIAN_SYSTEM = 'string_29824';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_MANUALCARDNUMBERNOTALLOWED = 'string_29825';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_TENDERLINECANNOTBEVOIDED = 'string_29826';
ErrorTypeEnum.INVALID_INCOME_EXPENSE_LINE_COLLECTION = 'string_4122';
ErrorTypeEnum.CARTLINE_DISCOUNTINUED = 'string_4452';
ErrorTypeEnum.SCALE_UNSPECIFIED_WITHOUT_MANUAL_ENTRY = 'string_5316';
ErrorTypeEnum.SCALE_RETURNED_ZERO_WITHOUT_MANUAL_ENTRY = 'string_5317';
ErrorTypeEnum.APPLICATION_CONFIGURATION_LOADING_ERROR = 'string_29951';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_DEMOMODEOPERATIONNOTSUPPORTED = 'string_29952';
ErrorTypeEnum.MANAGER_OVERRIDE_CANCELED_ERROR = 'string_29029';
ErrorTypeEnum.SELECTED_CARTLINES_CONTAINS_VOIDED_PRODUCTS = 'string_29832';
ErrorTypeEnum.CUSTOMERORDER_MANUAL_DEPOSIT_REQUIRED = 'string_29833';
// Client error codes - store information and details
ErrorTypeEnum.STORE_NOT_FOUND = 'string_29016';
// Client error codes - return - mapped to local strings
ErrorTypeEnum.DIMENSION_SELECTION_NOT_COMPLETED = 'string_821';
ErrorTypeEnum.RETURN_MULTIPLE_REASON_CODE_SETS_ARE_DEFINED = 'RETURN_MULTIPLE_REASON_CODE_SETS_ARE_DEFINED';
ErrorTypeEnum.RETURN_NO_ORDERS_FOUND = 'string_1218';
ErrorTypeEnum.RETURN_NO_SALES_LINES_IN_ORDER = 'string_1220';
ErrorTypeEnum.RETURN_ALL_SALES_LINES_IN_ORDER_RETURN = 'string_1237';
ErrorTypeEnum.RETURN_NO_REASON_CODES_ARE_DEFINED = 'string_1244';
ErrorTypeEnum.RETURN_NO_ITEM_SELECTED = 'string_1246';
ErrorTypeEnum.RETURN_MAX_RETURN_LINE_AMOUNT_EXCEEDED = 'string_29370';
ErrorTypeEnum.RETURN_MAX_RETURN_TOTAL_AMOUNT_EXCEEDED = 'string_29371';
// Client error codes - payment
ErrorTypeEnum.CREDIT_MEMO_INVALID_AMOUNT = 'string_29800';
// Client error codes - inventory
ErrorTypeEnum.NO_PRICECHECK_ON_PRODUCTS = 'string_3523';
ErrorTypeEnum.NO_PRODUCT_INFORMATION = 'string_3873';
// Client error codes - salesOrders
ErrorTypeEnum.CART_UNAVAILABLE_FOR_PICK_UP = 'string_4539';
ErrorTypeEnum.CART_LINES_UNAVAILABLE_FOR_PICK_UP = 'string_4540';
ErrorTypeEnum.PICK_LIST_CAN_NOT_BE_CREATED = 'string_4544';
ErrorTypeEnum.PACK_SLIP_CAN_NOT_BE_CREATED = 'string_4546';
ErrorTypeEnum.CUSTOMER_ORDER_OPERATION_INVOICE_NOT_SUPPORTED = 'string_29028';
ErrorTypeEnum.CREATE_OR_EDIT_CUSTOMER_ORDER_OR_QUOTATION_ONLY = 'string_29032';
ErrorTypeEnum.EDIT_CUSTOMER_ORDER_OR_QUOTATION_ONLY = 'string_29033';
ErrorTypeEnum.NO_STORE_SELECTED_FOR_PICKUP = 'string_29034';
ErrorTypeEnum.ALL_PRODUCTS_SELECTED_PICKUP_OR_SHIP_SELECTED = 'string_29035';
ErrorTypeEnum.INVALID_SHIPPING_CHARGES = 'string_2543';
ErrorTypeEnum.NO_ADDRESSES_SELECTED_FOR_SHIP = 'string_29036';
ErrorTypeEnum.NO_SHIPPING_METHODS_SELECTED_FOR_SHIP = 'string_29037';
ErrorTypeEnum.CREATE_OR_EDIT_QUOTATION_ONLY = 'string_29042';
ErrorTypeEnum.EDIT_CUSTOMER_ORDER_ONLY = 'string_29043';
// Client error codes - customer order cancellation
ErrorTypeEnum.CANCELLATION_CHARGE_IS_NOT_VALID = 'string_4542';
ErrorTypeEnum.CANCELLATION_CHARGE_INVALID_NEGATIVE_AMOUNT = 'string_29026';
ErrorTypeEnum.CANCELLATION_CHARGE_INVALID_OPERATION = 'string_29027';
ErrorTypeEnum.CART_UNAVAILABLE_FOR_CANCEL = 'string_4541';
// Client error codes - discounts
ErrorTypeEnum.UNSUPPORTED_APPLY_DISCOUNT_OPERATION = 'string_5600';
ErrorTypeEnum.MISSING_CARTLINE_ON_APPLY_DISCOUNT = 'string_5601';
ErrorTypeEnum.MAXIMUM_LINE_DISCOUNT_AMOUNT_EXCEEDED = 'string_5602';
ErrorTypeEnum.MAXIMUM_LINE_DISCOUNT_PERCENT_EXCEEDED = 'string_5603';
ErrorTypeEnum.MAXIMUM_TOTAL_DISCOUNT_AMOUNT_EXCEEDED = 'string_5604';
ErrorTypeEnum.MAXIMUM_TOTAL_DISCOUNT_PERCENT_EXCEEDED = 'string_5605';
ErrorTypeEnum.MAXIMUM_LINE_DISCOUNT_AMOUNT_EXCEEDED_PRICE = 'string_5617';
ErrorTypeEnum.MAXIMUM_TOTAL_DISCOUNT_AMOUNT_EXCEEDED_SUBTOTAL = 'string_5618';
// Client error codes - price override
ErrorTypeEnum.PRICE_OVERRIDE_NOT_VALID_ONE_OR_MORE_ITEMS = 'string_5705';
ErrorTypeEnum.PRICE_OVERRIDE_PRICE_EXCEEDS_MAXIMUM_DEVICE_PRICE = 'string_5715';
ErrorTypeEnum.PRICE_OVERRIDE_NO_ITEM_SELECTED = 'string_5716';
ErrorTypeEnum.PRICE_OVERRIDE_PRODUCT_IS_VOIDED = 'string_29803';
ErrorTypeEnum.PRICE_OVERRIDE_PRICE_CANNOT_BE_NEGATIVE = 'string_29009';
ErrorTypeEnum.PRICE_OVERRIDE_INVALID_PRICE = 'string_29010';
ErrorTypeEnum.PRICE_OVERRIDE_PRICE_NOT_A_NUMBER = 'string_29011';
ErrorTypeEnum.PRICE_OVERRIDE_PRODUCT_IS_FOR_A_RECEIPT = 'string_29804';
ErrorTypeEnum.PRICE_OVERRIDE_PRODUCT_IS_FOR_A_GIFT_CERTIFICATE = 'string_29805';
ErrorTypeEnum.PRICE_OVERRIDE_PRICE_CANNOT_BE_ZERO = 'string_5717';
ErrorTypeEnum.PRICE_OVERRIDE_PRICE_MUST_BE_POSITIVE = 'string_29020';
ErrorTypeEnum.PRICE_OVERRIDE_ONLY_LOWER_AMOUNTS_ALLOWED = 'string_5718';
ErrorTypeEnum.PRICE_OVERRIDE_ONLY_HIGHER_AMOUNTS_ALLOWED = 'string_5719';
ErrorTypeEnum.PRICE_OVERRIDE_ONLY_LOWER_OR_EQUAL_AMOUNTS_ALLOWED = 'string_5720';
ErrorTypeEnum.PRICE_OVERRIDE_ONLY_HIGHER_OR_EQUAL_AMOUNTS_ALLOWED = 'string_5721';
ErrorTypeEnum.PRICE_OVERRIDE_NOT_ALLOWED_FOR_PRODUCT = 'string_5722';
ErrorTypeEnum.PRICE_OVERRIDE_NONE_ALLOWED = 'string_5723';
// Client error codes - change sales person
ErrorTypeEnum.CHANGE_SALES_PERSON_INVALID_CART_MODE = 'string_5733';
// Client error codes - set quantity
ErrorTypeEnum.SET_QUANTITY_NOT_VALID_ONE_OR_MORE_ITEMS = 'string_5305';
ErrorTypeEnum.SET_QUANTITY_NOT_GREATER_THAN_ZERO = 'string_5308';
ErrorTypeEnum.SET_QUANTITY_NOT_IN_RANGE = 'string_5309';
ErrorTypeEnum.SET_QUANTITY_NOT_VALID_NO_ITEM_SELECTED = 'string_5310';
ErrorTypeEnum.SET_QUANTITY_NOT_A_NUMBER = 'string_5311';
ErrorTypeEnum.SET_QUANTITY_NOT_ZERO = 'string_5312';
ErrorTypeEnum.SET_QUANTITY_QUANTITY_EXCEEDS_MAXIMUM_DEVICE_QUANTITY = 'string_5313';
ErrorTypeEnum.SET_QUANTITY_NOT_VALID_FOR_UNIT_OF_MEASURE = 'string_5314';
ErrorTypeEnum.SET_QUANTITY_NOT_VALID_FOR_SERIALIZED_ITEM = 'string_5315';
// Client error codes - unit of measure
ErrorTypeEnum.UNIT_OF_MEASURE_NOT_VALID_ONE_OR_MORE_ITEMS = 'string_3205';
ErrorTypeEnum.UNIT_OF_MEASURE_NOT_VALID_NO_ITEM_SELECTED = 'string_3206';
ErrorTypeEnum.UNIT_OF_MEASURE_NOT_VALID_NO_UNIT_OF_MEASURE = 'string_3207';
ErrorTypeEnum.UNIT_OF_MEASURE_NOT_VALID_NO_UNIT_OF_MEASURE_CONVERSIONS = 'string_3208';
ErrorTypeEnum.NOT_IMPLEMENTED = 'string_29003';
// Client error codes - kit disassembly
ErrorTypeEnum.KIT_BLOCKED_FOR_DISASSEMBLY_AT_REGISTER = 'string_420';
// Client error codes - peripherals
ErrorTypeEnum.PERIPHERALS_HARDWARESTATION_NOTCONFIGURED = 'string_4908';
ErrorTypeEnum.PERIPHERALS_BARCODE_SCANNER_NOTFOUND = 'string_4900';
ErrorTypeEnum.PERIPHERALS_BARCODE_SCANNER_ENABLE_FAILED = 'string_4901';
ErrorTypeEnum.PERIPHERALS_MSR_NOTFOUND = 'string_4902';
ErrorTypeEnum.PERIPHERALS_MSR_ENABLE_FAILED = 'string_4903';
ErrorTypeEnum.PERIPHERALS_PRINTER_FAILED = 'string_4904';
ErrorTypeEnum.PERIPHERAL_PAYMENT_UNKNOWN_ERROR = 'string_4919';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_HARDWARESTATION_PRINTER_ERROR = 'string_4904';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_HARDWARESTATION_CASHDRAWER_ERROR = 'string_4905';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_HARDWARESTATION_SCALE_ERROR = 'string_4906';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_HARDWARESTATION_PAYMENTTERMINAL_ERROR = 'string_4907';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_HARDWARESTATION_DUALDISPLAY_ERROR = 'string_4918';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_HARDWARESTATION_PERIPHERALNOTFOUND = 'string_4917';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_HARDWARESTATION_PAIRINGERROR = 'string_6011';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_HARDWARESTATION_TOKENVALIDATIONFAILED = 'string_6011';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_HARDWARESTATION_PINPAD_ERROR = 'string_4923';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_HARDWARESTATION_SIGNATURECAPTURE_ERROR = 'string_4924';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_HARDWARESTATION_LINEDISPLAY_ERROR = 'string_4925';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_HARDWARESTATION_LINEDISPLAY_CHARACTERSETNOTSUPPORTED = 'string_4926';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_HARDWARESTATION_PERIPHERALISLOCKED = 'string_4927';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_PAYMENTCONNECTORNOTFOUND = 'string_4929';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CUSTOMERACCOUNTPAYMENTISNOTALLOWEDFORCUSTOMERORDERDEPOSITANDCANCELLATION = 'string_29021';
// Client error codes - store operations
ErrorTypeEnum.AMOUNT_IS_NOT_VALID = 'string_4102';
ErrorTypeEnum.SHIFT_IS_NOT_VALID = 'string_4103';
ErrorTypeEnum.OPERATION_NOT_ALLOWED_PRODUCT_IS_VOIDED = 'string_29803';
ErrorTypeEnum.OPERATION_NOT_ALLOWED_PRODUCT_IS_FOR_A_RECEIPT = 'string_29804';
ErrorTypeEnum.OPERATION_NOT_ALLOWED_PRODUCT_IS_FOR_A_GIFT_CERTIFICATE = 'string_29805';
ErrorTypeEnum.OPERATION_NOT_ALLOWED_LINKED_PRODUCT = 'string_29806';
ErrorTypeEnum.OPERATION_NOT_ALLOWED_MULTIPLE_CART_LINES = 'string_29807';
ErrorTypeEnum.OPERATION_NOT_ALLOWED_INCOME_EXPENSE_TRANSACTION = 'string_29808';
ErrorTypeEnum.OPERATION_NOT_ALLOWED_TIME_CLOCK_DISABLED = 'string_29809';
ErrorTypeEnum.OPERATION_NOT_ALLOWED_FINISH_CURRENT_TRANSACTION = 'string_4125';
ErrorTypeEnum.RECEIPT_EMAIL_IS_EMPTY = 'string_4126';
ErrorTypeEnum.OPERATION_NOT_ALLOWED_PERMISSION_DENIED = 'string_511';
ErrorTypeEnum.OPERATION_VALIDATION_INVALID_ARGUMENTS = 'string_29018';
ErrorTypeEnum.OPERATION_NOT_ALLOWED_NO_CART_LINE_SELECTED = 'string_29822';
ErrorTypeEnum.OPERATION_NOT_ALLOWED_IN_NONDRAWER_MODE = 'string_4141';
ErrorTypeEnum.OPERATION_NOT_ALLOWED_NO_PAYMENT_LINE_SELECTED = 'string_29828';
ErrorTypeEnum.OPERATION_NOT_ALLOWED_MULTIPLE_PAYMENT_LINES = 'string_29829';
ErrorTypeEnum.OPERATION_NOT_ALLOWED_IN_OFFLINE_STATE = 'string_29831';
// Client error codes - affiliation
ErrorTypeEnum.INVALID_AFFILIATION_COLLECTION = 'string_5205';
// Client error codes - sales tax override
ErrorTypeEnum.MISSING_CARTLINE_ON_APPLY_TAX_OVERRDE = 'string_4423';
ErrorTypeEnum.NO_TAX_OVERRIDE_REASON_CODES_CONFIGURED = 'string_4422';
// Client error code - Offline
ErrorTypeEnum.CANNOT_SWITCH_ONLINE_CART_IN_PROGRESS = 'string_6607';
ErrorTypeEnum.CANNOT_SWITCH_OFFLINE_NOT_AVAILABLE = 'string_6608';
ErrorTypeEnum.CANNOT_SWITCH_TRANSFER_FAILED = 'string_6609';
// Client error codes - Signature
ErrorTypeEnum.SIGNATURE_INVALID_FORMAT = 'string_6906';
// Async Client codes
ErrorTypeEnum.ASYNC_CLIENT_ZERO_DOWNLOAD_SESSION = 'string_29375';
ErrorTypeEnum.ASYNC_CLIENT_EMPTY_UPLOAD_JOB_DEFINITION = 'string_29376';
ErrorTypeEnum.ASYNC_CLIENT_NO_TRANSACTION_DATA = 'string_29377';
ErrorTypeEnum.ASYNC_CLIENT_CANNOT_LOAD_OFFLINE_TRANSACTION_DATA = 'string_29378';
ErrorTypeEnum.ASYNC_CLIENT_FAIL_PURGE_OFFLINE_TRANSACTION_DATA = 'string_29379';
ErrorTypeEnum.ASYNC_CLIENT_FAIL_DOWNLOAD_FILE = 'string_29380';
ErrorTypeEnum.ASYNC_CLIENT_FAIL_APPLY_FILE_TO_OFFLINE_DATABASE = 'string_29381';
ErrorTypeEnum.ASYNC_CLIENT_FAIL_UPDATE_DOWNLOAD_SESSION_STATUS = 'string_29382';
ErrorTypeEnum.ASYNC_CLIENT_FAIL_RETRIEVE_INITIAL_DATA_SYNC_INDICATOR = 'string_29383';
ErrorTypeEnum.ASYNC_CLIENT_OFFLINE_NOT_ENABLED_ON_TERMINAL = 'string_29384';
// Transaction service errors comes already localized from Retail Server.
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_HEADQUARTERTRANSACTIONSERVICEMETHODCALLFAILURE = 'MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_HEADQUARTERTRANSACTIONSERVICEMETHODCALLFAILURE';
// Employee errors
ErrorTypeEnum.EMPLOYEE_NOT_FOUND = 'string_29320';
// Retail Server Error codes mapped to locale strings
ErrorTypeEnum.GENERICCOMMERCEERROR = 'string_29200';
ErrorTypeEnum.GENERICERRORMESSAGE = 'string_29201';
ErrorTypeEnum.GENERICVALIDATIONERROR = 'string_29202';
ErrorTypeEnum.ISASSOCIATEDVALIDATIONERROR = 'string_29203';
ErrorTypeEnum.ISNOTASSOCIATEDVALIDATIONERROR = 'string_29204';
ErrorTypeEnum.ISNOTAUTHENTICATEDVALIDATIONERROR = 'string_29205';
ErrorTypeEnum.ITEMSVALIDATIONERROR = 'string_29206';
ErrorTypeEnum.LINEIDSVALIDATIONERROR = 'string_29207';
ErrorTypeEnum.LISTINGSVALIDATIONERROR = 'string_29208';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_DUPLICATEOBJECT = 'string_29209';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INSUFFICIENTQUANTITYONHAND = 'string_29210';
ErrorTypeEnum.INVALID_CURRENCY_AMOUNT = 'string_29012';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDCARTSALESLINEADD = 'string_29211';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDFORMAT = 'string_29212';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_LASTCHANGEVERSIONMISMATCH = 'string_29213';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_OBJECTNOTFOUND = 'string_29214';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_REQUIREDVALUENOTFOUND = 'string_29215';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_UNKNOWNREQUEST = 'string_29216';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_UNSUPPORTEDLANGUAGE = 'string_29217';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_VALUEOUTOFRANGE = 'string_29218';
ErrorTypeEnum.NAMEVALIDATIONERROR = 'string_29219';
ErrorTypeEnum.PAYMENTSVALIDATIONERROR = 'string_29220';
ErrorTypeEnum.PROMOTIONCODEVALIDATIONERROR = 'string_29221';
ErrorTypeEnum.SAVEDSHOPPINGCARTIDVALIDATIONERROR = 'string_29222';
ErrorTypeEnum.SHIPPINGOPTIONSLINEITEMSELECTION = 'string_29223';
ErrorTypeEnum.SHIPPINGOPTIONSPICKUP = 'string_29224';
ErrorTypeEnum.SHIPPINGOPTIONSSHIPTONEWADDRESS = 'string_29225';
ErrorTypeEnum.SHIPPINGOPTIONSVALIDATIONERROR = 'string_29226';
ErrorTypeEnum.SHOPPINGCARTIDSVALIDATIONERROR = 'string_29227';
ErrorTypeEnum.SHOPPINGCARTIDVALIDATIONERROR = 'string_29228';
ErrorTypeEnum.STORELOCATORBINGMAPSTOKENEMPTY = 'string_29229';
ErrorTypeEnum.STORELOCATORINVALIDDISTANCE = 'string_29230';
ErrorTypeEnum.STORELOCATORINVALIDLOCATION = 'string_29231';
ErrorTypeEnum.STORELOCATORUNABLETOGETCOORDINATES = 'string_29232';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDCARTVERSION = 'string_29233';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_AGGREGATECOMMUNICATIONERROR = 'string_29234';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_APPLICATIONCOMPOSITIONFAILED = 'string_29235';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CANNOTRETURNMORETHANPURCHASED = 'string_5303';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CONFIGURATIONSETTINGNOTFOUND = 'string_29236';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_DUPLICATEDEFAULTNOTIFICATIONHANDLERENCOUNTERED = 'string_29237';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_EMPTYINVENTORYUNITOFMEASUREFORITEM = 'string_29238';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_EXTERNALPROVIDERERROR = 'string_29239';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_HEADQUARTERCOMMUNICATIONFAILURE = 'string_29240';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_HEADQUARTERRESPONSEPARSINGERROR = 'string_29241';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDCARTSTATE = 'string_29008';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDCHANNELCONFIGURATION = 'string_29242';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDCONFIGURATIONKEYFORMAT = 'string_29243';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDCONNECTIONSTRING = 'string_29244';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDPIPELINECONFIGURATION = 'string_29245';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDPROVIDERCONFIGURATION = 'string_29246';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDRUNTIMECONTEXT = 'string_29247';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDSERVERRESPONSE = 'string_29248';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_SERVICEINITIALIZATIONFAILED = 'string_29249';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_SERVICENOTFOUND = 'string_29250';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_UNABLETOCOMPUTESALESTAXGROUPFORADDRESS = 'string_29251';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_UNABLETOFINDDEFAULTHANDLER = 'string_29252';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_UNABLETOFINDDELIVERYOPTIONS = 'string_29253';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_UNABLETOFINDINVENTORYFORITEM = 'string_29254';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_AGGREGATEVALIDATIONERROR = 'string_29255';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_AUTHENTICATIONFAILED = 'string_29256';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_AUTHORIZATIONFAILED = 'string_29257';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_COUPONISVALIDFORCURRENTSESSION = 'string_29258';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CRITICALSTORAGEERROR = 'string_29259';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_DISCOUNTAMOUNTINVALIDATED = 'string_29260';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_DISCOUNTISALLOWEDONLYFORCREATIONANDEDITION = 'string_5613';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_IDMISMATCH = 'string_29261';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INSUFFICIENTQUANTITYAVAILABLE = 'string_29262';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDCHANGETRACKINGCONFIGURATION = 'string_29263';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDPRICEENCOUNTERED = 'string_29264';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDREQUEST = 'string_29265';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDSHIPPINGADDRESS = 'string_29266';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDSQLCOMMAND = 'string_29267';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_ITEMDISCONTINUEDFROMCHANNEL = 'string_29268';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_OBJECTVERSIONMISMATCHERROR = 'string_29269';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_PROVIDERCOMMUNICATIONFAILURE = 'string_29270';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_REQUESTEDITEMISOUTOFSTOCK = 'string_29271';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_UNITOFMEASURECONVERSIONNOTFOUND = 'string_29272';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_DEFAULTCUSTOMERNOTFOUND = 'string_29273';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_BARCODENOTFOUND = 'string_29368';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_LINKEDITEMSEARCHBYBARCODENOTSUPPORTED = 'MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_LINKEDITEMSEARCHBYBARCODENOTSUPPORTED';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_QUOTEMUSTNOTHAVEDEPOSITOVERRIDE = 'string_29282';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_QUOTEMUSTHAVEVALIDQUOTATIONEXPIRYDATE = 'string_4321';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDPICKUPDEPOSITOVERRIDEAMOUNT = 'string_29283';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_SALESLINEMUSTHAVEPICKUPDELIVERYMODE = 'string_29820';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_UNABLETOPICKUPMORETHANQTYREMAINING = 'string_29821';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_ORDERWASNOTCREATEDWITHDEPOSITOVERRIDE = 'string_29299';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_MULTIPLEEMPLOYEETOTALDISCOUNTSNOTALLOWED = 'string_29294';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_MULTIPLEEMPLOYEELINEDISCOUNTSNOTALLOWED = 'string_29295';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CANNOTSUSPENDCARTWITHACTIVETENDERLINES = 'string_29304';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_TERMINALHASANOPENSHIFT = 'string_29338';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CASHDRAWERHASANOPENSHIFT = 'string_29306';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_SHIFTVALIDATIONERROR = 'string_29307';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_SHIFTALREADYOPENONDIFFERENTTERMINAL = 'string_29334';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_SHIFTSTARTINGAMOUNTNOTENTERED = 'string_29308';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_SHIFTTENDERDECLARATIONAMOUNTNOTENTERED = 'string_29309';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CANNOTCHANGECUSTOMERIDWHENEDITINGCUSTOMERORDER = 'string_4420';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDSHIPPINGDATE = 'string_29810';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CURRENCYCHANNELORDERMISMATCH = 'string_29319';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CURRENCYNOTFOUND = 'string_29374';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_DEPOSITMUSTBEGREATERTHANZERO = 'string_29325';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_DEPOSITOVERRIDEMUSTNOTBEGREATERTHANTOTALAMOUNT = 'string_29324';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_DEPOSITOVERRIDEMAYNOTBECHANGED = 'string_29326';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_DEPOSITOVERRIDEMAYNOTBECLEARED = 'string_29327';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_PAYMENTUSINGUNAUTHORIZEDACCOUNT = 'string_29351';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_EMPLOYEEDISCOUNTEXCEEDED = 'string_29346';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_MAXCOUNTINGDIFFERENCEEXCEEDED = 'string_29367';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CHANGETENDERTYPENOTSUPPORTED = 'string_29369';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_ELEVATEDUSERSAMEASLOGGEDONUSER = 'string_29256';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_UNABLETOGENERATETOKEN = 'string_1175';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_PAYMENTKEYNOTFOUND = 'string_1195';
ErrorTypeEnum.BAD_REQUEST = 'string_29274';
ErrorTypeEnum.NOT_AUTHORIZIED = 'string_29275';
ErrorTypeEnum.FORBIDDEN = 'string_29276';
ErrorTypeEnum.PRECONDITION_FAILED = 'string_29277';
ErrorTypeEnum.SERVICE_UNAVAILABLE = 'string_29278';
ErrorTypeEnum.SERVER_TIMEOUT = 'string_29279';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDCARDEXPIRATIONDATE = 'string_1180';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_UNABLETOAUTHORIZEPAYMENT = 'string_29280';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_VOIDTRANSACTIONCONTAINSTENDEREDLINES = 'string_29281';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_LOYALTYCARDALREADYISSUED = 'string_29284';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CUSTOMERNOTFOUND = 'string_29285';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDLOYALTYCARDNUMBER = 'string_29286';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_BLOCKEDLOYALTYCARD = 'string_29287';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_NOTENDERLOYALTYCARD = 'string_29322';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CONFLICTLOYALTYCARDCUSTOMERANDTRANSACTIONCUSTOMER = 'string_29288';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDDEVICE = 'string_507';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CDXREALTIMESERVICEFAILURE = 'string_29289';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_NOTENOUGHREWARDPOINTS = 'string_29290';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_NOMORETHANONELOYALTYTENDER = 'string_29291';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_AMOUNTDUEMUSTBEPAIDBEFORECHECKOUT = 'string_29292';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDPAYMENTREQUEST = 'string_29293';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_PAYMENTALREADYVOIDED = 'string_29293';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDADDRESS = 'string_29296';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_NOPRICEOVERRIDEFORRETURNS = 'string_29297';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_NOPRICEOVERRIDEFORGIFTCARDS = 'string_29298';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_PAYMENTAMOUNTEXCEEDSGIFTBALANCE = 'string_29301';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_NOMORETHANONEOPERATIONWITHAGIFTCARD = 'string_29302';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CANNOTADDNONPRODUCTITEMTOCUSTOMERORDER = 'string_29811';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_USEEXISTINGSHIFTPERMISSIONDENIED = 'string_29303';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_SHIFTNOTFOUND = 'string_29336';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CUSTOMERACCOUNTDEPOSITMULTIPLECARTLINESNOTALLOWED = 'string_29340';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CUSTOMERACCOUNTNUMBERISNOTSET = 'string_29341';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CANNOTPAYFORCUSTOMERACCOUNTDEPOSITWITHCUSTOMERACCOUNTPAYMENTMETHOD = 'string_29342';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CUSTOMERACCOUNTDEPOSITCANNOTBEVOIDED = 'string_29344';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CUSTOMERACCOUNTDEPOSITCARTTYPEMISMATCH = 'string_29345';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_NONDRAWEROPERATIONSONLY = 'string_2123';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CANNOTSUSPENDCARTWITHACTIVEGIFTCARDSALESLINES = 'string_29305';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_RETURNITEMPRICEEXCEEDED = 'string_29310';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_RETURNTRANSACTIONTOTALEXCEEDED = 'string_29311';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_PROPERTYUPDATENOTALLOWED = 'string_29312';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDQUANTITY = 'string_29313';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDCUSTOMERORDERMODEFORADDCARTLINE = 'string_29314';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDCUSTOMERORDERMODEFORVOIDPRODUCTS = 'string_29335';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_REMOTEPRODUCTSNOTSUPPORTEDWITHCURRENTTRANSACTIONTYPE = 'string_29315';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_REFUNDAMOUNTMORETHANALLOWED = 'string_29316';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_VIEWTIMECLOCKNOTENABLED = 'string_29317';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_ITEMQUANTITYEXCEEDED = 'string_29318';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_TIMECLOCKNOTENABLED = 'string_29321';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDUNITOFMEASURE = 'string_29323';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_SETTLEINVOICEFAILED = 'string_29330';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_NOPRICEOVERRIDEFORINVOICELINES = 'string_29331';
ErrorTypeEnum.MICROSOFT_DYNAMICS_SERVER_INTERNAL_ERROR = ErrorTypeEnum.GENERICERRORMESSAGE;
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CARTNOTFOUND = 'string_29314';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CARTNOTACTIVE = 'string_29834';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INCOMEEXPENSECARTDOESNOTALLOWSALESLINE = 'string_29337';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INCOMEEXPENSECARTDOESNOTALLOWCUSTOMER = 'string_29348';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_GIFTCARDUNLOCKFAILED = 'string_29339';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_UNLOCKREGISTERFAILED = 'string_29352';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_OFFLINEDATABASECHUNKFILENOTFOUND = 'string_29830';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CHANGEPASSWORDFAILED = 'string_6613';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_RESETPASSWORDFAILED = 'string_6614';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CHANNELDATABASECONNECTIONFAILED = 'string_1429';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_REALTIMESERVICECONNECTIONFAILED = 'string_1429';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_SALESMUSTHAVEQUANTITYGREATERTHANZERO = 'string_29353';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_UNABLETOCAPTUREPAYMENT = 'string_29354';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_SESSIONEXPIRED = 'string_29256';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDAMOUNT = 'string_29372';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INVALIDBAGNUMBER = 'string_29373';
// Retail Server Payment Error codes mapped to locale strings
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_BALANCEAMOUNTEXCEEDSMAXIMUMALLOWEDVALUE = 'string_29355';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CHANGEBACKISNOTALLOWED = 'string_29356';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INCORRECTPAYMENTAMOUNTSIGN = 'string_29357';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_OVERTENDERAMOUNTEXCEEDSMAXIMUMALLOWEDVALUE = 'string_29358';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_PAYMENTEXCEEDSMAXIMUMAMOUNTPERLINE = 'string_29359';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_PAYMENTEXCEEDSMAXIMUMAMOUNTPERTRANSACTION = 'string_29360';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_PAYMENTEXCEEDSMINIMUMAMOUNTPERLINE = 'string_29361';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_PAYMENTEXCEEDSMINIMUMAMOUNTPERTRANSACTION = 'string_29362';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_PAYMENTMUSTBEUSEDTOFINALIZETRANSACTION = 'string_29363';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_PRODUCTISNOTACTIVE = 'string_29364';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_PRODUCTISBLOCKED = 'string_29365';
ErrorTypeEnum.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CANNOTRETURNMULTIPLETRANSACTIONS = 'string_29366';
/**
 * Forward links for device activation errors.
 */
export class DeviceActivationErrorsForwardLinks {
}
DeviceActivationErrorsForwardLinks.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_CHANNELDATABASECONNECTIONFAILED = 'http://go.microsoft.com/fwlink/?LinkId=403590';
DeviceActivationErrorsForwardLinks.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_REALTIMESERVICECONNECTIONFAILED = 'http://go.microsoft.com/fwlink/?LinkId=403591';
DeviceActivationErrorsForwardLinks.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_TERMINALNOTASSIGNEDTOSTORE = 'http://go.microsoft.com/fwlink/?LinkId=403592';
DeviceActivationErrorsForwardLinks.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_USERNOTASSIGNEDTOSTORE = 'http://go.microsoft.com/fwlink/?LinkId=403593';
DeviceActivationErrorsForwardLinks.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_INCORRECTSTAFFIDORPASSWORD = 'http://go.microsoft.com/fwlink/?LinkId=403594';
DeviceActivationErrorsForwardLinks.MICROSOFT_DYNAMICS_COMMERCE_RUNTIME_DEVICEALREADYACTIVATED = 'http://go.microsoft.com/fwlink/?LinkId=519136';
DeviceActivationErrorsForwardLinks.SERVER_ERROR = 'http://go.microsoft.com/fwlink/?LinkId=519137';
export class ErrorHelper {
    static isAggregatedErrorResourceId(errorResourceId) {
        return ErrorHelper.AGGREGATED_ERROR_RESOUCEIDS.indexOf(errorResourceId) !== -1;
    }
    /**
     * Map response status code to error.
     *
     * @param {any} Error returned by retail server call.
     */
    static MapResponseStatusCodeToError(errorMessage, statusCode, err) {
        let errorCode = '';
        let canRetry = false;
        switch (statusCode) {
            case 400:
                errorCode = ErrorTypeEnum.BAD_REQUEST;
                break;
            case 401:
                errorCode = ErrorTypeEnum.NOT_AUTHORIZIED;
                break;
            case 403:
                errorCode = ErrorTypeEnum.FORBIDDEN;
                break;
            case 412:
                errorCode = ErrorTypeEnum.PRECONDITION_FAILED;
                break;
            case 503:
                errorCode = ErrorTypeEnum.SERVICE_UNAVAILABLE;
                canRetry = true;
                break;
            case 500:
            default:
                errorCode = ErrorTypeEnum.SERVER_ERROR;
                errorMessage = 'The server error format is not supported or it does not have enough information.';
                // Temporary until the fix for Bug 681427 goes in
                // If no orders were found, than an error did not occur
                if (err && err.response && err.response.body && err.response.requestUri) {
                    if (err.response.requestUri.indexOf('GetOrderByReceiptId') > -1) {
                        if (err.response.body.indexOf('No orders were found') > -1) {
                            errorMessage = ErrorTypeEnum.RETURN_NO_ORDERS_FOUND;
                        }
                    }
                }
        }
        if (errorMessage && errorMessage.toUpperCase() === 'TIMEOUT') {
            errorCode = ErrorTypeEnum.SERVER_TIMEOUT;
        }
        return new ProxyError(errorCode, errorMessage || EMPTY, EMPTY, canRetry, err);
    }
    /**
     * Gets whether an error code is present in the error collection or not.
     * @param {string} errorType The error resource identifier.
     * @return {boolean} Whether an error code is present in the error collection or not.
     */
    static hasError(errors, errorType) {
        if (hasElements(errors)) {
            for (const error of errors) {
                const errorTypeValue = error.ErrorCode != null ? ErrorTypeEnum[error.ErrorCode.toUpperCase()] : null;
                // compares both error resource id as well as label resource id
                if (!compare(errorType, error.ErrorCode, true) || !compare(errorType, errorTypeValue, true)) {
                    // error found
                    return true;
                }
            }
        }
        return false;
    }
    /**
     * Checks whether all the provided errors are retryable. If no error code is provided or an error is undefined,
     * then the errors will be treated as not retryable.
     *
     * @return {boolean} True if the error codes are retryable, false if the error codes are not retryable or not defined
     */
    static isRetryable(errors) {
        const numErrors = !errors ? 0 : errors.length;
        let isRetryable = numErrors > 0;
        for (let i = 0; i < numErrors; i++) {
            if (errors[i]) {
                isRetryable = isRetryable && errors[i].CanRetry;
            }
            else {
                isRetryable = false;
            }
        }
        return isRetryable;
    }
    /**
     * Gets a string with all error codes.
     * @param {ProxyError[]} errors the error collection.
     * @return {string} a formated string containing all error codes.
     */
    static getErrorResourceIds(errors) {
        let result = '';
        if (hasElements(errors)) {
            const errorResourceIds = [];
            for (const error of errors) {
                errorResourceIds.push(error.ErrorCode);
            }
            result = errorResourceIds.join(', ');
        }
        return result;
    }
    /**
     * Gets a string with all error messages.
     * @param {ProxyError[]} errors the error collection.
     * @return {string} a formated string containing all error messages.
     */
    static getErrorMessages(errors) {
        let result = '';
        if (hasElements(errors)) {
            const errorResourceIds = [];
            for (const error of errors) {
                errorResourceIds.push(error.LocalizedErrorMessage);
            }
            result = errorResourceIds.join(', ');
        }
        return result;
    }
}
ErrorHelper.AGGREGATED_ERROR_RESOUCEIDS = [
    'Microsoft_Dynamics_Commerce_Runtime_AggregateValidationError',
    'Microsoft_Dynamics_Commerce_Runtime_AggregateCommunicationError',
    'Microsoft_Dynamics_Commerce_Runtime_InvalidCartLinesAggregateError'
];
ErrorHelper.MICROSOFT_DYNAMICS_SERVER_INTERNAL_ERROR = 'Microsoft_Dynamics_Internal_Server_Error';
