// @msdyn365-commerce-modules/order-management/src/modules/order-history/tests/__mock__.ts
import { OrderShipments, SalesLine, SimpleProduct } from '@msdyn365-commerce/retail-proxy';

export const mockSalesLine: SalesLine = {
    ProductId: 22565430654,
    TotalAmount: 300,
    SalesStatusValue: 0,
    Quantity: 1
};

export const mockOrderShipment: OrderShipments = {
    SalesId: undefined,
    ChannelReferenceId: '0641E90GLDEN',
    CurrencyCode: 'USD',
    SalesLines: [mockSalesLine],
    TaxAmount: 19.5,
    TotalAmount: 319.5,
    CustomerId: '004258'
};

export const mockProducts: SimpleProduct[] = [
    {
        RecordId: 22565430654,
        ItemId: '81307',
        Name: 'Red Leather Bag',
        Description:
            'Whether you need a carry-everything or evening clutch, our selection covers all of your needs with a variety of handbags to match any occasion.',
        ProductTypeValue: 4,
        BasePrice: 300,
        Price: 300,
        AdjustedPrice: 300,
        PrimaryImageUrl: 'Products/81307_000_001.png'
    }
];

export const mockOrderShipments: OrderShipments[] = [
    mockOrderShipment
];

export const mockRequestContextJson = <unknown>{
    url: {
        onerfPageUrl: 'https://localhost:3000'
    },
    locale: 'en-us',
    apiSettings: {
        baseUrl: '',
        baseImageUrl: '',
        channelId: 0,
        oun: '',
        ratingsReviewsEndpoint: ''
    },
    user: {
        token: 'Dummy token',
        isAuthenticated: true,
        customerAccountNumber:'004258'
    },
    urlTokens: {
        locale: 'en-us',
        categories: ['fashion-sunglasses']
    }
};
