// @msdyn365-commerce-modules/checkout-utilities/dist/lib/checkout-state.js
import { __decorate, __metadata, __spreadArrays } from "tslib";
import { action, computed, observable } from 'mobx';
/**
 * State information related to what is needed to succesfully perform a checkout
 */
var CheckoutState = /** @class */ (function () {
    function CheckoutState() {
        this._giftCards = [];
        this._giftCardExtends = [];
        this._loyaltyAmount = 0;
        this._guestCheckoutEmail = '';
        this._isTermsAndConditionAccepted = false;
        this._customerAccountAmount = 0;
    }
    Object.defineProperty(CheckoutState.prototype, "tokenizedPaymentCard", {
        get: function () {
            return this._tokenizedPaymentCard;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CheckoutState.prototype, "tenderLine", {
        get: function () {
            return this._tenderLine;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CheckoutState.prototype, "billingAddress", {
        get: function () {
            return this._billingAddress;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CheckoutState.prototype, "shippingAddress", {
        get: function () {
            return this._shippingAddress;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CheckoutState.prototype, "cardPrefix", {
        get: function () {
            return this._cardPrefix;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CheckoutState.prototype, "giftCards", {
        get: function () {
            return this._giftCards;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CheckoutState.prototype, "giftCardExtends", {
        get: function () {
            return this._giftCardExtends;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CheckoutState.prototype, "loyaltyCardNumber", {
        get: function () {
            return this._loyaltyCardNumber;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CheckoutState.prototype, "loyaltyAmount", {
        get: function () {
            return this._loyaltyAmount;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CheckoutState.prototype, "guestCheckoutEmail", {
        get: function () {
            return this._guestCheckoutEmail;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CheckoutState.prototype, "isTermsAndConditionAccepted", {
        get: function () {
            return this._isTermsAndConditionAccepted;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CheckoutState.prototype, "customerAccountAmount", {
        get: function () {
            return this._customerAccountAmount;
        },
        enumerable: true,
        configurable: true
    });
    CheckoutState.prototype.updateTokenizedPaymentCard = function (newTokenizedPaymentCard) {
        this._tokenizedPaymentCard = newTokenizedPaymentCard;
    };
    CheckoutState.prototype.updateTenderLine = function (newTenderLine) {
        this._tenderLine = newTenderLine;
    };
    CheckoutState.prototype.updateShippingAddress = function (newShippingAddress) {
        this._shippingAddress = newShippingAddress;
    };
    CheckoutState.prototype.updateBillingAddress = function (newBillingAddress) {
        this._billingAddress = newBillingAddress;
    };
    CheckoutState.prototype.updateCardPrefix = function (newCardPrefix) {
        this._cardPrefix = newCardPrefix;
    };
    CheckoutState.prototype.removeGiftCard = function (giftCardNumber) {
        this._giftCards = this._giftCards.filter(function (giftCard) { return giftCard.Id !== giftCardNumber; });
    };
    CheckoutState.prototype.removeGiftCardExtend = function (giftCardNumber) {
        this._giftCardExtends = this._giftCardExtends.filter(function (giftCard) { return giftCard.Id !== giftCardNumber; });
    };
    CheckoutState.prototype.addGiftCard = function (giftCard) {
        this._giftCards = __spreadArrays([giftCard], this._giftCards);
    };
    CheckoutState.prototype.addGiftCardExtend = function (giftCard) {
        this._giftCardExtends = __spreadArrays([giftCard], this._giftCards);
    };
    CheckoutState.prototype.updateLoyaltyCardNumber = function (newLoyaltyCardNumber) {
        this._loyaltyCardNumber = newLoyaltyCardNumber;
    };
    CheckoutState.prototype.updateLoyaltyAmount = function (newAmount) {
        this._loyaltyAmount = newAmount;
    };
    CheckoutState.prototype.updateGuestCheckoutEmail = function (newGuestCheckoutEmail) {
        this._guestCheckoutEmail = newGuestCheckoutEmail;
    };
    CheckoutState.prototype.updateTermsAndConditionsAcceptance = function (newIsTermsAndConditionAccepted) {
        this._isTermsAndConditionAccepted = newIsTermsAndConditionAccepted;
    };
    CheckoutState.prototype.updateCustomerAccountAmount = function (newAmount) {
        this._customerAccountAmount = newAmount;
    };
    __decorate([
        observable,
        __metadata("design:type", Object)
    ], CheckoutState.prototype, "_tokenizedPaymentCard", void 0);
    __decorate([
        computed,
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [])
    ], CheckoutState.prototype, "tokenizedPaymentCard", null);
    __decorate([
        observable,
        __metadata("design:type", Object)
    ], CheckoutState.prototype, "_tenderLine", void 0);
    __decorate([
        computed,
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [])
    ], CheckoutState.prototype, "tenderLine", null);
    __decorate([
        observable,
        __metadata("design:type", Object)
    ], CheckoutState.prototype, "_billingAddress", void 0);
    __decorate([
        computed,
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [])
    ], CheckoutState.prototype, "billingAddress", null);
    __decorate([
        observable,
        __metadata("design:type", Object)
    ], CheckoutState.prototype, "_shippingAddress", void 0);
    __decorate([
        computed,
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [])
    ], CheckoutState.prototype, "shippingAddress", null);
    __decorate([
        observable,
        __metadata("design:type", String)
    ], CheckoutState.prototype, "_cardPrefix", void 0);
    __decorate([
        computed,
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [])
    ], CheckoutState.prototype, "cardPrefix", null);
    __decorate([
        observable,
        __metadata("design:type", Array)
    ], CheckoutState.prototype, "_giftCards", void 0);
    __decorate([
        computed,
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [])
    ], CheckoutState.prototype, "giftCards", null);
    __decorate([
        observable,
        __metadata("design:type", Array)
    ], CheckoutState.prototype, "_giftCardExtends", void 0);
    __decorate([
        computed,
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [])
    ], CheckoutState.prototype, "giftCardExtends", null);
    __decorate([
        observable,
        __metadata("design:type", String)
    ], CheckoutState.prototype, "_loyaltyCardNumber", void 0);
    __decorate([
        computed,
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [])
    ], CheckoutState.prototype, "loyaltyCardNumber", null);
    __decorate([
        observable,
        __metadata("design:type", Number)
    ], CheckoutState.prototype, "_loyaltyAmount", void 0);
    __decorate([
        computed,
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [])
    ], CheckoutState.prototype, "loyaltyAmount", null);
    __decorate([
        observable,
        __metadata("design:type", String)
    ], CheckoutState.prototype, "_guestCheckoutEmail", void 0);
    __decorate([
        computed,
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [])
    ], CheckoutState.prototype, "guestCheckoutEmail", null);
    __decorate([
        observable,
        __metadata("design:type", Boolean)
    ], CheckoutState.prototype, "_isTermsAndConditionAccepted", void 0);
    __decorate([
        computed,
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [])
    ], CheckoutState.prototype, "isTermsAndConditionAccepted", null);
    __decorate([
        observable,
        __metadata("design:type", Number)
    ], CheckoutState.prototype, "_customerAccountAmount", void 0);
    __decorate([
        computed,
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [])
    ], CheckoutState.prototype, "customerAccountAmount", null);
    __decorate([
        action,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], CheckoutState.prototype, "updateTokenizedPaymentCard", null);
    __decorate([
        action,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], CheckoutState.prototype, "updateTenderLine", null);
    __decorate([
        action,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], CheckoutState.prototype, "updateShippingAddress", null);
    __decorate([
        action,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], CheckoutState.prototype, "updateBillingAddress", null);
    __decorate([
        action,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", void 0)
    ], CheckoutState.prototype, "updateCardPrefix", null);
    __decorate([
        action,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", void 0)
    ], CheckoutState.prototype, "removeGiftCard", null);
    __decorate([
        action,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", void 0)
    ], CheckoutState.prototype, "removeGiftCardExtend", null);
    __decorate([
        action,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], CheckoutState.prototype, "addGiftCard", null);
    __decorate([
        action,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], CheckoutState.prototype, "addGiftCardExtend", null);
    __decorate([
        action,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", void 0)
    ], CheckoutState.prototype, "updateLoyaltyCardNumber", null);
    __decorate([
        action,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Number]),
        __metadata("design:returntype", void 0)
    ], CheckoutState.prototype, "updateLoyaltyAmount", null);
    __decorate([
        action,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", void 0)
    ], CheckoutState.prototype, "updateGuestCheckoutEmail", null);
    __decorate([
        action,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Boolean]),
        __metadata("design:returntype", void 0)
    ], CheckoutState.prototype, "updateTermsAndConditionsAcceptance", null);
    __decorate([
        action,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Number]),
        __metadata("design:returntype", void 0)
    ], CheckoutState.prototype, "updateCustomerAccountAmount", null);
    return CheckoutState;
}());
export { CheckoutState };
