// @msdyn365-commerce-modules/checkout-utilities/dist/lib/checkout-state-data-action.js
import { __awaiter, __generator } from "tslib";
import { buildCacheKey } from '@msdyn365-commerce-modules/retail-actions';
import { createObservableDataAction } from '@msdyn365-commerce/core';
import { CheckoutState } from './checkout-state';
/**
 * CheckoutState - action input
 */
var CheckoutStateInput = /** @class */ (function () {
    function CheckoutStateInput(apiSettings) {
        var _this = this;
        this.getCacheKey = function () { return buildCacheKey("CheckoutState", _this.apiSettings); };
        this.getCacheObjectType = function () { return 'CheckoutState'; };
        this.dataCacheType = function () { return 'request'; };
        this.apiSettings = apiSettings;
    }
    return CheckoutStateInput;
}());
export { CheckoutStateInput };
/**
 * CheckoutState - action
 */
export function checkoutStateAction(input, ctx) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, new CheckoutState()];
        });
    });
}
/**
 * CheckoutState - create new input for create action
 */
var createInput = function (inputData) {
    return new CheckoutStateInput(inputData.requestContext.apiSettings);
};
/**
 * CheckoutState - create action
 */
export default createObservableDataAction({
    action: checkoutStateAction,
    input: createInput
});
