// @msdyn365-commerce-modules/order-template/dist/lib/modules/order-template/components/add-line/dropdown.js
import * as React from 'react';
export class Dropdown extends React.PureComponent {
    constructor(props, state) {
        super(props);
        this.selectMenu = React.createRef();
        this._onChanged = (event) => {
            this.setState({ selectedIndex: event.target.value });
            if (this.props.onChange) {
                this.props.onChange({
                    dropdownId: this.props.dropdownId,
                    selectId: event.target.value,
                    selectedValue: event.target.innerText
                });
            }
        };
        this.state = {
            selectedIndex: ''
        };
    }
    componentDidMount() {
        if (this.selectMenu.current) {
            this.selectMenu.current.selectedIndex = 0;
        }
    }
    render() {
        const { dropdownId, dropdownList, dropdownName } = this.props;
        return (React.createElement("select", { id: `msc-dropdown__entry-${dropdownId}`, ref: this.selectMenu, "aria-label": dropdownName, className: 'msc-dropdown__select', onChange: this._onChanged }, dropdownList.map((item) => (item &&
            React.createElement("option", { value: item.id, key: item.id, selected: +this.state.selectedIndex === +item.id, "aria-selected": +this.state.selectedIndex === +item.id }, item.value)))));
    }
}
