// @msdyn365-commerce-modules/header/dist/lib/modules/header/components/logo.js
import { getPayloadObject, getTelemetryAttributes, TelemetryConstant } from '@msdyn365-commerce-modules/utilities';
import * as Msdyn365 from '@msdyn365-commerce/core';
import * as React from 'react';
export class Logo extends React.PureComponent {
    constructor(props) {
        super(props);
        const payLoad = getPayloadObject('click', props.telemetryContent, TelemetryConstant.Logo);
        this.attributes = getTelemetryAttributes(props.telemetryContent, payLoad);
    }
    render() {
        return (React.createElement("div", { className: this.props.className }, this._renderLogo(this.props)));
    }
    _renderLogo(config) {
        if (config.link && config.link.linkUrl.destinationUrl) {
            return (React.createElement("a", Object.assign({ href: config.link.linkUrl.destinationUrl, "aria-label": config.link.ariaLabel, target: config.link.openInNewTab ? '_blank' : undefined }, this.attributes), this._renderImage(config)));
        }
        return (this._renderImage(config));
    }
    _renderImage(config) {
        const defaultImageSettings = {
            viewports: {
                xs: { q: `w=132&h=28&m=6`, w: 0, h: 0 },
                lg: { q: `w=160&h=48&m=6`, w: 0, h: 0 }
            },
            lazyload: true
        };
        return (React.createElement(Msdyn365.Image, Object.assign({}, config.image, { gridSettings: this.props.gridSettings, imageSettings: config && config.image && config.image.imageSettings || defaultImageSettings, loadFailureBehavior: 'default', editProps: { key: config.image || {}, requestContext: this.props.requestContext } })));
    }
}
export default Logo;
