// @msdyn365-commerce/global-state/dist/lib/data-actions/global-state-data-action.js
import { createObservableDataAction } from '@msdyn365-commerce/core';
import { createGlobalState } from '../global-state/i-global-state';
/**
 * Shared class that handles global state inputs
 */
export class GlobalStateInput {
    constructor(key, stateConstructor, apiSettings) {
        this.getCacheKey = () => `${this.key}-chanId:${this.apiSettings.channelId}-catId:${this.apiSettings.catalogId}`;
        this.getCacheObjectType = () => this.key;
        this.dataCacheType = () => 'instance';
        this.apiSettings = apiSettings;
        this.stateConstructor = stateConstructor;
        this.key = key;
    }
}
export async function getState(input, ctx) {
    const globalStateObject = createGlobalState(input.stateConstructor, ctx);
    await globalStateObject.initialize();
    return globalStateObject;
}
export default function createGlobalStateDataAction(key, stateConstructor) {
    return createObservableDataAction({
        action: getState,
        input: (args) => {
            return new GlobalStateInput(key, stateConstructor, args.requestContext.apiSettings);
        }
    });
}
