// @msdyn365-commerce/cli-internal/jest.config.js
/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */
const path = require('path');

module.exports = {
    roots: [path.join(process.cwd(), '__tests__'), path.join(process.cwd(), 'scripts')],
    collectCoverage: true,
    collectCoverageFrom: ['scripts/**/*.{js,jsx,ts,tsx}', '!scripts/**/*.test.{js,jsx,ts,tsx}'],
    coverageReporters: ['cobertura', 'html'],
    moduleNameMapper: {
        '\\.(css|less|scss|sass)$': 'identity-obj-proxy'
    },
    moduleFileExtensions: ['ts', 'tsx', 'js', 'json'],
    reporters: ['default', 'jest-junit'],
    testMatch: ['<rootDir>/__tests__/**/*.(ts|js)?(x)', '<rootDir>/?(*.)(spec|test).(ts|js)?(x)'],
    testPathIgnorePatterns: ['<rootDir>/dist/', '<rootDir>/__tests__/__mocks__', '<rootDir>/__tests__/__constants__'],
    transform: {
        '^.+\\.(js|jsx|ts|tsx)$': 'ts-jest'
    },
    transformIgnorePatterns: ['/node_modules/(?!@msdyn365-commerce/core-internal)/'],
    globals: {
        'ts-jest': {
            tsConfig: {
                allowJs: true
            }
        }
    }
};
