// @msdyn365-commerce-modules/order-management/dist/lib/modules/order-history/order-history.view.js
import { Module, Node } from '@msdyn365-commerce-modules/utilities';
import * as React from 'react';
export const OrderHistoryOrderInfomation = ({ orderInformationProps, salesId, receiptId, channelName, createdDate, count, amount, channelReferenceId }) => (React.createElement(Node, Object.assign({}, orderInformationProps),
    channelName,
    salesId,
    receiptId,
    createdDate,
    count,
    amount,
    channelReferenceId));
export const OrderHistoryGroupDelivery = ({ deliveryProps, heading, count, processing, address, trackingInfo }) => (React.createElement(Node, Object.assign({}, deliveryProps),
    heading,
    count,
    processing,
    address,
    trackingInfo));
export const OrderHistoryGroup = ({ groupProps, delivery, salesLinesProps, salesLines }) => (React.createElement(Node, Object.assign({}, groupProps),
    delivery && React.createElement(OrderHistoryGroupDelivery, Object.assign({}, delivery)),
    salesLines && (React.createElement(Node, Object.assign({}, salesLinesProps), salesLines.map(salesLine => (React.createElement(React.Fragment, { key: salesLine.data.salesLine.LineId }, salesLine.salesLine)))))));
export const OrderHistoryGroups = ({ groupsProps, groups }) => (React.createElement(Node, Object.assign({}, groupsProps), groups.map((group, index) => (React.createElement(OrderHistoryGroup, Object.assign({ key: index }, group))))));
export const OrderHistoryHeader = ({ headerProps, heading, orderCountLabel }) => (React.createElement(Node, Object.assign({}, headerProps),
    heading,
    orderCountLabel));
export const OrderHistorySalesOder = ({ salesOrderProps, orderInfomation, groups, orderDetailsLink }) => (React.createElement(Node, Object.assign({}, salesOrderProps),
    orderInfomation && React.createElement(OrderHistoryOrderInfomation, Object.assign({}, orderInfomation)),
    groups && React.createElement(OrderHistoryGroups, Object.assign({}, groups)),
    orderDetailsLink));
export const OrderHistoryList = ({ listProps, salesOrders }) => (React.createElement(Node, Object.assign({}, listProps), salesOrders && salesOrders.map((salesOrder, index) => React.createElement(OrderHistorySalesOder, Object.assign({ key: index }, salesOrder)))));
const OrderHistoryView = ({ orderHistoryProps, header, alert, loading, emptyMessage, backToShoppingLink, list, moreButton }) => (React.createElement(Module, Object.assign({}, orderHistoryProps),
    header && React.createElement(OrderHistoryHeader, Object.assign({}, header)),
    loading,
    alert && (React.createElement(React.Fragment, null,
        alert,
        backToShoppingLink)),
    emptyMessage && (React.createElement(React.Fragment, null,
        emptyMessage,
        backToShoppingLink)),
    list && React.createElement(OrderHistoryList, Object.assign({}, list)),
    moreButton && moreButton));
export default OrderHistoryView;
