// @msdyn365-commerce-modules/address/dist/lib/common/components/address-select.js
import * as React from 'react';
import AddressButtonComponent from './address-button';
import AddressInputComponent from './address-input';
import { AddressShow } from './address-show';
const getInput = (index, address, props) => {
    const { addresses, onAddressOptionChange, selectedAddress } = props;
    const ichecked = address.RecordId === selectedAddress.RecordId;
    const additionalAttributes = {
        checked: ichecked,
        'aria-checked': ichecked,
        'aria-setsize': addresses.length,
        'aria-posinset': index + 1
    };
    return (React.createElement(AddressInputComponent, Object.assign({}, {
        className: 'msc-address-select',
        name: 'selectAddressOptions',
        type: 'radio',
        value: (address.RecordId || '').toString(),
        onChange: onAddressOptionChange,
        additionalAddributes: additionalAttributes,
        telemetryContent: props.telemetryContent
    })));
};
const getAddressSelectItems = (props) => {
    const { addresses, addressFormat, addressPurposes } = props;
    return addresses.map((address, index) => {
        return {
            key: address.RecordId || 0,
            SelectItem: { className: 'msc-address-select__item' },
            input: getInput(index, address, props),
            showItems: AddressShow({ addressFormat: addressFormat.getAddressFormat(address.ThreeLetterISORegionName || ''), address: address, addressPurposes: addressPurposes })
        };
    });
};
export const AddressSelect = (props) => {
    const { resources, onCancel, onSave, onAddAddress, hasExternalSubmitGroup } = props;
    return {
        SelectAddress: { className: 'msc-address-select' },
        addButton: (React.createElement(AddressButtonComponent, Object.assign({}, {
            className: 'msc-address-select__button-add',
            text: resources.addressAddButtonText,
            ariaLabel: resources.addressAddButtonAriaLabel,
            onClick: onAddAddress,
            telemetryContent: props.telemetryContent
        }))),
        isShowSaveButton: !hasExternalSubmitGroup,
        saveButton: (React.createElement(AddressButtonComponent, Object.assign({}, {
            className: 'msc-address-select__button-save',
            text: resources.addressSaveButtonText,
            ariaLabel: resources.addressSaveButtonAriaLabel,
            onClick: onSave,
            telemetryContent: props.telemetryContent
        }))),
        isShowCancelButton: !hasExternalSubmitGroup,
        cancelButton: (React.createElement(AddressButtonComponent, Object.assign({}, {
            className: 'msc-address-select__button-cancel',
            text: resources.addressCancelButtonText,
            ariaLabel: resources.addressCancelButtonAriaLabel,
            onClick: onCancel,
            telemetryContent: props.telemetryContent
        }))),
        items: getAddressSelectItems(props)
    };
};
