// @msdyn365-commerce-modules/buybox/dist/lib/modules/buybox/components/dropdown/dropdown.js
import { getPayloadObject, getTelemetryAttributes } from '@msdyn365-commerce-modules/utilities';
import * as React from 'react';
export class Dropdown extends React.PureComponent {
    constructor(props, state) {
        super(props);
        this.selectMenu = React.createRef();
        this._onChanged = async (event) => {
            this.setState({ selectedIndex: event.target.value });
            if (this.props.onChange) {
                await this.props.onChange({
                    dropdownId: this.props.dropdownId,
                    selectId: event.target.value,
                    selectedValue: event.target.innerText
                });
            }
        };
        this.state = {
            selectedIndex: ''
        };
        this.payLoad = getPayloadObject('click', props.telemetryContent, this.props.dropdownName);
    }
    componentDidMount() {
        if (this.selectMenu.current) {
            this.selectMenu.current.selectedIndex = 0;
        }
    }
    render() {
        const { dropdownId, dropdownList, dropdownName, dropdownToggleName } = this.props;
        const attribute = getTelemetryAttributes(this.props.telemetryContent, this.payLoad);
        return (React.createElement("select", Object.assign({ id: `msc-dropdown__entry-${dropdownId}`, ref: this.selectMenu, "aria-label": dropdownName, className: 'msc-dropdown__select', onChange: this._onChanged }, attribute),
            React.createElement("option", { value: '', "aria-selected": this.state.selectedIndex === '', selected: this.state.selectedIndex === '', hidden: true, disabled: true }, dropdownToggleName),
            dropdownList.map((item) => item && React.createElement("option", { value: item.id, key: item.id, selected: this.state.selectedIndex === item.id.toString(), "aria-selected": this.state.selectedIndex === item.id.toString() }, item.value))));
    }
}
