// @msdyn365-commerce-modules/utilities/src/drawer/Drawer.examples.tsx
import * as React from 'react';
import {
    Button,
   /*  Card,
    CardBody,
    CardDeck,
    CardImg,
    CardSubtitle,
    CardText,
    CardTitle,
    ICollapseProps */
} from '../';
import Drawer from './Drawer';

export interface IDrawerParentControlExampleProps {
    /* collapseProps: ICollapseProps; */
    initState?: boolean;
}

export interface IDrawerParentControlState {
    isOpen: boolean;
}
// tslint:disable:no-http-string
/**
 * DrawerParentControlExample - This is an example where the parent controls
 * the toggling of the drawer and the drawer can toggle itself too on clicking of it
 */

export default class DrawerParentControlExample extends React.Component<
    IDrawerParentControlExampleProps,
    IDrawerParentControlState
> {
    constructor(props: IDrawerParentControlExampleProps) {
        super(props);
        this._toggle    = this._toggle.bind(this);
        this._onToggle  = this._onToggle.bind(this);
        this.state = { isOpen: this.props.initState || false };
    }

    public render(): JSX.Element {
        // const drawerCollapseProps = { ...this.props.collapseProps } as ICollapseProps;
        // drawerCollapseProps.isOpen = this.state.isOpen;
        // drawerCollapseProps.timeout = 700;

        return (
            <div>
                <Button
                    className='mb-3'
                    color='warning'
                    onClick={this._toggle}
                    /*aria-expanded={drawerCollapseProps.isOpen}*/
                >
                    Click here to toggle the drawer
                </Button>
                <Drawer
                    toggleButtonText={'Select Shipping Address'}
                    toggleButtonProps={{
                        color: 'secondary',
                        tag: 'button',
                        className: 'position-relative'
                    }}
                    onToggle={this._onToggle}
                    openGlyph={'fa fa-chevron-up'}
                    closeGlyph={'fa fa-chevron-down'}
                    glyphPlacement={'end'}
                >
                    <div className='card drawer__card'>
                        body mock content
                    </div>
                </Drawer>
               {/*  <Drawer
                    toggleButtonText={'Select Shipping Address'}
                    collapseProps={drawerCollapseProps}
                    toggleButtonProps={{
                        color: 'secondary',
                        tag: 'button',
                        className: 'position-relative'
                    }}
                    onToggle={this._onToggle}
                    openGlyph={'fa fa-chevron-up'}
                    closeGlyph={'fa fa-chevron-down'}
                    glyphPlacement={'end'}
                >
                    <div className='card drawer__card'>
                        <CardDeck>
                            <Card>
                                <CardImg
                                    top
                                    width='100%'
                                    src='https://img-prod-cms-mr-microsoft-com.akamaized.net/cms/api/fabrikam/imageFileData/MA1VFr?ver=cdf2'
                                    alt='Card image cap'
                                />
                                <CardBody>
                                    <CardTitle>Home</CardTitle>
                                    <CardSubtitle>Milky Way Galaxy</CardSubtitle>
                                    <CardText>Somewhere in outerspace</CardText>
                                    <Button>Ship to this address</Button>
                                </CardBody>
                            </Card>
                            <Card>
                                <CardImg
                                    top
                                    width='100%'
                                    src='https://img-prod-cms-mr-microsoft-com.akamaized.net/cms/api/fabrikam/imageFileData/MA1VFr?ver=cdf2'
                                    alt='Card image cap'
                                />
                                <CardBody>
                                    <CardTitle>Work</CardTitle>
                                    <CardSubtitle>Andromeda Galaxy</CardSubtitle>
                                    <CardText>Somewhere in outerspace</CardText>
                                    <Button>Ship to this address</Button>
                                </CardBody>
                            </Card>
                            <Card>
                                <CardImg
                                    top
                                    width='100%'
                                    src='https://img-prod-cms-mr-microsoft-com.akamaized.net/cms/api/fabrikam/imageFileData/MA1VFr?ver=cdf2'
                                    alt='Card image cap'
                                />
                                <CardBody>
                                    <CardTitle>Store Pickup</CardTitle>
                                    <CardSubtitle>Pinwheel Galaxy</CardSubtitle>
                                    <CardText>Somewhere in Ursula Major.</CardText>
                                    <Button>Ship to this address</Button>
                                </CardBody>
                            </Card>
                        </CardDeck>
                    </div>
                </Drawer> */}
            </div>
        );
    }
    private _toggle(e: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLElement>): void {
        e.preventDefault();
        this.setState({ isOpen: !this.state.isOpen });
    }
    private _onToggle(state:IDrawerParentControlState): void {
        this.setState({isOpen:!state.isOpen});
    }
}
