// @msdyn365-commerce-modules/checkout-utilities/dist/lib/module-state-manager/module-state.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
/**
 * GetModulesInput
 */
var ModuleStatesCacheKey = /** @class */ (function () {
    function ModuleStatesCacheKey() {
        this.getCacheKey = function () { return 'ModuleState'; };
        this.getCacheObjectType = function () { return 'ModuleState'; };
        this.dataCacheType = function () { return 'none'; };
    }
    return ModuleStatesCacheKey;
}());
export { ModuleStatesCacheKey };
export function getModuleStates(ctx) {
    var cacheKey = new ModuleStatesCacheKey();
    // @ts-ignore
    var moduleStates = ctx.get(cacheKey.getCacheObjectType(), cacheKey.getCacheKey());
    if (moduleStates) {
        return moduleStates;
    }
    ctx.update(cacheKey, {});
    moduleStates = ctx.get(cacheKey.getCacheObjectType(), cacheKey.getCacheKey());
    return moduleStates;
}
export function updateModuleStates(value, ctx) {
    var cacheKey = new ModuleStatesCacheKey();
    ctx.update(cacheKey, value);
}
