// @msdyn365-commerce-modules/checkout/src/modules/gift-card-balance-check/tests/__mock__.ts
import { GiftCard } from '@msdyn365-commerce/retail-proxy/dist/Entities/CommerceTypes.g';
import { IGiftCardBalanceCheckResources } from '../gift-card-balance-check.props.autogenerated';

export const mockResources: IGiftCardBalanceCheckResources = {
    emptyInputError: 'Please enterd a gift card number',
    invalidCodeError: 'Invalid gift card code entered',
    invalidCardInfoError: 'Invalid gift card information entered',
    invalidCardTypeError: 'The gift card type is not suported',
    noCardPinError: 'Gift card PIN is required',
    noCardExpError: 'Gift card expiration date is required',
    noBalanceError: 'Gift card has no balance left',
    giftCardFormLabel: 'Gift card',
    giftCardNumberLabel: 'Gift card number',
    giftCardPinLabel: 'PIN',
    giftCardExpLabel: 'Expiration date',
    giftCardAlertLabel: 'You gift card may or may not include these fields',
    balanceCheckButton: 'Balance check',
    giftCardAvailableBalanceLabel: 'Available balance',
    giftCardPinPlaceholderText: 'PIN',
    giftCardExpPlaceholderText: 'MM/YYYY',
    enteredGiftCardLabel: 'Gift card ending in'
};

export const mockGiftCard1: GiftCard = {
    Id: 'abc',
    Balance: 100,
    BalanceCurrencyCode: 'USD',
    BalanceInCardCurrency: 100,
    CardCurrencyCode: 'USD'
};

export const mockGiftCard2: GiftCard = {
    Id: 'efg',
    Balance: 100,
    BalanceCurrencyCode: 'USD',
    BalanceInCardCurrency: 100,
    CardCurrencyCode: 'USD'
};
