// @msdyn365-commerce/build-scripts-internal/dist/lib/__mocks__/installed/msdyn365-exp-test/msdyn365-exp-test.provider.js
"use strict";
/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */
// tslint:disable: no-any
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * A basic implementation of the ExperimentationProvider interface used for testing
 */
var MsDyn365ExpTestConnector = /** @class */ (function () {
    function MsDyn365ExpTestConnector() {
    }
    MsDyn365ExpTestConnector.prototype.getConfigForClientSideInit = function () {
        return {};
    };
    MsDyn365ExpTestConnector.prototype.initializeClientSide = function (config) {
        console.log("Initialize client side called on MsDyn365ExpTestConnector with config " + config);
        return true;
    };
    MsDyn365ExpTestConnector.prototype.getExperiments = function () {
        return {
            name: 'msdyn365-exp-test-2',
            experiments: [
                {
                    id: 'expid1',
                    friendlyName: 'experiment-1',
                    status: 1,
                    variations: [
                        {
                            id: 'var1',
                            friendlyName: 'variant-1',
                            status: 1
                        },
                        {
                            id: 'var2',
                            friendlyName: 'variant-2',
                            status: 2
                        }
                    ]
                },
                {
                    id: 'expid1',
                    friendlyName: 'experiment-2',
                    status: 0,
                    variations: [
                        {
                            id: 'var3',
                            friendlyName: 'variant-3',
                            status: 0
                        }
                    ]
                }
            ]
        };
    };
    MsDyn365ExpTestConnector.prototype.getVariantsForUser = function (userId, attributes) {
        return [
            {
                variantId: 'var1',
                experimentId: 'expid1'
            }
        ];
    };
    MsDyn365ExpTestConnector.prototype.activateExperiment = function (userId, experiments, attributes) {
        console.log("Active experiment called on MsDyn365ExpTestConnector with userId: " + userId + ", experiments: " + JSON.stringify(experiments) + ", attributes: " + JSON.stringify(attributes));
        return true;
    };
    MsDyn365ExpTestConnector.prototype.initialize = function (config) {
        console.log("MsDyn365 Exp Test Connector called with config: " + JSON.stringify(config));
        return true;
    };
    return MsDyn365ExpTestConnector;
}());
var connector = new MsDyn365ExpTestConnector();
exports.default = connector;
