// @msdyn365-commerce-modules/utilities/dist/lib/labeled-dropdown/LabeledDropdown.examples.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = tslib_1.__importStar(require("react"));
var LabeledDropdown_1 = tslib_1.__importDefault(require("./LabeledDropdown"));
var LabeledDropdownExample = (function (_super) {
    tslib_1.__extends(LabeledDropdownExample, _super);
    function LabeledDropdownExample(props) {
        var _this = _super.call(this, props) || this;
        _this._onSelectedOption = function (notification) {
            alert("Selected option key=" + notification.selectedOption.key + " value=" + notification.selectedOption.value);
        };
        _this._onSelectedOption = _this._onSelectedOption.bind(_this);
        return _this;
    }
    LabeledDropdownExample.prototype.render = function () {
        var options = [
            { key: 'a', value: 'apple' },
            { key: 'b', value: 'banana' },
            { key: 'c', value: 'cucumber' },
            { key: 'delta', value: 'pear' },
            { key: 'e', value: 'tomato' },
        ];
        return (React.createElement(LabeledDropdown_1.default, { dropdownOptions: options, selectedOption: options[1], onSelectOption: this._onSelectedOption, labelText: 'fruit', dropdownId: 'testDropdownId' }));
    };
    return LabeledDropdownExample;
}(React.Component));
exports.LabeledDropdownExample = LabeledDropdownExample;
