// @msdyn365-commerce-modules/wishlist/src/modules/wishlist-items/wishlist-items.props.autogenerated.ts
/**
 * Copyright (c) 2018 Microsoft Corporation
 * IWishlistItems contentModule Interface Properties
 * THIS FILE IS AUTO-GENERATED - MANUAL MODIFICATIONS WILL BE LOST
 */

import * as Msdyn365 from '@msdyn365-commerce/core';

export interface IWishlistItemsConfig extends Msdyn365.IModuleConfig {
    heading?: IHeadingData;
    imageSettings?: Msdyn365.IImageSettings;
    className?: string;
    enableImageProductLink?: boolean;
}

export interface IWishlistItemsResources {
    emptyWishlistText: string;
    priceFree: string;
    unableToGetWishlist: string;
    removeButtonText: string;
    addToCartButtonText: string;
    removeButtonAriaLabel: string;
    waitingClass: string;
    outOfStockText: string;
    addedToCartSuccessMessage: string;
    addedToCartFailureMessage: string;
    maxQuantityLimitText: string;
    originalPriceText: string;
    currentPriceText: string;
    productDimensionTypeColor: string;
    productDimensionTypeSize: string;
    productDimensionTypeStyle: string;
    wishlistErrorGettingWishlist: string;
}

export const enum HeadingTag {
    h1 = 'h1',
    h2 = 'h2',
    h3 = 'h3',
    h4 = 'h4',
    h5 = 'h5',
    h6 = 'h6'
}

export interface IHeadingData {
    text: string;
    tag?: HeadingTag;
}

export interface IWishlistItemsProps<T> extends Msdyn365.IModule<T> {
    resources: IWishlistItemsResources;
    config: IWishlistItemsConfig;
}
