// @msdyn365-commerce-modules/checkout/dist/lib/modules/checkout-guest-profile/components/get-add-contact-info.js
import { Button } from '@msdyn365-commerce-modules/utilities';
import classnames from 'classnames';
import * as React from 'react';
const getAddContactInfo = ({ hasError, inputRef, email, canSubmit, canCancel, onChange, onSubmit, onCancel, resources: { emailLabel, emailErrortext, saveBtnLabel, cancelBtnLabel } }) => {
    const onInputChange = (e) => {
        onChange(e.target.value || '');
    };
    return {
        addFormProps: { className: 'ms-checkout-guest-profile__add-contact-info' },
        submitButton: canSubmit && (React.createElement(Button, { className: 'ms-checkout-guest-profile__btn-save', title: saveBtnLabel, color: 'primary', onClick: onSubmit }, saveBtnLabel)),
        cancelButton: canCancel && (React.createElement(Button, { className: 'ms-checkout-guest-profile__btn-cancel', title: cancelBtnLabel, color: 'secondary', onClick: onCancel }, cancelBtnLabel)),
        inputGroupProps: { className: classnames('ms-checkout-guest-profile__input', { 'is-invalid': hasError }) },
        label: (React.createElement("label", { id: 'ms-checkout-guest-profile__label', className: 'ms-checkout-guest-profile__input-label' }, emailLabel)),
        error: hasError && (React.createElement("span", { id: 'ms-checkout-guest-profile__error', className: 'ms-checkout-guest-profile__input-error', role: 'alert', "aria-live": 'assertive' }, emailErrortext)),
        input: (React.createElement("input", { ref: inputRef, type: 'text', className: 'ms-checkout-guest-profile__input-text form-control', "aria-label": emailLabel, onChange: onInputChange, value: email, "aria-labelledby": 'ms-checkout-guest-profile__label ms-checkout-guest-profile__error' }))
    };
};
export default getAddContactInfo;
