// @msdyn365-commerce-modules/search-result-container/dist/lib/modules/search-result-container/components/modal.js
import { Button, Modal, ModalBody, ModalFooter, ModalHeader } from '@msdyn365-commerce-modules/utilities';
import React from 'react';
export const ModalToggle = (props) => {
    const { text, onClick, ariaLabel, innerRef, id } = props;
    return React.createElement("button", { id: id, className: 'msc-button', "aria-label": ariaLabel, onClick: onClick, ref: innerRef }, text);
};
export const SearchResultModal = (props) => {
    return {
        modal: modalNode(props),
        modalHeader: modalHeaderNode(props),
        modalFooter: modalFooterNode(props),
        modalBody: React.createElement(ModalBody, { className: 'msc-review-modal-body' })
    };
};
const modalNode = (props) => {
    return (React.createElement(Modal, { autoFocus: true, applicationNode: 'renderPage', returnFocusRef: props.returnRef, isOpen: props.isOpen, toggle: props.onModalToggle, className: 'msc-search-result-modal', zIndex: 1050 }));
};
const modalHeaderNode = (props) => {
    return (React.createElement(ModalHeader, { toggle: props.onModalToggle }, props.resources.modalTitle));
};
const modalFooterNode = (props) => {
    return (React.createElement(ModalFooter, null,
        React.createElement(Button, { onClick: props.onModalToggle, className: 'ms-sort-and-filter-modal-close' }, props.resources.modalCloseButtonText)));
};
