// @msdyn365-commerce-modules/account-management/dist/lib/modules/account-order-templates-tile/account-order-templates-tile.js
import { format, getTelemetryObject } from '@msdyn365-commerce-modules/utilities';
import * as Msdyn365 from '@msdyn365-commerce/core';
import classnames from 'classnames';
import * as React from 'react';
import { AccountOrderTemplatesTileDescription, AccountOrderTemplatesTileLinks } from './components';
export class AccountOrderTemplatesTile extends React.PureComponent {
    constructor(props) {
        super(props);
        this.handleHeadingChange = (event) => {
            this.props.config.heading.text = event.target.value;
        };
        this.handleLinkTextChange = (linkIndex) => (event) => {
            if (this.props.config.links && this.props.config.links[linkIndex]) {
                this.props.config.links[linkIndex].linkText = event.target.value;
            }
        };
        this.telemetryContent = getTelemetryObject(this.props.context.request.telemetryPageName, this.props.friendlyName, this.props.telemetry);
    }
    render() {
        const { config, resources, data } = this.props;
        const RetailB2BEcommerceFeatureState = this.props.data.featureState.result?.find(featureState => featureState.Name === 'Dynamics.AX.Application.RetailB2BEcommerceFeature');
        if (!RetailB2BEcommerceFeatureState?.IsEnabled || !this.props.data.customerInformation?.result?.IsB2b) {
            return React.createElement(React.Fragment, null);
        }
        const { className, heading, links } = config;
        const { accountOrderTemplatesTileDescription } = resources;
        const orderTemplates = data.orderTemplates;
        const orderTemplatesCount = (orderTemplates?.result || []).length;
        const description = format(accountOrderTemplatesTileDescription, orderTemplatesCount);
        const viewProps = {
            ...this.props,
            className: className,
            AccountOrderTemplatesTile: {
                moduleProps: this.props,
                className: classnames('ms-account-order-templates-tile', config.className)
            },
            heading: heading && heading.text && (React.createElement(Msdyn365.Text, { className: 'ms-account-order-templates-tile__heading', tag: heading.tag, text: heading.text, editProps: { onEdit: this.handleHeadingChange, requestContext: this.props.context.request } })),
            links: links && links.length > 0 && React.createElement(AccountOrderTemplatesTileLinks, Object.assign({}, { links: links, onTextChange: this.handleLinkTextChange, requestContext: this.props.context.request, telemetryContent: this.telemetryContent })),
            description: React.createElement(AccountOrderTemplatesTileDescription, { description: description })
        };
        return this.props.renderView(viewProps);
    }
}
export default AccountOrderTemplatesTile;
