// @msdyn365-commerce-modules/order-management/src/modules/order-confirmation/tests/__mock__.ts
import { Address, SalesOrder, SimpleProduct, TenderLine, TenderType } from '@msdyn365-commerce/retail-proxy/dist/Entities/CommerceTypes.g';

export const mockShippingAddress: Address = {
    Name: 'Gloria Deng',
    Street: 'One Microsoft Way',
    City: 'Redmond',
    State: 'wa',
    ZipCode: '98052',
    ThreeLetterISORegionName: 'usa'
};

export const mockStoreAddress: Address = {
    Name: 'Fabrikam Company Store',
    FullAddress: '15010 NE 36th St↵Redmond, WA 98052↵USA',
    Street: '15010 NE 36th St',
    City: 'Redmond',
    County: 'KING',
    ThreeLetterISORegionName: 'USA',
    State: 'WA',
    ZipCode: '98052'
};

export const mockShipments = [
    {
        ShipmentId: '123',
        TrackingNumber: '123456',
        TrackingUrl: '/tracking'
    }
];

export const mockTenderLines: TenderLine[] = [
    {
        TenderTypeId: '3',
        MaskedCardNumber: '************1111',
        Currency: 'USD',
        CardTypeId: 'VISA'
    },
    {
        TenderTypeId: '8',
        GiftCardId: 'gldeng_100_3',
        Currency: 'USD',
        Amount: 100
    },
    {
        TenderTypeId: '1',
        Currency: 'USD',
        Amount: 100
    },
    {
        TenderTypeId: '10',
        LoyaltyCardId: '890609',
        Currency: 'USD',
        Amount: 100
    }
];

export const mockSalesOrder: SalesOrder = {
    Id: '0HHRKonY4riLeLAjxZAZTltqsZrMVPvt',
    ReceiptEmail: 'john@example.com',
    ChannelId: 5637145359,
    ChannelReferenceId: 'C764D04GLDEN',
    CurrencyCode: 'USD',
    SalesId: '012660',
    SalesLines: [
        {
            ProductId: 22565430654,
            TotalAmount: 300,
            SalesStatusValue: 0,
            Quantity: 1,
            DeliveryMode: '60',
            ShippingAddress: mockShippingAddress,
            TrackingId: '123'
        },
        {
            ProductId: 22565430655,
            TotalAmount: 200,
            SalesStatusValue: 0,
            Quantity: 2,
            DeliveryMode: '70',
            ShippingAddress: mockStoreAddress
        }
    ],
    Shipments: mockShipments,
    SubtotalAmount: 300,
    ChargeAmount: 0,
    SubtotalAmountWithoutTax: 300,
    TaxAmount: 19.5,
    TotalAmount: 319.5,
    TenderLines: mockTenderLines
};

export const mockTransaction: SalesOrder = {
    Id: 'UiE~EWeXya9o0byTIE7KeP8bIEjpGv9-',
    ReceiptEmail: 'john@example.com',
    ChannelId: 5637145359,
    ChannelReferenceId: '0641E90GLDEN',
    CurrencyCode: 'USD',
    SalesId: undefined,
    SalesLines: [
        {
            ProductId: 22565430654,
            TotalAmount: 300,
            SalesStatusValue: 0,
            Quantity: 1,
            DeliveryMode: '60',
            ShippingAddress: mockShippingAddress
        },
        {
            ProductId: 22565430655,
            TotalAmount: 200,
            SalesStatusValue: 0,
            Quantity: 2,
            DeliveryMode: '70',
            ShippingAddress: mockStoreAddress
        }
    ],
    SubtotalAmount: 300,
    ChargeAmount: 0,
    SubtotalAmountWithoutTax: 300,
    TaxAmount: 19.5,
    TotalAmount: 319.5,
    TenderLines: mockTenderLines
};

export const mockProducts: SimpleProduct[] = [
    {
        RecordId: 22565430654,
        ItemId: '81307',
        Name: 'Red Leather Bag',
        ProductTypeValue: 4,
        BasePrice: 300,
        Price: 300,
        AdjustedPrice: 300,
        PrimaryImageUrl:
            'https://img-prod-cms-mr-microsoft-com.akamaized.net/cms/api/fabrikam/imageFileData/search?fileName=/Products%2F81307_000_001.png'
    },
    {
        RecordId: 22565430655,
        ItemId: '81308',
        Name: 'Brown Leather Bag',
        ProductTypeValue: 4,
        BasePrice: 200,
        Price: 200,
        AdjustedPrice: 200,
        PrimaryImageUrl:
            'https://img-prod-cms-mr-microsoft-com.akamaized.net/cms/api/fabrikam/imageFileData/search?fileName=/Products%2F81307_000_001.png',
        // @ts-ignore: Mock partial value
        Dimensions: [
            {
                DimensionTypeValue: 1,
                // @ts-ignore: Mock partial value
                DimensionValue: {
                    Value: 'Brown'
                }
            },
            {
                DimensionTypeValue: 3,
                // @ts-ignore: Mock partial value
                DimensionValue: {
                    Value: 'Map'
                }
            },
            {
                DimensionTypeValue: 4,
                // @ts-ignore: Mock partial value
                DimensionValue: {
                    Value: 'Cool'
                }
            }
        ]
    }
];

export const mockRequestContextJson = <unknown>{
    url: {
        serverPageUrl: 'https://localhost:3000'
    },
    locale: 'en-us',
    apiSettings: {
        baseUrl: '',
        baseImageUrl: '',
        channelId: 0,
        oun: '',
        ratingsReviewsEndpoint: ''
    },
    user: {
        token: 'Dummy token',
        isAuthenticated: true
    },
    urlTokens: {
        locale: 'en-us',
        categories: ['fashion-sunglasses']
    }
};

export const mockTenderTypes: TenderType[] = [
    {
        Function: 0,
        TenderTypeId: '1',
        Name: 'Cash',
        OperationId: 200,
        OperationName: 'Pay cash'
    },
    {
        Function: 1,
        TenderTypeId: '3',
        Name: 'Cards',
        OperationId: 201,
        OperationName: 'Pay card'
    },
    {
        Function: 0,
        TenderTypeId: '8',
        Name: 'Gift Card',
        OperationId: 214,
        OperationName: 'Pay gift card'
    },
    {
        Function: 1,
        TenderTypeId: '10',
        Name: 'Loyalty Cards',
        OperationId: 207,
        OperationName: 'Pay loyalty card'
    }
];
