// @msdyn365-commerce/telemetry-internal/dist/lib/events/server-render.js
/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */
export const EVENT_SERVER_RENDER_START = 'ServerRenderEventStart';
export const EVENT_SERVER_RENDER_STOP = 'ServerRenderEventStop';
export const EVENT_VALIDATE_ROUTE_START = 'ValidateRouteEventStart';
export const EVENT_VALIDATE_ROUTE_STOP = 'ValidateRouteEventStop';
export const EVENT_RENDER_PAGE_START = 'RenderPageEventStart';
export const EVENT_RENDER_PAGE_STOP = 'RenderPageEventStop';
export const EVENT_RENDER_TO_STRING_START = 'RenderToStringEventStart';
export const EVENT_RENDER_TO_STRING_STOP = 'RenderToStringEventStop';
export const EVENT_BASEPAGE_CREATION_START = 'BasePageCreationEventStart';
export const EVENT_BASEPAGE_CREATION_STOP = 'BasePageCreationEventStop';
export const EVENT_JSON_STRINGIFY_START = 'JsonStringifyEventStart';
export const EVENT_JSON_STRINGIFY_STOP = 'JsonStringifyEventStop';
export const EVENT_REQUEST_PREP_START = 'RequestPrepEventStart';
export const EVENT_REQUEST_PREP_STOP = 'RequestPrepEventStop';
export const EVENT_COMPLEX_TYPE_START = 'ComplexTypeEventStart';
export const EVENT_COMPLEX_TYPE_STOP = 'ComplexTypeEventStop';
