// @msdyn365-commerce-modules/account-management/src/modules/account-customer-balance/account-customer-balance.props.autogenerated.ts
/**
 * Copyright (c) 2018 Microsoft Corporation
 * IAccountCustomerBalance contentModule Interface Properties
 * THIS FILE IS AUTO-GENERATED - MANUAL MODIFICATIONS WILL BE LOST
 */

import * as Msdyn365 from '@msdyn365-commerce/core';

export interface IAccountCustomerBalanceConfig extends Msdyn365.IModuleConfig {
    heading: IHeadingData;
}

export interface IAccountCustomerBalanceResources {
    creditSummarySectionHeading: string;
    accountSectionHeading: string;
    backToShoppingText: string;
    customerSincePrefix: string;
    availableCreditLabel: string;
    balanceOwedLabel: string;
    creditExcessLabel: string;
}

export const enum HeadingTag {
    h1 = 'h1',
    h2 = 'h2',
    h3 = 'h3',
    h4 = 'h4',
    h5 = 'h5',
    h6 = 'h6'
}

export interface IHeadingData {
    text: string;
    tag?: HeadingTag;
}

export interface IAccountCustomerBalanceProps<T> extends Msdyn365.IModule<T> {
    resources: IAccountCustomerBalanceResources;
    config: IAccountCustomerBalanceConfig;
}
