// @msdyn365-commerce-modules/utilities/dist/lib/common/KeyCodes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var KeyCodes;
(function (KeyCodes) {
    KeyCodes[KeyCodes["Back"] = 8] = "Back";
    KeyCodes[KeyCodes["Tab"] = 9] = "Tab";
    KeyCodes[KeyCodes["Enter"] = 13] = "Enter";
    KeyCodes[KeyCodes["Shift"] = 16] = "Shift";
    KeyCodes[KeyCodes["Ctrl"] = 17] = "Ctrl";
    KeyCodes[KeyCodes["Alt"] = 18] = "Alt";
    KeyCodes[KeyCodes["Break"] = 19] = "Break";
    KeyCodes[KeyCodes["CapsLock"] = 20] = "CapsLock";
    KeyCodes[KeyCodes["Escape"] = 27] = "Escape";
    KeyCodes[KeyCodes["Space"] = 32] = "Space";
    KeyCodes[KeyCodes["PageUp"] = 33] = "PageUp";
    KeyCodes[KeyCodes["PageDown"] = 34] = "PageDown";
    KeyCodes[KeyCodes["End"] = 35] = "End";
    KeyCodes[KeyCodes["Home"] = 36] = "Home";
    KeyCodes[KeyCodes["ArrowLeft"] = 37] = "ArrowLeft";
    KeyCodes[KeyCodes["ArrowUp"] = 38] = "ArrowUp";
    KeyCodes[KeyCodes["ArrowRight"] = 39] = "ArrowRight";
    KeyCodes[KeyCodes["ArrowDown"] = 40] = "ArrowDown";
    KeyCodes[KeyCodes["Print"] = 44] = "Print";
    KeyCodes[KeyCodes["Insert"] = 45] = "Insert";
    KeyCodes[KeyCodes["Delete"] = 46] = "Delete";
    KeyCodes[KeyCodes["Colon2"] = 59] = "Colon2";
    KeyCodes[KeyCodes["Equals2"] = 61] = "Equals2";
    KeyCodes[KeyCodes["Equals3"] = 107] = "Equals3";
    KeyCodes[KeyCodes["Minus2"] = 109] = "Minus2";
    KeyCodes[KeyCodes["Period"] = 190] = "Period";
    KeyCodes[KeyCodes["WindowsLeft"] = 91] = "WindowsLeft";
    KeyCodes[KeyCodes["WindowsRight"] = 92] = "WindowsRight";
    KeyCodes[KeyCodes["WindowsOpera"] = 219] = "WindowsOpera";
    KeyCodes[KeyCodes["Menu"] = 93] = "Menu";
    KeyCodes[KeyCodes["NumPad0"] = 96] = "NumPad0";
    KeyCodes[KeyCodes["NumPad1"] = 97] = "NumPad1";
    KeyCodes[KeyCodes["NumPad2"] = 98] = "NumPad2";
    KeyCodes[KeyCodes["NumPad3"] = 99] = "NumPad3";
    KeyCodes[KeyCodes["NumPad4"] = 100] = "NumPad4";
    KeyCodes[KeyCodes["NumPad5"] = 101] = "NumPad5";
    KeyCodes[KeyCodes["NumPad6"] = 102] = "NumPad6";
    KeyCodes[KeyCodes["NumPad7"] = 103] = "NumPad7";
    KeyCodes[KeyCodes["NumPad8"] = 104] = "NumPad8";
    KeyCodes[KeyCodes["NumPad9"] = 105] = "NumPad9";
    KeyCodes[KeyCodes["NumPadMultiply"] = 106] = "NumPadMultiply";
    KeyCodes[KeyCodes["NumPadPlus"] = 107] = "NumPadPlus";
    KeyCodes[KeyCodes["NumPadMinus"] = 109] = "NumPadMinus";
    KeyCodes[KeyCodes["NumPadDot"] = 110] = "NumPadDot";
    KeyCodes[KeyCodes["NumPadDivide"] = 111] = "NumPadDivide";
    KeyCodes[KeyCodes["Function1"] = 112] = "Function1";
    KeyCodes[KeyCodes["Function2"] = 113] = "Function2";
    KeyCodes[KeyCodes["Function3"] = 114] = "Function3";
    KeyCodes[KeyCodes["Function4"] = 115] = "Function4";
    KeyCodes[KeyCodes["Function5"] = 116] = "Function5";
    KeyCodes[KeyCodes["Function6"] = 117] = "Function6";
    KeyCodes[KeyCodes["Function7"] = 118] = "Function7";
    KeyCodes[KeyCodes["Function8"] = 119] = "Function8";
    KeyCodes[KeyCodes["Function9"] = 120] = "Function9";
    KeyCodes[KeyCodes["Function10"] = 121] = "Function10";
    KeyCodes[KeyCodes["Function11"] = 122] = "Function11";
    KeyCodes[KeyCodes["Function12"] = 123] = "Function12";
    KeyCodes[KeyCodes["NavigationUp"] = 138] = "NavigationUp";
    KeyCodes[KeyCodes["NavigationDown"] = 139] = "NavigationDown";
    KeyCodes[KeyCodes["NavigationLeft"] = 140] = "NavigationLeft";
    KeyCodes[KeyCodes["NavigationRight"] = 141] = "NavigationRight";
    KeyCodes[KeyCodes["NavigationAccept"] = 142] = "NavigationAccept";
    KeyCodes[KeyCodes["NumLock"] = 144] = "NumLock";
    KeyCodes[KeyCodes["ScrollLock"] = 145] = "ScrollLock";
    KeyCodes[KeyCodes["Colon"] = 186] = "Colon";
    KeyCodes[KeyCodes["Equals"] = 187] = "Equals";
    KeyCodes[KeyCodes["Comma"] = 188] = "Comma";
    KeyCodes[KeyCodes["Minus"] = 189] = "Minus";
    KeyCodes[KeyCodes["ForwardSlash"] = 191] = "ForwardSlash";
    KeyCodes[KeyCodes["Tilde"] = 192] = "Tilde";
    KeyCodes[KeyCodes["GamepadA"] = 195] = "GamepadA";
    KeyCodes[KeyCodes["GamepadB"] = 196] = "GamepadB";
    KeyCodes[KeyCodes["GamepadX"] = 197] = "GamepadX";
    KeyCodes[KeyCodes["GamepadY"] = 198] = "GamepadY";
    KeyCodes[KeyCodes["GamepadRightButton"] = 199] = "GamepadRightButton";
    KeyCodes[KeyCodes["GamepadLeftButton"] = 200] = "GamepadLeftButton";
    KeyCodes[KeyCodes["GamepadLeftTrigger"] = 201] = "GamepadLeftTrigger";
    KeyCodes[KeyCodes["GamepadRightTrigger"] = 202] = "GamepadRightTrigger";
    KeyCodes[KeyCodes["GamepadDPadUp"] = 203] = "GamepadDPadUp";
    KeyCodes[KeyCodes["GamepadDPadDown"] = 204] = "GamepadDPadDown";
    KeyCodes[KeyCodes["GamepadDPadLeft"] = 205] = "GamepadDPadLeft";
    KeyCodes[KeyCodes["GamepadDPadRight"] = 206] = "GamepadDPadRight";
    KeyCodes[KeyCodes["GamepadSelect"] = 207] = "GamepadSelect";
    KeyCodes[KeyCodes["GamepadStart"] = 208] = "GamepadStart";
    KeyCodes[KeyCodes["GamepadLeftThumbstick"] = 209] = "GamepadLeftThumbstick";
    KeyCodes[KeyCodes["GamepadRightThumbstick"] = 210] = "GamepadRightThumbstick";
    KeyCodes[KeyCodes["GamepadLeftThumbstickUp"] = 211] = "GamepadLeftThumbstickUp";
    KeyCodes[KeyCodes["GamepadLeftThumbstickDown"] = 212] = "GamepadLeftThumbstickDown";
    KeyCodes[KeyCodes["GamepadLeftThumbstickRight"] = 213] = "GamepadLeftThumbstickRight";
    KeyCodes[KeyCodes["GamepadLeftThumbstickLeft"] = 214] = "GamepadLeftThumbstickLeft";
    KeyCodes[KeyCodes["GamepadRightThumbstickUp"] = 215] = "GamepadRightThumbstickUp";
    KeyCodes[KeyCodes["GamepadRightThumbstickDown"] = 216] = "GamepadRightThumbstickDown";
    KeyCodes[KeyCodes["GamepadRightThumbstickRight"] = 217] = "GamepadRightThumbstickRight";
    KeyCodes[KeyCodes["GamepadRightThumbstickLeft"] = 218] = "GamepadRightThumbstickLeft";
    KeyCodes[KeyCodes["OpenBracket"] = 219] = "OpenBracket";
    KeyCodes[KeyCodes["BackSlash"] = 220] = "BackSlash";
    KeyCodes[KeyCodes["CloseBracket"] = 221] = "CloseBracket";
    KeyCodes[KeyCodes["Quote"] = 222] = "Quote";
})(KeyCodes = exports.KeyCodes || (exports.KeyCodes = {}));
var KeycodesClass = (function () {
    function KeycodesClass() {
    }
    KeycodesClass.prototype.testCoverage = function () {
        return ':(';
    };
    return KeycodesClass;
}());
exports.KeycodesClass = KeycodesClass;
