// @msdyn365-commerce/runtime-internal/dist/lib/utils/query-utils-test-data.js
// tslint:disable
export const queryUtilsTestMockData = {
    dataProviders: [
        {
            Type: 'coreui-hero-data-provider',
            id: 'coreui-hero-zul6d1k_primary',
            data: {
                Items: [
                    {
                        Heading: 'Michael Hill',
                        Subheading: 'THIS IS A PLACEHOLDER',
                        Subheading1: null,
                        Subheading2: null,
                        Paragraph: 'We think about you and the reasons behind your purchase – designing timeless pieces that bring out your best.',
                        PrimaryImage: {
                            Title: 'BeautifulNecklace',
                            Href: 'http://img-prod-cms-rt-microsoft-com.akamaized.net/cms/api/am/imageFileData/RE238nJ?ver=aa0c',
                            AltText: 'Feather layered necklace',
                            Attribution: null,
                            Caption: null,
                            Description: null,
                            FocalRegion: {
                                X1: 299,
                                Y1: 288,
                                X2: 299,
                                Y2: 288
                            },
                            CropRegions: [],
                            FilmstripMetaData: null,
                            IsLocalized: false,
                            Height: 600,
                            Photographer: null,
                            Quality: 100,
                            SourceMimeType: null,
                            Width: 1600
                        },
                        SecondaryImage: null,
                        Video: null,
                        BackgroundColor: '808080',
                        ForegroundColor: null,
                        Mask: 0,
                        TextPosition: 5,
                        Theme: 1,
                        ProductType: 0,
                        Logo: {
                            Image: null,
                            DestinationUrl: null,
                            Text: null,
                            IconType: 0
                        },
                        Badge: {
                            Text: null,
                            BadgeSize: 1,
                            BadgeType: 0,
                            BadgeCategory: 0
                        },
                        Prices: null,
                        Rating: null,
                        Links: [
                            {
                                Title: 'Shop necklaces',
                                Href: '/en-us/rushmore/listpage?categoryid=68719478073',
                                AltText: null,
                                AriaLabel: 'Shop necklaces',
                                Image: null,
                                SubLinks: null,
                                IsSelected: false,
                                Language: null,
                                TelemetrySource: 0,
                                TelemetryId: null,
                                TelemetryProperties: null,
                                LinkActionType: 0,
                                OpenInNewTab: false
                            },
                            {
                                Title: 'Shop collections',
                                Href: '/en-us/rushmore/listpage?categoryid=68719478072',
                                AltText: null,
                                AriaLabel: 'Shop collections',
                                Image: null,
                                SubLinks: null,
                                IsSelected: false,
                                Language: null,
                                TelemetrySource: 0,
                                TelemetryId: null,
                                TelemetryProperties: null,
                                LinkActionType: 0,
                                OpenInNewTab: false
                            }
                        ],
                        PdpLinkModel: null,
                        IsExternalSource: false,
                        ProductName: null,
                        ProductId: null,
                        SkuId: null,
                        AvailabilityId: null,
                        BackgroundImage: null,
                        GlyphId: null,
                        Expiration: null,
                        OverrideDefaultGlyphWithImage: false,
                        TargetingInfo: {
                            TelemetryTags: {
                                varId: 'dbg8kki',
                                choiceId: '3xbbcb0'
                            },
                            TargetingTags: {}
                        },
                        ProductFamilyName: null,
                        PackageFamilyNames: null,
                        ContentIds: null,
                        AlternateIds: null
                    }
                ],
                TotalNumberOfItemsInCollection: 0,
                Title: null,
                Description: null,
                BackgroundImage: null,
                BackgroundColor: null,
                LinkModel: null,
                SkipCount: 0
            },
            config: {
                targetingInfo: [
                    {
                        TelemetryTags: {
                            varId: 'dbg8kki',
                            choiceId: '3xbbcb0'
                        },
                        TargetingTags: {}
                    }
                ],
                'hrefs[0]': 'cms/api/embedded/promotionalItem/cP8ZcandVgJ',
                href: 'cms/api/embedded/promotionalItem/cP8ZcandVgJ'
            }
        }
    ],
    childeren: null,
    id: 'R1Hero',
    config: {
        size: 'Medium',
        isprimary: true
    },
    Type: 'partner-hero',
    dataBinder: 'partner-hero-data-binder'
};
