// @msdyn365-commerce-modules/promo-banner/dist/lib/modules/promo-banner/components/links.js
import { getPayloadObject, getTelemetryAttributes, onTelemetryClick } from '@msdyn365-commerce-modules/utilities';
import * as Msdyn365 from '@msdyn365-commerce/core';
import * as React from 'react';
export const PromoBannerLinks = ({ links, requestContext, onTextChange, telemetryContent }) => {
    const editableLinks = _mapEditableLinks(links, telemetryContent);
    return (React.createElement("span", { className: 'ms-promo-banner__links' }, editableLinks && editableLinks.length ?
        React.createElement(Msdyn365.Links, { links: editableLinks, editProps: { onTextChange: onTextChange, requestContext: requestContext } })
        : null));
};
const _mapEditableLinks = (linkdata, telemetryContent) => {
    if (!linkdata || linkdata.length === 0) {
        return null;
    }
    const editableLinks = [];
    linkdata.forEach((link, index) => {
        const payLoad = getPayloadObject('click', telemetryContent, '', '');
        const linkText = link.linkText ? link.linkText : '';
        payLoad.contentAction.etext = linkText;
        const attributes = getTelemetryAttributes(telemetryContent, payLoad);
        const editableLink = {
            ariaLabel: link.ariaLabel,
            className: 'ms-promo-banner__link',
            linkText: link.linkText,
            linkUrl: link.linkUrl.destinationUrl,
            openInNewTab: link.openInNewTab,
            role: 'button',
            additionalProperties: attributes,
            onClick: onTelemetryClick(telemetryContent, payLoad, linkText)
        };
        editableLinks.push(editableLink);
    });
    return editableLinks;
};
