// @msdyn365-commerce-modules/fabrikam-design-kit/dist/lib/modules/fabrikam/utilities/quantity/Quantity.js
import { __decorate } from "tslib";
import { getPayloadObject, getTelemetryAttributes, TelemetryConstant } from '@msdyn365-commerce-modules/utilities';
import { debounce } from 'lodash';
import { observer } from 'mobx-react';
import * as React from 'react';
let Quantity = class Quantity extends React.PureComponent {
    constructor(props) {
        super(props);
        this.inputRef = React.createRef();
        this._validateMin = () => {
            const minValue = this.props.min === undefined ? 1 : this.props.min;
            if (isNaN(this.state.currentInput) || (this.state.currentInput < minValue)) {
                this.setState({ currentInput: minValue }, () => { this.props.onChange && this.props.onChange(this.state.currentInput); });
            }
            else {
                this.props.onChange && this.props.onChange(this.state.currentInput);
            }
        };
        this.state = { currentInput: props.currentCount || 1, isUpdateing: false };
        this._onIncrement = this._onIncrement.bind(this);
        this._onDecrement = this._onDecrement.bind(this);
        this._handleChange = this._handleChange.bind(this);
        this.payLoad = getPayloadObject('click', this.props.telemetryContent, '');
    }
    componentDidUpdate() {
        if (this.props.disabled && !this.state.isUpdateing) {
            this.setState({ isUpdateing: true });
        }
        if (!this.props.disabled && this.state.isUpdateing) {
            this.setState({ currentInput: this.props.currentCount || 1, isUpdateing: false });
        }
    }
    render() {
        const glyphMinusClassName = `${this.props.decrementGlyphClass} quantity__controls-glyph`;
        const glyphPlusClassName = `${this.props.incrementGlyphClass} quantity__controls-glyph`;
        const disabled = this.state.currentInput === this.props.min;
        const incDisabled = this.state.currentInput === this.props.max;
        const currentValue = this.state.currentInput;
        this.payLoad.contentAction.etext = TelemetryConstant.DecrementQuantity;
        const decrementAttributes = getTelemetryAttributes(this.props.telemetryContent, this.payLoad);
        this.payLoad.contentAction.etext = TelemetryConstant.IncrementQuantity;
        const incrementAttributes = getTelemetryAttributes(this.props.telemetryContent, this.payLoad);
        let extraClass = '';
        if (disabled) {
            extraClass = 'disabled';
        }
        else if (this.props.disabled) {
            extraClass = 'transition';
        }
        let extraClassInc = '';
        if (incDisabled) {
            extraClassInc = 'disabled';
        }
        else if (this.props.disabled) {
            extraClassInc = 'transition';
        }
        return (React.createElement(React.Fragment, null,
            React.createElement("div", { className: 'quantity', id: this.props.id },
                React.createElement("button", Object.assign({ disabled: this.props.disabled || disabled, title: disabled ? '' : this.props.decrementButtonAriaLabel, className: `decrement quantity__controls ${extraClass}`, onClick: this._onDecrement, "aria-hidden": true, "aria-label": `${this.props.decrementButtonAriaLabel}`, tabIndex: -1, color: 'secondary' }, decrementAttributes),
                    React.createElement("span", { className: glyphMinusClassName })),
                React.createElement("input", { type: 'number', className: 'quantity-input', pattern: '[0-9]*', value: this.state.currentInput, onChange: this._handleChange, onBlur: this._validateMin, "aria-live": 'polite', "aria-label": `${this.props.inputQuantityAriaLabel}`, role: 'spinbutton', "aria-valuemin": this.props.min, "aria-valuemax": this.props.max, "aria-valuenow": currentValue, ref: this.inputRef, disabled: this.props.disabled }),
                React.createElement("button", Object.assign({ disabled: this.props.disabled || incDisabled, title: incDisabled ? '' : this.props.incrementButtonAriaLabel, className: `increment quantity__controls ${extraClassInc}`, onClick: this._onIncrement, "aria-hidden": true, "aria-label": `${this.props.incrementButtonAriaLabel}`, tabIndex: -1, color: 'secondary' }, incrementAttributes),
                    React.createElement("span", { className: glyphPlusClassName })))));
    }
    _onIncrement() {
        let invokeCallback = false;
        const currQuantity = this.state.currentInput;
        let updatedQuantity;
        if (currQuantity < this.props.max) {
            invokeCallback = true;
            updatedQuantity = currQuantity + 1;
        }
        else {
            invokeCallback = false;
            updatedQuantity = this.props.max;
        }
        this.setState({ currentInput: updatedQuantity });
        invokeCallback && this.props.onChange && this.props.onChange(updatedQuantity);
    }
    _onDecrement() {
        let invokeCallback = false;
        const currQuantity = this.state.currentInput;
        let updatedQuantity;
        if (currQuantity > 1) {
            invokeCallback = true;
            updatedQuantity = currQuantity - 1;
        }
        else {
            invokeCallback = false;
            updatedQuantity = 1;
        }
        this.setState({ currentInput: updatedQuantity });
        invokeCallback && this.props.onChange && this.props.onChange(updatedQuantity);
    }
    _handleChange(e) {
        const currentValue = parseInt((e.target.value), 10);
        const minValue = this.props.min === undefined ? 1 : this.props.min;
        const inputElement = this.inputRef && this.inputRef.current && this.inputRef.current instanceof HTMLInputElement && this.inputRef.current;
        if (currentValue > this.props.max) {
            this.setState({ currentInput: this.props.max }, () => {
                debounce(() => {
                    this.props.onChange && this.props.onChange(this.state.currentInput);
                }, 200)();
            });
        }
        else {
            this.setState({ currentInput: currentValue }, () => {
                debounce(() => {
                    if (!isNaN(this.state.currentInput) && !(this.state.currentInput < minValue)) {
                        this.props.onChange && this.props.onChange(this.state.currentInput);
                        if (inputElement) {
                            inputElement.setAttribute('aria-valuenow', currentValue.toString());
                            inputElement.setAttribute('value', currentValue.toString());
                        }
                    }
                }, 200)();
            });
        }
    }
};
Quantity.defaultProps = {
    min: 1,
    decrementGlyphClass: 'fas fa-minus',
    incrementGlyphClass: 'fas fa-plus'
};
Quantity = __decorate([
    observer
], Quantity);
export default Quantity;
