// @msdyn365-commerce/runtime-internal/dist/lib/resources/constants.js
/*******************************************************************************
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 *******************************************************************************/
import path from 'path';
const currentDirectory = process.cwd();
const msdyn365CommerceModuleNamespace = '@msdyn365-commerce-modules';
const resolvedHoistedPath = path.resolve(currentDirectory, '../..').replace(/\\/gi, '/');
export const installedModulesResourcesPathPattern = `${currentDirectory}/node_modules/${msdyn365CommerceModuleNamespace}/*/dist/lib/**/resources/modules/*.json`;
export const installedAuthoringResourcesPathPattern = `${currentDirectory}/node_modules/${msdyn365CommerceModuleNamespace}/*/dist/lib/**/resources/authoring/*.json`;
export const hoistedModulesResourcesPathPattern = `${resolvedHoistedPath}/node_modules/${msdyn365CommerceModuleNamespace}/*/dist/lib/**/resources/modules/*.json`;
export const hoistedAuthoringResourcesPathPattern = `${resolvedHoistedPath}/node_modules/${msdyn365CommerceModuleNamespace}/*/dist/lib/**/resources/authoring/*.json`;
export const localModulesResourcesPathPattern = `${currentDirectory}/src/resources/modules/*.json`;
export const localAuthoringResourcesPathPattern = `${currentDirectory}/src/resources/authoring/*.json`;
