// @msdyn365-commerce-modules/utilities/src/back-to-top/BackToTop.props.ts
import * as Msdyn365 from '@msdyn365-commerce/core';
import { ITelemetryContent } from '../common/telemetryHelper';
/**
 * BackToTop properties.
 */
export interface IBackToTopProps {
    /** Sets the css classes */
    className?: string;
    /** Sets the #href to treat as the "top" and jump to. Must start with # */
    href?: string;
    /** Sets the localized string to use for the aria-label attribute value */
    'aria-label': string;
    /** Sets the localized string to use for the title attribute value */
    title: string;
    /** Sets the threshold value. If not specified twice the viewport height is used */
    threshold?: number;
    /** Sets the id on the element */
    id?: string;
    /** Sets the scroll steps to scrollup the element */
    scrollStepInPx: number;
    /** Sets the delay to scroll the element */
    delayInMs?: number;
    /** text to be displayed in back to top */
    text: string;
    /** Telemetry Content */
    telemetryContent?: ITelemetryContent;
    /** editProps */
    editProps?: Msdyn365.ITextEditProps | undefined;

}
