// @msdyn365-commerce-modules/core-components/src/modules/default-container/default-container.props.autogenerated.ts
/**
 * Copyright (c) 2018 Microsoft Corporation
 * IDefaultContainer containerModule Interface Properties
 * THIS FILE IS AUTO-GENERATED - MANUAL MODIFICATIONS WILL BE LOST
 */

import * as Msdyn365 from '@msdyn365-commerce/core';
import * as React from 'react';

export const enum layout {
    stacked = 'stacked',
    flow = 'flow'
}

export const enum containerType {
    fluid = 'fluid',
    container = 'container'
}

export const enum childrenWidth {
    auto = 'auto',
    one = 'one',
    two = 'two',
    three = 'three',
    four = 'four',
    six = 'six',
    twelve = 'twelve'
}

export interface IDefaultContainerConfig extends Msdyn365.IModuleConfig {
    heading?: IHeadingData;
    layout?: layout;
    containerType?: containerType;
    childrenWidth?: childrenWidth;
    className?: string;
}

export const enum HeadingTag {
    h1 = 'h1',
    h2 = 'h2',
    h3 = 'h3',
    h4 = 'h4',
    h5 = 'h5',
    h6 = 'h6'
}

export interface IHeadingData {
    headingText?: string;
    tag?: HeadingTag;
}

export interface IDefaultContainerProps<T> extends Msdyn365.IModule<T> {
    config: IDefaultContainerConfig;
    slots: {
        content: React.ReactNode[];
    };
}
