// @msdyn365-commerce-modules/account-management/src/modules/password-reset-verification/password-reset-verification.props.autogenerated.ts
/**
 * Copyright (c) 2018 Microsoft Corporation
 * IPasswordResetVerification contentModule Interface Properties
 * THIS FILE IS AUTO-GENERATED - MANUAL MODIFICATIONS WILL BE LOST
 */

import * as Msdyn365 from '@msdyn365-commerce/core';

export interface IPasswordResetVerificationConfig extends Msdyn365.IModuleConfig {
    heading?: IHeadingData;
    className?: string;
}

export interface IPasswordResetVerificationResources {
    emailAddressLabelText: string;
    verificationCodeLabelText: string;
    sendCodeButtonText: string;
    sendCodeButtonAriaLabel: string;
    resendCodeButtonText: string;
    resendCodeButtonAriaLabel: string;
    verifyCodeButtonText: string;
    verifyCodeButtonAriaLabel: string;
    changeEmailButtonText: string;
    changeEmailButtonAriaLabel: string;
    continueButtonText: string;
    continueButtonAriaLabel: string;
    cancelButtonText: string;
    cancelButtonAriaLabel: string;
    verificationCodeSendSuccess: string;
    emailAddressVerifiedSuccess: string;
    requriedEmailError: string;
    invalidEmailError: string;
    retryError: string;
    retryNotAllowedError: string;
    throttledError: string;
    codeExpiredError: string;
    serverError: string;
    loadingMessage: string;
}

export const enum HeadingTag {
    h1 = 'h1',
    h2 = 'h2',
    h3 = 'h3',
    h4 = 'h4',
    h5 = 'h5',
    h6 = 'h6'
}

export interface IHeadingData {
    text: string;
    tag?: HeadingTag;
}

export interface IPasswordResetVerificationProps<T> extends Msdyn365.IModule<T> {
    resources: IPasswordResetVerificationResources;
    config: IPasswordResetVerificationConfig;
}
