// @msdyn365-commerce-modules/ratings-reviews/src/modules/reviews-list/reviews-list.data.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
import { RatingsSummary, Review, Reviews } from '@msdyn365-commerce/commerce-entities';
import { AsyncResult, SimpleProduct } from '@msdyn365-commerce/retail-proxy';
import { IRatingsReviewsState } from '../../actions/ratings-reviews-state-data-action';

export interface IReviewsListData {
    /**
     * @friendlyName Reviews List
     * @description List of reviews to be displayed on a page.
     */
    reviewsList: AsyncResult<Reviews>;

    /**
     * @description The user review data.
     */
    userReview: AsyncResult<Review>;

    /**
     * @description The product that the reviews are on.
     */
    product: AsyncResult<SimpleProduct>;
    ratingsSummary: AsyncResult<RatingsSummary>;
    ratingsReviewsState: AsyncResult<IRatingsReviewsState>;
}
