// @msdyn365-commerce-modules/order-management/dist/lib/modules/order-details/order-details.view.js
import { Module, Node } from '@msdyn365-commerce-modules/utilities';
import * as React from 'react';
export const OrderDetailsOrderInfomation = ({ orderInformationProps, salesId, receiptId, createdDate, count, amount, channelReferenceId, channelName, channelAddress }) => (React.createElement(Node, Object.assign({}, orderInformationProps),
    channelName,
    channelAddress,
    channelReferenceId,
    salesId,
    receiptId,
    createdDate,
    count,
    amount));
export const OrderDetailsSalesLine = ({ salesLineProps, salesLine, buyAgainButton }) => (React.createElement(Node, Object.assign({}, salesLineProps),
    salesLine,
    buyAgainButton));
export const OrderDetailsGroupDelivery = ({ deliveryProps, heading, count, trackingInfo }) => (React.createElement(Node, Object.assign({}, deliveryProps),
    heading,
    count,
    trackingInfo));
export const OrderDetailsGroup = ({ groupProps, delivery, address, salesLinesProps, salesLines }) => (React.createElement(Node, Object.assign({}, groupProps),
    delivery && React.createElement(OrderDetailsGroupDelivery, Object.assign({}, delivery)),
    salesLines && (React.createElement(Node, Object.assign({}, salesLinesProps), salesLines.map(salesLine => (React.createElement(React.Fragment, { key: salesLine.data.salesLine.LineId },
        React.createElement(OrderDetailsSalesLine, Object.assign({}, salesLine)),
        salesLine.salesStatus))))),
    address));
export const OrderDetailsGroups = ({ groupsProps, groups }) => (React.createElement(Node, Object.assign({}, groupsProps), groups.map((group, index) => (React.createElement(OrderDetailsGroup, Object.assign({ key: index }, group))))));
export const OrderDetailsOrderSummary = ({ orderSummaryProps, heading, subtotal, shipping, tax, totalAmount, earnedPoints }) => (React.createElement(Node, Object.assign({}, orderSummaryProps),
    heading,
    subtotal,
    shipping,
    tax,
    totalAmount,
    earnedPoints));
export const OrderDetailsPayment = ({ paymentMethodsProps, title, methods }) => (React.createElement(Node, Object.assign({}, paymentMethodsProps),
    title,
    methods));
export const OrderDetailsHelp = ({ helpProps, needHelpLabel, helpLineNumberLabel, contactNumber }) => (React.createElement(Node, Object.assign({}, helpProps),
    needHelpLabel,
    helpLineNumberLabel,
    contactNumber));
const OrderDetailsView = ({ moduleProps, heading, alert, loading, orderInfomation, orderSummary, payment, help, groups }) => {
    return (React.createElement(Module, Object.assign({}, moduleProps),
        heading,
        alert,
        loading,
        orderInfomation && React.createElement(OrderDetailsOrderInfomation, Object.assign({}, orderInfomation)),
        groups && React.createElement(OrderDetailsGroups, Object.assign({}, groups)),
        orderSummary && React.createElement(OrderDetailsOrderSummary, Object.assign({}, orderSummary)),
        payment && React.createElement(OrderDetailsPayment, Object.assign({}, payment)),
        help && React.createElement(OrderDetailsHelp, Object.assign({}, help))));
};
export default OrderDetailsView;
